{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import dataclasses\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import toml\n",
    "import pathlib2 as pathlib\n",
    "\n",
    "\n",
    "with open(\"config.toml\") as f:\n",
    "    config = toml.load(f)\n",
    "\n",
    "@dataclasses.dataclass\n",
    "class Thresholds:\n",
    "    GQ: int = config['QC_Thresholds']['GQ']\n",
    "    DP: int = config['QC_Thresholds']['DP']\n",
    "    SIFT: float = config['In_silico_predictions']['SIFT']\n",
    "    PP2: float = config['In_silico_predictions']['PolyPhen-2']\n",
    "    CADD: float = config['In_silico_predictions']['CADD']\n",
    "    MutationTaster: list = dataclasses.field(default_factory=list)\n",
    "    GGM_AF: float = config['MAF_Thresholds']['GGM']\n",
    "\n",
    "# Define the thresholds of 'MutationTaster' as a list of strings\n",
    "Thresholds.MutationTaster = config['In_silico_predictions']['exclude_MutationTaster']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "input_dir = '/Volumes/vol/work/Github/TestData/ggm'\n",
    "input_csv = f\"{input_dir}/ggm_trio_SEL1L.csv\"\n",
    "input_csv = pathlib.Path(input_csv)\n",
    "input_basename = input_csv.stem"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "def _split_qc_col(df: pd.DataFrame) -> pd.DataFrame:\n",
    "    for rel in ['pro', 'pat', 'mat']:\n",
    "        df = pd.concat(\n",
    "            [df, df[f'GQ:DP:AD({rel})'].str.split(':', expand=True)], axis=1)\n",
    "        for i in range(3):\n",
    "            df[i] = df[i].replace('.', np.nan)\n",
    "            df[i] = df[i].replace('-', np.nan)\n",
    "        df = df.astype({0: float, 1: float, 2: float})\n",
    "        df.rename(columns={0: f'GQ({rel})', 1: f'DP({rel})', 2: f'AD({rel})'}, \n",
    "                inplace=True)\n",
    "    return df\n",
    "\n",
    "def _split_ggmacan_col(df: pd.DataFrame) -> pd.DataFrame:\n",
    "    df = pd.concat(\n",
    "        [df, df['GGM(AC/AN)'].str.split('/', expand=True)], axis=1)\n",
    "    df = df.astype({0: float, 1: float})\n",
    "    df.rename(columns={0: 'GGM(AC)', 1: 'GGM(AN)'}, inplace=True)\n",
    "\n",
    "    return df\n",
    "\n",
    "\n",
    "def _add_ggmaf_col(df: pd.DataFrame) -> pd.DataFrame:\n",
    "    df['GGM(AF)'] = df['GGM(AC)'] / df['GGM(AN)']\n",
    "    \n",
    "    return df\n",
    "\n",
    "def _rename_cols(df: pd.DataFrame) -> pd.DataFrame:\n",
    "    rename_dict: dict = {\n",
    "        'Chr': 'CHROM',\n",
    "        'Position': 'POS',\n",
    "        'Ref': 'REF',\n",
    "        'Alt': 'ALT'\n",
    "        }\n",
    "    df.rename(columns=rename_dict, inplace=True)\n",
    "\n",
    "    return df\n",
    "\n",
    "def generate_variant_id(df: pd.DataFrame) -> pd.DataFrame:\n",
    "    df['variant_id'] = df['Chr'] + ':' + \\\n",
    "                       df['Position'] + '-' + df['Ref'] + '-' + df['Alt']\n",
    "    return df\n",
    "\n",
    "def add_qc_filter(df: pd.DataFrame, thresholds: Thresholds) -> pd.DataFrame:\n",
    "    df.loc[\n",
    "        (\n",
    "            (df[f'GQ(pro)'] >= thresholds.GQ) \n",
    "            & (df[f'DP(pro)'] >= thresholds.DP)\n",
    "        ),\n",
    "        'QC_FILTER'] = 'PASS'\n",
    "    \n",
    "    return df\n",
    "\n",
    "def add_insilico_filter(row, thresholds: Thresholds) -> str:\n",
    "    if ((row['SIFT'] >= thresholds.SIFT) \n",
    "        and (row['PolyPhen-2'] <= thresholds.PP2)\n",
    "        and (row['CADD'] < thresholds.CADD)\n",
    "        and (row['MutationTaster'] in thresholds.MutationTaster)):\n",
    "        return 'FAIL' \n",
    "    else:\n",
    "        return 'PASS'\n",
    "\n",
    "def add_identified_filter(df: pd.DataFrame) -> pd.DataFrame:\n",
    "    df.loc[\n",
    "        df['Analysis status'] != 'Identified', \n",
    "        'Not_Identified_FILTER'] = 'PASS'\n",
    "\n",
    "    return df\n",
    "    \n",
    "def add_hard_filter(df: pd.DataFrame) -> pd.DataFrame:\n",
    "    df.loc[\n",
    "        (\n",
    "            (df['QC_FILTER'] == 'PASS')\n",
    "            & (df['Not_Identified_FILTER'] == 'PASS')\n",
    "            & (df['insilico_FILTER'] == 'PASS')\n",
    "        ),\n",
    "        'HARD_FILTER'] = 'PASS'\n",
    "    \n",
    "    return df\n",
    "\n",
    "def add_maf_filter(df: pd.DataFrame, thresholds: Thresholds) -> pd.DataFrame:\n",
    "    df.loc[\n",
    "        df['GGM(AF)'] < thresholds.GGM_AF,\n",
    "        'GGM_FILTER'] = 'PASS'\n",
    "    \n",
    "    return df\n",
    "\n",
    "def preprocess(df: pd.DataFrame) -> pd.DataFrame:\n",
    "    df = _split_qc_col(df)\n",
    "    df = _split_ggmacan_col(df)\n",
    "    df = _add_ggmaf_col(df)\n",
    "    df = generate_variant_id(df)\n",
    "    df.replace({'SIFT': '-', 'PolyPhen-2': '-', 'CADD': '-'}, np.nan, inplace=True)\n",
    "    df = df.astype(\n",
    "        {'SIFT': float, 'PolyPhen-2': float, 'CADD': float, 'GGM(AF)': float}\n",
    "        )\n",
    "\n",
    "    return df\n",
    "\n",
    "def reorder_columns(df: pd.DataFrame) -> pd.DataFrame:\n",
    "    # Drop the 1st column\n",
    "    df = df.drop(columns=df.columns[0])\n",
    "    \n",
    "    # Rename the columns\n",
    "    df = _rename_cols(df)\n",
    "\n",
    "    # Reorder the columns\n",
    "    reodered_columns = [\n",
    "        'Gene', 'Transcript', 'Family', 'Sample', 'Disease', 'Impact',\n",
    "        'Vtype', 'variant_id', 'Amino acid change2', 'Effect',  \n",
    "        'Distance',  'SIFT', 'PolyPhen-2', 'MutationTaster', 'CADD', 'HARD_FILTER', \n",
    "        'gnomAD(AF)', 'ExAC(AF)', 'ToMMo3.5K(AF)', 'GGM(AF)', 'JPNCTL(SC)', \n",
    "        'GGM(AC)', 'gnomAD(AC)', 'ToMMo3.5K(AC)',  \n",
    "        'ID(pro)', 'AS(pro)', 'GT(pro)', 'GQ(pro)', 'DP(pro)', 'AD(pro)', \n",
    "        'ID(pat)', 'AS(pat)', 'GT(pat)', 'GQ(pat)', 'DP(pat)', 'AD(pat)', \n",
    "        'ID(mat)', 'AS(mat)', 'GT(mat)', 'GQ(mat)', 'DP(mat)', 'AD(mat)', \n",
    "        'QC_FILTER', 'Not_Identified_FILTER', 'insilico_FILTER',\n",
    "        'Analysis status', 'Identified gene', 'Variant description', \n",
    "        'CHROM', 'POS', 'REF', 'ALT'\n",
    "        ]\n",
    "    \n",
    "    df = df[reodered_columns]\n",
    "\n",
    "    return df\n",
    "\n",
    "def postprocess(df: pd.DataFrame) -> pd.DataFrame:\n",
    "    df.replace(np.nan, '.', inplace=True)\n",
    "\n",
    "    return df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "df = pd.read_csv(input_csv, header=0, dtype=str)\n",
    "\n",
    "df = preprocess(df)\n",
    "df = add_qc_filter(df, Thresholds())\n",
    "df = add_identified_filter(df)\n",
    "df['insilico_FILTER'] = df.apply(add_insilico_filter, args=(Thresholds(),), axis=1)\n",
    "df = add_hard_filter(df)\n",
    "df = reorder_columns(df)\n",
    "df = postprocess(df)\n",
    "df.to_excel(f\"{input_dir}/{input_basename}.parsed.xlsx\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "wesanno",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.18"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
