{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [],
   "source": [
    "import dataclasses\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import toml\n",
    "import pathlib2 as pathlib\n",
    "\n",
    "\n",
    "with open(\"config.toml\") as f:\n",
    "    config = toml.load(f)\n",
    "\n",
    "@dataclasses.dataclass\n",
    "class Thresholds:\n",
    "    GQ: int = config['QC_Thresholds']['GQ']\n",
    "    DP: int = config['QC_Thresholds']['DP']\n",
    "    AB_het = config['QC_Thresholds']['AB_het']\n",
    "    AB_hom = config['QC_Thresholds']['AB_hom']\n",
    "    SIFT: float = config['In_silico_predictions']['SIFT']\n",
    "    PP2: float = config['In_silico_predictions']['PolyPhen-2']\n",
    "    CADD: float = config['In_silico_predictions']['CADD']\n",
    "    MutationTaster: list = dataclasses.field(default_factory=list)\n",
    "    GGM_AF: float = config['AF_Thresholds']['GGM']\n",
    "\n",
    "\n",
    "# Define the thresholds of 'MutationTaster' as a list of strings\n",
    "Thresholds.MutationTaster = config['In_silico_predictions']['exclude_MutationTaster']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [],
   "source": [
    "input_dir = '/Volumes/vol/work/Github/TestData/ggm'\n",
    "input_csv = f\"{input_dir}/ggm_trio_SEL1L.csv\"\n",
    "input_csv = pathlib.Path(input_csv)\n",
    "input_basename = input_csv.stem"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [],
   "source": [
    "def _split_qc_col(df: pd.DataFrame) -> pd.DataFrame:\n",
    "    for rel in ['pro', 'pat', 'mat']:\n",
    "        df = pd.concat(\n",
    "            [df, df[f'GQ:DP:AD({rel})'].str.split(':', expand=True)], axis=1)\n",
    "        for i in range(3):\n",
    "            df[i] = df[i].replace('.', np.nan)\n",
    "            df[i] = df[i].replace('-', np.nan)\n",
    "        df = df.astype({0: float, 1: float, 2: float})\n",
    "        df.rename(columns={0: f'GQ({rel})', 1: f'DP({rel})', 2: f'AD({rel})'}, \n",
    "                inplace=True)\n",
    "    return df\n",
    "\n",
    "def _add_ab_col(df: pd.DataFrame) -> pd.DataFrame:\n",
    "    for rel in ['pro', 'pat', 'mat']:\n",
    "        df[f'AB({rel})'] = df[f'AD({rel})'] / df[f'DP({rel})']\n",
    "    return df\n",
    "\n",
    "def _split_ggmacan_col(df: pd.DataFrame) -> pd.DataFrame:\n",
    "    df = pd.concat(\n",
    "        [df, df['GGM(AC/AN)'].str.split('/', expand=True)], axis=1)\n",
    "    df = df.astype({0: float, 1: float})\n",
    "    df.rename(columns={0: 'GGM(AC)', 1: 'GGM(AN)'}, inplace=True)\n",
    "\n",
    "    return df\n",
    "\n",
    "\n",
    "def _add_ggmaf_col(df: pd.DataFrame) -> pd.DataFrame:\n",
    "    df['GGM(AF)'] = df['GGM(AC)'] / df['GGM(AN)']\n",
    "    \n",
    "    return df\n",
    "\n",
    "def _rename_cols(df: pd.DataFrame) -> pd.DataFrame:\n",
    "    rename_dict: dict = {\n",
    "        'Chr': 'CHROM',\n",
    "        'Position': 'POS',\n",
    "        'Ref': 'REF',\n",
    "        'Alt': 'ALT'\n",
    "        }\n",
    "    df.rename(columns=rename_dict, inplace=True)\n",
    "\n",
    "    return df\n",
    "\n",
    "def generate_variant_id(df: pd.DataFrame) -> pd.DataFrame:\n",
    "    df['variant_id'] = df['Chr'] + ':' + \\\n",
    "                       df['Position'] + '-' + df['Ref'] + '-' + df['Alt']\n",
    "    return df\n",
    "\n",
    "def add_qc_filter(row, thresholds: Thresholds) -> pd.DataFrame:    \n",
    "    if row['GQ(pro)'] < thresholds.GQ:\n",
    "        return '.'\n",
    "    else:\n",
    "        if row['Vtype'] == 'homo':\n",
    "            if row['AB(pro)'] >= (1 - thresholds.AB_hom):\n",
    "                return 'PASS'\n",
    "            else:\n",
    "                return '.'\n",
    "        else:\n",
    "            if thresholds.AB_het <= row['AB(pro)'] <= (1 - thresholds.AB_het):\n",
    "                return 'PASS'\n",
    "            else:\n",
    "                return '.'\n",
    "\n",
    "def add_insilico_filter(row, thresholds: Thresholds) -> str:\n",
    "    if ((row['SIFT'] >= thresholds.SIFT) \n",
    "        and (row['PolyPhen-2'] <= thresholds.PP2)\n",
    "        and (row['CADD'] < thresholds.CADD)\n",
    "        and (row['MutationTaster'] in thresholds.MutationTaster)):\n",
    "        return 'FAIL' \n",
    "    else:\n",
    "        return 'PASS'\n",
    "\n",
    "def add_identified_filter(df: pd.DataFrame) -> pd.DataFrame:\n",
    "    df.loc[\n",
    "        df['Analysis status'] != 'Identified', \n",
    "        'Not_Identified_FILTER'] = 'PASS'\n",
    "\n",
    "    return df\n",
    "\n",
    "def add_ggmmaf_filter(df: pd.DataFrame, thresholds: Thresholds) -> pd.DataFrame:\n",
    "    df.loc[\n",
    "        df['GGM(AF)'] < thresholds.GGM_AF,\n",
    "        'GGM_FILTER'] = 'PASS'\n",
    "    \n",
    "    return df\n",
    "    \n",
    "def add_hard_filter(df: pd.DataFrame) -> pd.DataFrame:\n",
    "    df.loc[\n",
    "        (\n",
    "            (df['QC_FILTER'] == 'PASS')\n",
    "            & (df['Not_Identified_FILTER'] == 'PASS')\n",
    "            & (df['insilico_FILTER'] == 'PASS')\n",
    "            & (df['GGM_FILTER'] == 'PASS')\n",
    "        ),\n",
    "        'HARD_FILTER'] = 'PASS'\n",
    "    \n",
    "    return df\n",
    "\n",
    "# Pre-processing functions\n",
    "def preprocess(df: pd.DataFrame) -> pd.DataFrame:\n",
    "    df = _split_qc_col(df)\n",
    "    df = _add_ab_col(df)\n",
    "    df = _split_ggmacan_col(df)\n",
    "    df = _add_ggmaf_col(df)\n",
    "    df = generate_variant_id(df)\n",
    "    df.replace({'SIFT': '-', 'PolyPhen-2': '-', 'CADD': '-'}, np.nan, inplace=True)\n",
    "    df = df.astype(\n",
    "        {'SIFT': float, 'PolyPhen-2': float, 'CADD': float, \n",
    "         'GQ(pro)': float, 'AB(pro)': float, 'GGM(AF)': float}\n",
    "        )\n",
    "\n",
    "    return df\n",
    "\n",
    "def reorder_columns(df: pd.DataFrame) -> pd.DataFrame:\n",
    "    # Drop the 1st column\n",
    "    df = df.drop(columns=df.columns[0])\n",
    "    \n",
    "    # Rename the columns\n",
    "    df = _rename_cols(df)\n",
    "\n",
    "    # Reorder the columns\n",
    "    reodered_columns = [\n",
    "        'Gene', 'Transcript', 'Family', 'Sample', 'Disease', \n",
    "        'Vtype', 'variant_id', 'Amino acid change2', 'Effect',  \n",
    "        'Distance',  'SIFT', 'PolyPhen-2', 'MutationTaster', 'CADD', 'HARD_FILTER', \n",
    "        'gnomAD(AF)', 'ExAC(AF)', 'ToMMo3.5K(AF)', 'GGM(AF)', 'JPNCTL(SC)', \n",
    "        'GGM(AC)', 'gnomAD(AC)', 'ToMMo3.5K(AC)',  \n",
    "        'ID(pro)', 'AS(pro)', 'GT(pro)', 'GQ(pro)', 'DP(pro)', 'AD(pro)', 'AB(pro)',\n",
    "        'ID(pat)', 'AS(pat)', 'GT(pat)', 'GQ(pat)', 'DP(pat)', 'AD(pat)', 'AB(pat)',\n",
    "        'ID(mat)', 'AS(mat)', 'GT(mat)', 'GQ(mat)', 'DP(mat)', 'AD(mat)', 'AB(mat)',\n",
    "        'Impact', 'QC_FILTER', 'Not_Identified_FILTER', 'insilico_FILTER',\n",
    "        'Analysis status', 'Identified gene', 'Variant description', \n",
    "        'CHROM', 'POS', 'REF', 'ALT'\n",
    "        ]\n",
    "    \n",
    "    df = df[reodered_columns]\n",
    "\n",
    "    return df\n",
    "\n",
    "def postprocess(df: pd.DataFrame) -> pd.DataFrame:\n",
    "    df.replace(np.nan, '.', inplace=True)\n",
    "\n",
    "    return df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [],
   "source": [
    "df = pd.read_csv(input_csv, header=0, dtype=str)\n",
    "\n",
    "df = preprocess(df)\n",
    "df = add_identified_filter(df)\n",
    "df = add_ggmmaf_filter(df, Thresholds())\n",
    "df['QC_FILTER'] = df.apply(add_qc_filter, args=(Thresholds(),), axis=1)\n",
    "df['insilico_FILTER'] = df.apply(add_insilico_filter, args=(Thresholds(),), axis=1)\n",
    "df = add_hard_filter(df)\n",
    "df = reorder_columns(df)\n",
    "df = postprocess(df)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene</th>\n",
       "      <th>Transcript</th>\n",
       "      <th>Family</th>\n",
       "      <th>Sample</th>\n",
       "      <th>Disease</th>\n",
       "      <th>Impact</th>\n",
       "      <th>Vtype</th>\n",
       "      <th>variant_id</th>\n",
       "      <th>Amino acid change2</th>\n",
       "      <th>Effect</th>\n",
       "      <th>...</th>\n",
       "      <th>QC_FILTER</th>\n",
       "      <th>Not_Identified_FILTER</th>\n",
       "      <th>insilico_FILTER</th>\n",
       "      <th>Analysis status</th>\n",
       "      <th>Identified gene</th>\n",
       "      <th>Variant description</th>\n",
       "      <th>CHROM</th>\n",
       "      <th>POS</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SEL1L</td>\n",
       "      <td>NM_005065.5</td>\n",
       "      <td>28443</td>\n",
       "      <td>Sample_28443</td>\n",
       "      <td>Kyphomelic_dysplasia</td>\n",
       "      <td>MODIFIER</td>\n",
       "      <td>denovo</td>\n",
       "      <td>14:81953889-GA-G</td>\n",
       "      <td>c.1484-8delT</td>\n",
       "      <td>intron_variant</td>\n",
       "      <td>...</td>\n",
       "      <td>.</td>\n",
       "      <td>PASS</td>\n",
       "      <td>PASS</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>14</td>\n",
       "      <td>81953889</td>\n",
       "      <td>GA</td>\n",
       "      <td>G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>SEL1L</td>\n",
       "      <td>NM_005065.5</td>\n",
       "      <td>33398</td>\n",
       "      <td>Sample_33398</td>\n",
       "      <td>Cerebellar_Ataxia</td>\n",
       "      <td>MODIFIER</td>\n",
       "      <td>denovo</td>\n",
       "      <td>14:81953889-G-GAAA</td>\n",
       "      <td>c.1484-7_1484-8insTTT</td>\n",
       "      <td>intron_variant</td>\n",
       "      <td>...</td>\n",
       "      <td>.</td>\n",
       "      <td>PASS</td>\n",
       "      <td>PASS</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>14</td>\n",
       "      <td>81953889</td>\n",
       "      <td>G</td>\n",
       "      <td>GAAA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>SEL1L</td>\n",
       "      <td>NM_005065.5</td>\n",
       "      <td>33703</td>\n",
       "      <td>Sample_33703</td>\n",
       "      <td>Dystonia</td>\n",
       "      <td>MODIFIER</td>\n",
       "      <td>denovo</td>\n",
       "      <td>14:81953889-G-GA</td>\n",
       "      <td>c.1484-7_1484-8insT</td>\n",
       "      <td>intron_variant</td>\n",
       "      <td>...</td>\n",
       "      <td>.</td>\n",
       "      <td>PASS</td>\n",
       "      <td>PASS</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>14</td>\n",
       "      <td>81953889</td>\n",
       "      <td>G</td>\n",
       "      <td>GA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>SEL1L</td>\n",
       "      <td>NM_005065.5</td>\n",
       "      <td>28529</td>\n",
       "      <td>Sample_28529</td>\n",
       "      <td>Hemidystonia_Regression</td>\n",
       "      <td>MODIFIER</td>\n",
       "      <td>denovo</td>\n",
       "      <td>14:81953889-GAA-G</td>\n",
       "      <td>c.1484-9_1484-8delTT</td>\n",
       "      <td>intron_variant</td>\n",
       "      <td>...</td>\n",
       "      <td>.</td>\n",
       "      <td>PASS</td>\n",
       "      <td>PASS</td>\n",
       "      <td>Undetermined</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>14</td>\n",
       "      <td>81953889</td>\n",
       "      <td>GAA</td>\n",
       "      <td>G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>SEL1L</td>\n",
       "      <td>NM_005065.5</td>\n",
       "      <td>29213</td>\n",
       "      <td>Sample_29213</td>\n",
       "      <td>Leigh_syndrome</td>\n",
       "      <td>MODIFIER</td>\n",
       "      <td>denovo</td>\n",
       "      <td>14:81953889-GA-G</td>\n",
       "      <td>c.1484-8delT</td>\n",
       "      <td>intron_variant</td>\n",
       "      <td>...</td>\n",
       "      <td>.</td>\n",
       "      <td>nan</td>\n",
       "      <td>PASS</td>\n",
       "      <td>Identified</td>\n",
       "      <td>MT_ND3</td>\n",
       "      <td>m.10191T&gt;C:p.(S45P)</td>\n",
       "      <td>14</td>\n",
       "      <td>81953889</td>\n",
       "      <td>GA</td>\n",
       "      <td>G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>244</th>\n",
       "      <td>SEL1L</td>\n",
       "      <td>NM_005065.5</td>\n",
       "      <td>13572</td>\n",
       "      <td>Sample_13572</td>\n",
       "      <td>Congenital_scoliosis</td>\n",
       "      <td>LOW</td>\n",
       "      <td>ch_mo</td>\n",
       "      <td>14:81970643-T-C</td>\n",
       "      <td>c.516A&gt;G:p.Glu172Glu</td>\n",
       "      <td>synonymous_variant</td>\n",
       "      <td>...</td>\n",
       "      <td>PASS</td>\n",
       "      <td>PASS</td>\n",
       "      <td>PASS</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>14</td>\n",
       "      <td>81970643</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>245</th>\n",
       "      <td>SEL1L</td>\n",
       "      <td>NM_005065.5</td>\n",
       "      <td>13572</td>\n",
       "      <td>Sample_13572</td>\n",
       "      <td>Congenital_scoliosis</td>\n",
       "      <td>MODIFIER</td>\n",
       "      <td>ch_fa</td>\n",
       "      <td>14:82000003-G-T</td>\n",
       "      <td>c.70+16C&gt;A</td>\n",
       "      <td>intron_variant</td>\n",
       "      <td>...</td>\n",
       "      <td>PASS</td>\n",
       "      <td>PASS</td>\n",
       "      <td>PASS</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>14</td>\n",
       "      <td>82000003</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>246</th>\n",
       "      <td>SEL1L</td>\n",
       "      <td>NM_005065.5</td>\n",
       "      <td>16736</td>\n",
       "      <td>Sample_16736</td>\n",
       "      <td>Pachygyria:Agenesis_of_corpus_callosum</td>\n",
       "      <td>MODIFIER</td>\n",
       "      <td>ch_fa</td>\n",
       "      <td>14:81972400-C-T</td>\n",
       "      <td>c.508+18G&gt;A</td>\n",
       "      <td>intron_variant</td>\n",
       "      <td>...</td>\n",
       "      <td>PASS</td>\n",
       "      <td>PASS</td>\n",
       "      <td>PASS</td>\n",
       "      <td>Undetermined</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>14</td>\n",
       "      <td>81972400</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>247</th>\n",
       "      <td>SEL1L</td>\n",
       "      <td>NM_005065.5</td>\n",
       "      <td>29339</td>\n",
       "      <td>Sample_29339</td>\n",
       "      <td>hyaloideoretinal_degeneration_heart_anormaly</td>\n",
       "      <td>MODIFIER</td>\n",
       "      <td>denovo</td>\n",
       "      <td>14:81953889-GA-G</td>\n",
       "      <td>c.1484-8delT</td>\n",
       "      <td>intron_variant</td>\n",
       "      <td>...</td>\n",
       "      <td>PASS</td>\n",
       "      <td>PASS</td>\n",
       "      <td>PASS</td>\n",
       "      <td>Undetermined</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>14</td>\n",
       "      <td>81953889</td>\n",
       "      <td>GA</td>\n",
       "      <td>G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>248</th>\n",
       "      <td>SEL1L</td>\n",
       "      <td>NM_005065.5</td>\n",
       "      <td>26765</td>\n",
       "      <td>Sample_26765</td>\n",
       "      <td>Epilepsy_complex_partial_seizure</td>\n",
       "      <td>MODIFIER</td>\n",
       "      <td>denovo</td>\n",
       "      <td>14:81953889-G-GA</td>\n",
       "      <td>c.1484-7_1484-8insT</td>\n",
       "      <td>intron_variant</td>\n",
       "      <td>...</td>\n",
       "      <td>.</td>\n",
       "      <td>PASS</td>\n",
       "      <td>PASS</td>\n",
       "      <td>Undetermined</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>14</td>\n",
       "      <td>81953889</td>\n",
       "      <td>G</td>\n",
       "      <td>GA</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>249 rows Ã— 55 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      Gene   Transcript Family        Sample  \\\n",
       "0    SEL1L  NM_005065.5  28443  Sample_28443   \n",
       "1    SEL1L  NM_005065.5  33398  Sample_33398   \n",
       "2    SEL1L  NM_005065.5  33703  Sample_33703   \n",
       "3    SEL1L  NM_005065.5  28529  Sample_28529   \n",
       "4    SEL1L  NM_005065.5  29213  Sample_29213   \n",
       "..     ...          ...    ...           ...   \n",
       "244  SEL1L  NM_005065.5  13572  Sample_13572   \n",
       "245  SEL1L  NM_005065.5  13572  Sample_13572   \n",
       "246  SEL1L  NM_005065.5  16736  Sample_16736   \n",
       "247  SEL1L  NM_005065.5  29339  Sample_29339   \n",
       "248  SEL1L  NM_005065.5  26765  Sample_26765   \n",
       "\n",
       "                                          Disease    Impact   Vtype  \\\n",
       "0                            Kyphomelic_dysplasia  MODIFIER  denovo   \n",
       "1                               Cerebellar_Ataxia  MODIFIER  denovo   \n",
       "2                                        Dystonia  MODIFIER  denovo   \n",
       "3                         Hemidystonia_Regression  MODIFIER  denovo   \n",
       "4                                  Leigh_syndrome  MODIFIER  denovo   \n",
       "..                                            ...       ...     ...   \n",
       "244                          Congenital_scoliosis       LOW   ch_mo   \n",
       "245                          Congenital_scoliosis  MODIFIER   ch_fa   \n",
       "246        Pachygyria:Agenesis_of_corpus_callosum  MODIFIER   ch_fa   \n",
       "247  hyaloideoretinal_degeneration_heart_anormaly  MODIFIER  denovo   \n",
       "248              Epilepsy_complex_partial_seizure  MODIFIER  denovo   \n",
       "\n",
       "             variant_id     Amino acid change2              Effect  ...  \\\n",
       "0      14:81953889-GA-G           c.1484-8delT      intron_variant  ...   \n",
       "1    14:81953889-G-GAAA  c.1484-7_1484-8insTTT      intron_variant  ...   \n",
       "2      14:81953889-G-GA    c.1484-7_1484-8insT      intron_variant  ...   \n",
       "3     14:81953889-GAA-G   c.1484-9_1484-8delTT      intron_variant  ...   \n",
       "4      14:81953889-GA-G           c.1484-8delT      intron_variant  ...   \n",
       "..                  ...                    ...                 ...  ...   \n",
       "244     14:81970643-T-C   c.516A>G:p.Glu172Glu  synonymous_variant  ...   \n",
       "245     14:82000003-G-T             c.70+16C>A      intron_variant  ...   \n",
       "246     14:81972400-C-T            c.508+18G>A      intron_variant  ...   \n",
       "247    14:81953889-GA-G           c.1484-8delT      intron_variant  ...   \n",
       "248    14:81953889-G-GA    c.1484-7_1484-8insT      intron_variant  ...   \n",
       "\n",
       "    QC_FILTER Not_Identified_FILTER insilico_FILTER Analysis status  \\\n",
       "0           .                  PASS            PASS               -   \n",
       "1           .                  PASS            PASS               -   \n",
       "2           .                  PASS            PASS               -   \n",
       "3           .                  PASS            PASS    Undetermined   \n",
       "4           .                   nan            PASS      Identified   \n",
       "..        ...                   ...             ...             ...   \n",
       "244      PASS                  PASS            PASS               -   \n",
       "245      PASS                  PASS            PASS               -   \n",
       "246      PASS                  PASS            PASS    Undetermined   \n",
       "247      PASS                  PASS            PASS    Undetermined   \n",
       "248         .                  PASS            PASS    Undetermined   \n",
       "\n",
       "    Identified gene  Variant description CHROM       POS  REF   ALT  \n",
       "0                 -                    -    14  81953889   GA     G  \n",
       "1                 -                    -    14  81953889    G  GAAA  \n",
       "2                 -                    -    14  81953889    G    GA  \n",
       "3                 -                    -    14  81953889  GAA     G  \n",
       "4            MT_ND3  m.10191T>C:p.(S45P)    14  81953889   GA     G  \n",
       "..              ...                  ...   ...       ...  ...   ...  \n",
       "244               -                    -    14  81970643    T     C  \n",
       "245               -                    -    14  82000003    G     T  \n",
       "246               -                    -    14  81972400    C     T  \n",
       "247               -                    -    14  81953889   GA     G  \n",
       "248               -                    -    14  81953889    G    GA  \n",
       "\n",
       "[249 rows x 55 columns]"
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df.to_excel(f\"{input_dir}/{input_basename}.parsed.xlsx\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "ename": "AttributeError",
     "evalue": "'Series' object has no attribute 'count_values'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mAttributeError\u001b[0m                            Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[18], line 1\u001b[0m\n\u001b[0;32m----> 1\u001b[0m \u001b[43mdf\u001b[49m\u001b[43m[\u001b[49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43mVtype\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m]\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mcount_values\u001b[49m()\n",
      "File \u001b[0;32m~/miniforge3/envs/wesanno/lib/python3.9/site-packages/pandas/core/generic.py:6204\u001b[0m, in \u001b[0;36mNDFrame.__getattr__\u001b[0;34m(self, name)\u001b[0m\n\u001b[1;32m   6197\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m (\n\u001b[1;32m   6198\u001b[0m     name \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;129;01min\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_internal_names_set\n\u001b[1;32m   6199\u001b[0m     \u001b[38;5;129;01mand\u001b[39;00m name \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;129;01min\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_metadata\n\u001b[1;32m   6200\u001b[0m     \u001b[38;5;129;01mand\u001b[39;00m name \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;129;01min\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_accessors\n\u001b[1;32m   6201\u001b[0m     \u001b[38;5;129;01mand\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_info_axis\u001b[38;5;241m.\u001b[39m_can_hold_identifiers_and_holds_name(name)\n\u001b[1;32m   6202\u001b[0m ):\n\u001b[1;32m   6203\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28mself\u001b[39m[name]\n\u001b[0;32m-> 6204\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;43mobject\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[38;5;21;43m__getattribute__\u001b[39;49m\u001b[43m(\u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mname\u001b[49m\u001b[43m)\u001b[49m\n",
      "\u001b[0;31mAttributeError\u001b[0m: 'Series' object has no attribute 'count_values'"
     ]
    }
   ],
   "source": [
    "df['Vtype'].count_values()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "wesanno",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.18"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
