{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import dataclasses\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import toml\n",
    "import pathlib2 as pathlib\n",
    "\n",
    "with open(\"config.toml\") as f:\n",
    "    config = toml.load(f)\n",
    "\n",
    "@dataclasses.dataclass\n",
    "class Thresholds:\n",
    "    GQ: int = config['QC_Thresholds']['GQ']\n",
    "    DP: int = config['QC_Thresholds']['DP']\n",
    "    AB_het = config['QC_Thresholds']['AB_het']\n",
    "    AB_hom = config['QC_Thresholds']['AB_hom']\n",
    "    SIFT: float = config['In_silico_predictions']['SIFT']\n",
    "    PP2: float = config['In_silico_predictions']['PolyPhen-2']\n",
    "    CADD: float = config['In_silico_predictions']['CADD']\n",
    "    MutationTaster: list = dataclasses.field(default_factory=list)\n",
    "    GGM_AF: float = config['AF_Thresholds']['GGM']\n",
    "\n",
    "# Define the thresholds of 'MutationTaster' as a list of strings\n",
    "Thresholds.MutationTaster = config['In_silico_predictions']['exclude_MutationTaster']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "input_dir = '/Volumes/vol/work/Github/TestData/ggm'\n",
    "input_csv = f\"{input_dir}/ggm_trio_BRS3.csv\"\n",
    "input_csv = pathlib.Path(input_csv)\n",
    "input_basename = input_csv.stem"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "def _split_qc_col(df: pd.DataFrame) -> pd.DataFrame:\n",
    "    for rel in ['pro', 'pat', 'mat']:\n",
    "        df = pd.concat(\n",
    "            [df, df[f'GQ:DP:AD({rel})'].str.split(':', expand=True)], axis=1)\n",
    "        for i in range(3):\n",
    "            df[i] = df[i].replace('.', np.nan)\n",
    "            df[i] = df[i].replace('-', np.nan)\n",
    "        df = df.astype({0: float, 1: float, 2: float})\n",
    "        df.rename(columns={0: f'GQ({rel})', 1: f'DP({rel})', 2: f'AD({rel})'}, \n",
    "                inplace=True)\n",
    "    return df\n",
    "\n",
    "def _add_ab_col(df: pd.DataFrame) -> pd.DataFrame:\n",
    "    for rel in ['pro', 'pat', 'mat']:\n",
    "        df[f'AB({rel})'] = df[f'AD({rel})'] / df[f'DP({rel})']\n",
    "    return df\n",
    "\n",
    "def _split_ggmacan_col(df: pd.DataFrame) -> pd.DataFrame:\n",
    "    df = pd.concat(\n",
    "        [df, df['GGM(AC/AN)'].str.split('/', expand=True)], axis=1)\n",
    "    df = df.astype({0: float, 1: float})\n",
    "    df.rename(columns={0: 'GGM(AC)', 1: 'GGM(AN)'}, inplace=True)\n",
    "\n",
    "    return df\n",
    "\n",
    "\n",
    "def _add_ggmaf_col(df: pd.DataFrame) -> pd.DataFrame:\n",
    "    df['GGM(AF)'] = df['GGM(AC)'] / df['GGM(AN)']\n",
    "    \n",
    "    return df\n",
    "\n",
    "def _rename_cols(df: pd.DataFrame) -> pd.DataFrame:\n",
    "    rename_dict: dict = {\n",
    "        'Chr': 'CHROM',\n",
    "        'Position': 'POS',\n",
    "        'Ref': 'REF',\n",
    "        'Alt': 'ALT'\n",
    "        }\n",
    "    df.rename(columns=rename_dict, inplace=True)\n",
    "\n",
    "    return df\n",
    "\n",
    "def generate_variant_id(df: pd.DataFrame) -> pd.DataFrame:\n",
    "    df['variant_id'] = df['Chr'] + ':' + \\\n",
    "                       df['Position'] + '-' + df['Ref'] + '-' + df['Alt']\n",
    "    return df\n",
    "\n",
    "def add_qc_filter(row, thresholds: Thresholds) -> pd.DataFrame:    \n",
    "    if row['GQ(pro)'] < thresholds.GQ:\n",
    "        return '.'\n",
    "    else:\n",
    "        if row['Vtype'] == 'homo':\n",
    "            if row['AB(pro)'] >= (1 - thresholds.AB_hom):\n",
    "                return 'PASS'\n",
    "            else:\n",
    "                return '.'\n",
    "        else:\n",
    "            if thresholds.AB_het <= row['AB(pro)'] <= (1 - thresholds.AB_het):\n",
    "                return 'PASS'\n",
    "            else:\n",
    "                return '.'\n",
    "\n",
    "def add_insilico_filter(row, thresholds: Thresholds) -> str:\n",
    "    if ((row['SIFT'] >= thresholds.SIFT) \n",
    "        and (row['PolyPhen-2'] <= thresholds.PP2)\n",
    "        and (row['CADD'] < thresholds.CADD)\n",
    "        and (row['MutationTaster'] in thresholds.MutationTaster)):\n",
    "        return 'FAIL' \n",
    "    else:\n",
    "        return 'PASS'\n",
    "\n",
    "def add_identified_filter(df: pd.DataFrame) -> pd.DataFrame:\n",
    "    df.loc[\n",
    "        df['Analysis status'] != 'Identified', \n",
    "        'Not_Identified_FILTER'] = 'PASS'\n",
    "\n",
    "    return df\n",
    "\n",
    "def add_ggmmaf_filter(df: pd.DataFrame, thresholds: Thresholds) -> pd.DataFrame:\n",
    "    df.loc[\n",
    "        df['GGM(AF)'] < thresholds.GGM_AF,\n",
    "        'GGM_FILTER'] = 'PASS'\n",
    "    \n",
    "    return df\n",
    "    \n",
    "def add_hard_filter(df: pd.DataFrame) -> pd.DataFrame:\n",
    "    df.loc[\n",
    "        (\n",
    "            (df['QC_FILTER'] == 'PASS')\n",
    "            & (df['Not_Identified_FILTER'] == 'PASS')\n",
    "            & (df['insilico_FILTER'] == 'PASS')\n",
    "            & (df['GGM_FILTER'] == 'PASS')\n",
    "        ),\n",
    "        'HARD_FILTER'] = 'PASS'\n",
    "    \n",
    "    return df\n",
    "\n",
    "# Pre-processing functions\n",
    "def preprocess(df: pd.DataFrame) -> pd.DataFrame:\n",
    "    df = _split_qc_col(df)\n",
    "    df = _add_ab_col(df)\n",
    "    df = _split_ggmacan_col(df)\n",
    "    df = _add_ggmaf_col(df)\n",
    "    df = generate_variant_id(df)\n",
    "    df.replace({'SIFT': '-', 'PolyPhen-2': '-', 'CADD': '-'}, np.nan, inplace=True)\n",
    "    df = df.astype(\n",
    "        {'SIFT': float, 'PolyPhen-2': float, 'CADD': float, \n",
    "         'GQ(pro)': float, 'AB(pro)': float, 'GGM(AF)': float}\n",
    "        )\n",
    "\n",
    "    return df\n",
    "\n",
    "def reorder_columns(df: pd.DataFrame) -> pd.DataFrame:\n",
    "    # Drop the 1st column\n",
    "    df = df.drop(columns=df.columns[0])\n",
    "    \n",
    "    # Rename the columns\n",
    "    df = _rename_cols(df)\n",
    "\n",
    "    # Reorder the columns\n",
    "    reodered_columns = [\n",
    "        'HARD_FILTER', 'Gene', 'Transcript', 'Family', 'Sample', 'Disease', \n",
    "        'Vtype', 'variant_id', 'Amino acid change2', 'Effect',  \n",
    "        'Distance', 'SIFT', 'PolyPhen-2', 'MutationTaster', 'CADD',\n",
    "        'gnomAD(AF)', 'ExAC(AF)', 'ToMMo3.5K(AF)', 'GGM(AF)', 'JPNCTL(SC)', \n",
    "        'GGM(AC)', 'gnomAD(AC)', 'ToMMo3.5K(AC)',  \n",
    "        'ID(pro)', 'AS(pro)', 'GT(pro)', 'GQ(pro)', 'DP(pro)', 'AD(pro)', 'AB(pro)',\n",
    "        'ID(pat)', 'AS(pat)', 'GT(pat)', 'GQ(pat)', 'DP(pat)', 'AD(pat)', 'AB(pat)',\n",
    "        'ID(mat)', 'AS(mat)', 'GT(mat)', 'GQ(mat)', 'DP(mat)', 'AD(mat)', 'AB(mat)',\n",
    "        'Impact', 'QC_FILTER', 'Not_Identified_FILTER', 'insilico_FILTER',\n",
    "        'Analysis status', 'Identified gene', 'Variant description', \n",
    "        'CHROM', 'POS', 'REF', 'ALT'\n",
    "        ]\n",
    "    \n",
    "    df = df[reodered_columns]\n",
    "\n",
    "    return df\n",
    "\n",
    "def postprocess(df: pd.DataFrame) -> pd.DataFrame:\n",
    "    df.replace(np.nan, '.', inplace=True)\n",
    "\n",
    "    return df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "df = pd.read_csv(input_csv, header=0, dtype=str)\n",
    "\n",
    "df = preprocess(df)\n",
    "df = add_identified_filter(df)\n",
    "df = add_ggmmaf_filter(df, Thresholds())\n",
    "df['QC_FILTER'] = df.apply(add_qc_filter, args=(Thresholds(),), axis=1)\n",
    "df['insilico_FILTER'] = df.apply(add_insilico_filter, args=(Thresholds(),), axis=1)\n",
    "df = add_hard_filter(df)\n",
    "df = reorder_columns(df)\n",
    "df = postprocess(df)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Sample</th>\n",
       "      <th>Disease</th>\n",
       "      <th>Vtype</th>\n",
       "      <th>GQ(pro)</th>\n",
       "      <th>DP(pro)</th>\n",
       "      <th>AD(pro)</th>\n",
       "      <th>AB(pro)</th>\n",
       "      <th>CADD</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Sample_21958</td>\n",
       "      <td>Ascending_spastic_paralysis</td>\n",
       "      <td>homo</td>\n",
       "      <td>99.0</td>\n",
       "      <td>77.0</td>\n",
       "      <td>77.0</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>26.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>Sample_26134</td>\n",
       "      <td>Acute_encephalitis_with_refractory_repetitive_...</td>\n",
       "      <td>denovo</td>\n",
       "      <td>23.0</td>\n",
       "      <td>10.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.200000</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>Sample_23722</td>\n",
       "      <td>Epilepsy_West_synd</td>\n",
       "      <td>homo</td>\n",
       "      <td>99.0</td>\n",
       "      <td>39.0</td>\n",
       "      <td>39.0</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>22.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>Sample_16449</td>\n",
       "      <td>Inherited_glycosilation_disorder</td>\n",
       "      <td>homo</td>\n",
       "      <td>99.0</td>\n",
       "      <td>61.0</td>\n",
       "      <td>61.0</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>Sample_2657</td>\n",
       "      <td>Epilepsy:Ohtahara_syndrome</td>\n",
       "      <td>homo</td>\n",
       "      <td>99.0</td>\n",
       "      <td>96.0</td>\n",
       "      <td>96.0</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>21.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>Sample_6093</td>\n",
       "      <td>Autism</td>\n",
       "      <td>denovo</td>\n",
       "      <td>99.0</td>\n",
       "      <td>144.0</td>\n",
       "      <td>40.0</td>\n",
       "      <td>0.277778</td>\n",
       "      <td>35.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>Sample_6093</td>\n",
       "      <td>Autism</td>\n",
       "      <td>denovo</td>\n",
       "      <td>99.0</td>\n",
       "      <td>141.0</td>\n",
       "      <td>37.0</td>\n",
       "      <td>0.262411</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>Sample_4491</td>\n",
       "      <td>Fibular_aplasia_tibial_campomelia_and_oligosyn...</td>\n",
       "      <td>homo</td>\n",
       "      <td>99.0</td>\n",
       "      <td>172.0</td>\n",
       "      <td>172.0</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>25.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>Sample_29017</td>\n",
       "      <td>Hypothalamic_Hamartoma</td>\n",
       "      <td>denovo</td>\n",
       "      <td>20.0</td>\n",
       "      <td>10.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.200000</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>Sample_33720</td>\n",
       "      <td>long_QT_syndrome</td>\n",
       "      <td>denovo</td>\n",
       "      <td>34.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.400000</td>\n",
       "      <td>34.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          Sample                                            Disease   Vtype  \\\n",
       "4   Sample_21958                        Ascending_spastic_paralysis    homo   \n",
       "9   Sample_26134  Acute_encephalitis_with_refractory_repetitive_...  denovo   \n",
       "10  Sample_23722                                 Epilepsy_West_synd    homo   \n",
       "11  Sample_16449                   Inherited_glycosilation_disorder    homo   \n",
       "13   Sample_2657                         Epilepsy:Ohtahara_syndrome    homo   \n",
       "15   Sample_6093                                             Autism  denovo   \n",
       "16   Sample_6093                                             Autism  denovo   \n",
       "19   Sample_4491  Fibular_aplasia_tibial_campomelia_and_oligosyn...    homo   \n",
       "21  Sample_29017                             Hypothalamic_Hamartoma  denovo   \n",
       "27  Sample_33720                                   long_QT_syndrome  denovo   \n",
       "\n",
       "    GQ(pro)  DP(pro)  AD(pro)   AB(pro)  CADD  \n",
       "4      99.0     77.0     77.0  1.000000  26.5  \n",
       "9      23.0     10.0      2.0  0.200000     .  \n",
       "10     99.0     39.0     39.0  1.000000  22.2  \n",
       "11     99.0     61.0     61.0  1.000000     .  \n",
       "13     99.0     96.0     96.0  1.000000  21.3  \n",
       "15     99.0    144.0     40.0  0.277778  35.0  \n",
       "16     99.0    141.0     37.0  0.262411     .  \n",
       "19     99.0    172.0    172.0  1.000000  25.2  \n",
       "21     20.0     10.0      2.0  0.200000     .  \n",
       "27     34.0      5.0      2.0  0.400000  34.0  "
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_hf = df[df['HARD_FILTER'] == 'PASS']\n",
    "df_hf[['Sample', 'Disease', 'Vtype', 'GQ(pro)', 'DP(pro)', 'AD(pro)', 'AB(pro)', 'CADD']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [],
   "source": [
    "df.to_excel(f\"{input_dir}/{input_basename}.parsed.xlsx\", index=False)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "wesanno",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.18"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
