{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "libs.filter.counter\n"
     ]
    }
   ],
   "source": [
    "# Standard modules\n",
    "from dataclasses import dataclass\n",
    "from logging import getLogger, StreamHandler, DEBUG, config\n",
    "import os\n",
    "\n",
    "# Third party modules\n",
    "import gffutils\n",
    "import pybedtools\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "from pathlib2 import Path\n",
    "import pysam\n",
    "from tqdm import tqdm\n",
    "import yaml\n",
    "\n",
    "# Local modules\n",
    "from libs.args_notebook import parser_setting\n",
    "from libs.utils import load_config, OutputSettings\n",
    "from libs.preprocess import PreProcessExomeSummary\n",
    "from libs.modesamples import ModeSamples\n",
    "from libs.annolib.anno import Anno\n",
    "from libs.annolib.genebased import GeneBasedAnno\n",
    "from libs.filter.maffilter import MafFilter\n",
    "from libs.filter.typefilter import TypeFilter\n",
    "from libs.filter.qcfilter import QcFilter\n",
    "from libs.filter.gtfilter import GtFilter\n",
    "from libs.filter.hardfilter import HardFilter\n",
    "from libs.filter.counter import counter\n",
    "\n",
    "# Settings\n",
    "# tqdm.pandas()\n",
    "\n",
    "#----- STEP 0. Logging settings\n",
    "# parent_directory = os.path.dirname(os.path.dirname(__file__))\n",
    "parent_directory = os.path.dirname(os.path.abspath('../__file__'))\n",
    "config_path: str = os.path.join(parent_directory, 'config/logging.yaml')\n",
    "with open(config_path, 'r') as f:\n",
    "    config.dictConfig(yaml.safe_load(f))\n",
    "logger = getLogger(__name__)\n",
    "\n",
    "# http://www.openbioinformatics.org/annovar/download/0wgxR2rIVP/annovar.latest.tar.gz\n",
    "# conda create -n env python=3.9 --yes\n",
    "# pip install jupyter cyvcf2 pybedtools gffutils biopython pathlib2 pyarrow openpyxl pandarallel pysam tqdm pandas scipy numpy pyhpo toml pyyaml \n",
    "# conda install --channel bioconda --channel conda-forge bedtools --yes "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "     input: /Volumes/vol/work/Github/TestData/proband/31070/annovar/exome_summary.20220607_145914.txt\n",
      "     output: None\n",
      "     xhmm: /work/Github/TestData/proband/xhmm/data.segdup.strvar.haplo.deciph.omim.xcnv.gene.uniq\n",
      "     phenotype: None\n",
      "     mode: auto\n",
      "     samples: auto\n",
      "     assembly: hg19\n",
      "     config: /Volumes/vol/work/Github/playground/wesanno/config/config.toml\n",
      "     resources: /Volumes/vol/work/Github/playground/wesanno/resources\n",
      "     no_gnomad: False\n",
      "     no_hgmd: False\n",
      "     no_decipher: False\n",
      "     no_ddg2p: False\n",
      "     no_jarvis: False\n",
      "     no_spliceai: False\n",
      "     no_syno: None\n",
      "     no_alphamissense: False\n",
      "     no_revel: False\n",
      "     no_trap: False\n",
      "     excel_formating: True\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "#-----   STEP 1. Argument settings\n",
    "args = parser_setting()\n",
    "# args['input'] = '/Volumes/vol/work/workspace/30829/exome_summary.20220427_194150.txt'\n",
    "args['input'] = '/Volumes/SSD_480GB/workspace/Github/TestData/proband/32805/exome_summary_20240410_001219.txt'\n",
    "args['input'] = '/Volumes/SSD_480GB/workspace/Github/TestData/proband/32968/exome_summary_20240410_080519.txt'\n",
    "args['input'] = '/Volumes/vol/work/Github/TestData/proband/32097/exome_summary_20240417_201801.txt'\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2024/04/18 01:33:07 [INFO   ] (libs.preprocess) - Drop ALL dot rows columns\n",
      "2024/04/18 01:33:08 [INFO   ] (libs.preprocess) - Extract MAF from each database cols\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Volumes/vol/work/Github/playground/wesanno/wesanno/libs/preprocess.py:26: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  df['HGVD_AF'] = df[col].str.extract('(\\.|[0-1]\\.\\d{,6})')\n",
      "/Volumes/vol/work/Github/playground/wesanno/wesanno/libs/preprocess.py:27: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  df['HGVD_AF'] = df['HGVD_AF'].replace('.', np.nan)\n",
      "/Volumes/vol/work/Github/playground/wesanno/wesanno/libs/preprocess.py:28: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  df['HGVD_AF'] = df['HGVD_AF'].astype(float)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2024/04/18 01:33:08 [INFO   ] (libs.preprocess) - Extract genotypeing info\n",
      "2024/04/18 01:33:08 [INFO   ] (libs.preprocess) - Split QC info\n",
      "2024/04/18 01:33:09 [INFO   ] (libs.preprocess) - Rename gnomAD MAF cols\n",
      "2024/04/18 01:33:11 [INFO   ] (libs.preprocess) - Replace \".\" to np.nan\n",
      "2024/04/18 01:33:11 [INFO   ] (libs.preprocess) - Split ALT column\n",
      "2024/04/18 01:33:13 [INFO   ] (libs.preprocess) - Extract SpliceAI\n",
      "2024/04/18 01:33:14 [INFO   ] (libs.preprocess) - Drop unused columns\n"
     ]
    }
   ],
   "source": [
    "\n",
    "#-----   STEP 2. Input file and Output settings\n",
    "df = pd.read_table(args['input'], header=0, dtype=str)\n",
    "configs: dict = load_config(args['config'])\n",
    "\n",
    "\n",
    "#-----   STEP 3. Get Mode and Samples information\n",
    "modesamples = ModeSamples(df=df, args=args)\n",
    "mode_samples_info = modesamples.get_mode_samples_info()\n",
    "logger.info(f\"mode_samples_info: {mode_samples_info.mode}\")\n",
    "\n",
    "#-----   STEP 4. Output settings\n",
    "output_settings = OutputSettings(\n",
    "    args=args, mode_samples_info=mode_samples_info)\n",
    "output_file_path = output_settings.get_saving_file_path()\n",
    "\n",
    "#-----   STEP 5. Pre-processing\n",
    "preprocessing = PreProcessExomeSummary(\n",
    "    df=df, args=args, mode_samples_info=mode_samples_info)\n",
    "df = preprocessing.all_pre_processing()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2024/04/18 01:33:16 [INFO   ] (libs.filter.qcfilter) - FILTER cutoffs: GQ >= 20  DP >= 10\n"
     ]
    }
   ],
   "source": [
    "#-----   STEP 6. Annotation\n",
    "gba = GeneBasedAnno(args)\n",
    "df = gba.anno_hgmd(df=df)\n",
    "df = gba.dm_filter(df=df)\n",
    "df = gba.anno_dcpr(df=df)\n",
    "df['MOI'] = df.apply(gba.summarize_moi, axis=1)\n",
    "\n",
    "\n",
    "#-----   STEP 7. Filtering\n",
    "maffilter = MafFilter(\n",
    "    df=df, mode_samples_info=mode_samples_info, configs=configs)\n",
    "df = maffilter.all_filtering()\n",
    "\n",
    "typefilter = TypeFilter(df=df)\n",
    "df = typefilter.exclude_hlamuc_and_exonicsyno()\n",
    "\n",
    "qcfilter = QcFilter(df=df, configs=configs)\n",
    "df = qcfilter.exclude_low_quality()\n",
    "\n",
    "gtfilter = GtFilter(\n",
    "    df=df, mode_samples_info=mode_samples_info)\n",
    "dfs = gtfilter.genotypeing_filter()\n",
    "\n",
    "\n",
    "#-----  Variant reduction and count variants of filtering process\n",
    "countsummery_file = str(Path(output_file_path).parent) + '/CountSummary.xlsx'\n",
    "dfs_filtered = counter(dfs=dfs, output_excel=countsummery_file)\n",
    "\n",
    "\n",
    "#-----   STEP 8. Hard filtering\n",
    "hard = HardFilter(dfs=dfs_filtered)\n",
    "dfs = hard.hard_filtering()\n",
    "\n",
    "\n",
    "#-----   STEP 8. Output as an Excel \n",
    "def df_to_excel(dfs: dataclass, output_xlsx) -> None:\n",
    "    sheet_names = ['AD', 'Homo', 'CH', 'XL']\n",
    "    with pd.ExcelWriter(output_xlsx) as writer:\n",
    "        for df, sheet_name in zip([dfs.AD, dfs.Hm, dfs.CH, dfs.XL], sheet_names):\n",
    "            df.to_excel(writer, sheet_name=sheet_name, index=False)\n",
    "\n",
    "df_to_excel(dfs, f\"{output_file_path}.xlsx\")\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chr</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Ref</th>\n",
       "      <th>Alt</th>\n",
       "      <th>Func.refGene</th>\n",
       "      <th>Gene.refGene</th>\n",
       "      <th>GeneDetail.refGene</th>\n",
       "      <th>ExonicFunc.refGene</th>\n",
       "      <th>AAChange.refGene</th>\n",
       "      <th>...</th>\n",
       "      <th>MAF_0.1%_FILTER</th>\n",
       "      <th>MAF_1%_FILTER</th>\n",
       "      <th>FlaggedSNP_FILTER</th>\n",
       "      <th>HLAMUC_FILTER</th>\n",
       "      <th>ExonicSyno_FILTER</th>\n",
       "      <th>GQ_FILTER</th>\n",
       "      <th>Hetero_count</th>\n",
       "      <th>GT_FILTER</th>\n",
       "      <th>MOI_FILTER</th>\n",
       "      <th>HARD_FILTER</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1571</th>\n",
       "      <td>1</td>\n",
       "      <td>150529731</td>\n",
       "      <td>150529731</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>exonic;splicing</td>\n",
       "      <td>ADAMTSL4;ADAMTSL4</td>\n",
       "      <td>NM_001288607:exon13:c.1857-7A&gt;G</td>\n",
       "      <td>nonsynonymous SNV</td>\n",
       "      <td>ADAMTSL4:NM_019032:exon12:c.A1967G:p.H656R,ADA...</td>\n",
       "      <td>...</td>\n",
       "      <td>PASS</td>\n",
       "      <td>PASS</td>\n",
       "      <td>PASS</td>\n",
       "      <td>PASS</td>\n",
       "      <td>PASS</td>\n",
       "      <td>PASS</td>\n",
       "      <td>1</td>\n",
       "      <td>PASS</td>\n",
       "      <td>PASS</td>\n",
       "      <td>PASS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1578</th>\n",
       "      <td>1</td>\n",
       "      <td>150940556</td>\n",
       "      <td>150940556</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>splicing</td>\n",
       "      <td>CERS2</td>\n",
       "      <td>NM_181746:exon4:c.410+3G&gt;T;NM_022075:exon4:c.4...</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>...</td>\n",
       "      <td>PASS</td>\n",
       "      <td>PASS</td>\n",
       "      <td>PASS</td>\n",
       "      <td>PASS</td>\n",
       "      <td>PASS</td>\n",
       "      <td>PASS</td>\n",
       "      <td>1</td>\n",
       "      <td>PASS</td>\n",
       "      <td>PASS</td>\n",
       "      <td>PASS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6023</th>\n",
       "      <td>3</td>\n",
       "      <td>132036365</td>\n",
       "      <td>132036365</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>exonic</td>\n",
       "      <td>ACPP</td>\n",
       "      <td>.</td>\n",
       "      <td>nonsynonymous SNV</td>\n",
       "      <td>ACPP:NM_001099:exon1:c.T65C:p.L22P,ACPP:NM_001...</td>\n",
       "      <td>...</td>\n",
       "      <td>PASS</td>\n",
       "      <td>PASS</td>\n",
       "      <td>PASS</td>\n",
       "      <td>PASS</td>\n",
       "      <td>PASS</td>\n",
       "      <td>PASS</td>\n",
       "      <td>1</td>\n",
       "      <td>PASS</td>\n",
       "      <td>PASS</td>\n",
       "      <td>PASS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7054</th>\n",
       "      <td>4</td>\n",
       "      <td>85678317</td>\n",
       "      <td>85678321</td>\n",
       "      <td>AAGAG</td>\n",
       "      <td>-</td>\n",
       "      <td>splicing</td>\n",
       "      <td>WDFY3</td>\n",
       "      <td>NM_014991:exon33:r.spl</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>...</td>\n",
       "      <td>PASS</td>\n",
       "      <td>PASS</td>\n",
       "      <td>PASS</td>\n",
       "      <td>PASS</td>\n",
       "      <td>PASS</td>\n",
       "      <td>PASS</td>\n",
       "      <td>1</td>\n",
       "      <td>PASS</td>\n",
       "      <td>PASS</td>\n",
       "      <td>PASS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8205</th>\n",
       "      <td>5</td>\n",
       "      <td>110456178</td>\n",
       "      <td>110456178</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>splicing</td>\n",
       "      <td>WDR36</td>\n",
       "      <td>NM_139281:exon18:c.2073-16T&gt;C</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>...</td>\n",
       "      <td>PASS</td>\n",
       "      <td>PASS</td>\n",
       "      <td>PASS</td>\n",
       "      <td>PASS</td>\n",
       "      <td>PASS</td>\n",
       "      <td>PASS</td>\n",
       "      <td>1</td>\n",
       "      <td>PASS</td>\n",
       "      <td>PASS</td>\n",
       "      <td>PASS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13358</th>\n",
       "      <td>9</td>\n",
       "      <td>21994323</td>\n",
       "      <td>21994323</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>exonic</td>\n",
       "      <td>CDKN2A</td>\n",
       "      <td>.</td>\n",
       "      <td>nonsynonymous SNV</td>\n",
       "      <td>CDKN2A:NM_058195:exon1:c.G8A:p.R3H</td>\n",
       "      <td>...</td>\n",
       "      <td>PASS</td>\n",
       "      <td>PASS</td>\n",
       "      <td>PASS</td>\n",
       "      <td>PASS</td>\n",
       "      <td>PASS</td>\n",
       "      <td>PASS</td>\n",
       "      <td>1</td>\n",
       "      <td>PASS</td>\n",
       "      <td>PASS</td>\n",
       "      <td>PASS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22095</th>\n",
       "      <td>16</td>\n",
       "      <td>30709488</td>\n",
       "      <td>30709488</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>exonic</td>\n",
       "      <td>LOC730183</td>\n",
       "      <td>.</td>\n",
       "      <td>nonsynonymous SNV</td>\n",
       "      <td>LOC730183:NM_001256932:exon1:c.C142T:p.R48C</td>\n",
       "      <td>...</td>\n",
       "      <td>PASS</td>\n",
       "      <td>PASS</td>\n",
       "      <td>PASS</td>\n",
       "      <td>PASS</td>\n",
       "      <td>PASS</td>\n",
       "      <td>PASS</td>\n",
       "      <td>1</td>\n",
       "      <td>PASS</td>\n",
       "      <td>PASS</td>\n",
       "      <td>PASS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25224</th>\n",
       "      <td>19</td>\n",
       "      <td>3959260</td>\n",
       "      <td>3959260</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>exonic</td>\n",
       "      <td>DAPK3</td>\n",
       "      <td>.</td>\n",
       "      <td>nonsynonymous SNV</td>\n",
       "      <td>DAPK3:NM_001348:exon9:c.G1204A:p.A402T</td>\n",
       "      <td>...</td>\n",
       "      <td>PASS</td>\n",
       "      <td>PASS</td>\n",
       "      <td>PASS</td>\n",
       "      <td>PASS</td>\n",
       "      <td>PASS</td>\n",
       "      <td>PASS</td>\n",
       "      <td>1</td>\n",
       "      <td>PASS</td>\n",
       "      <td>PASS</td>\n",
       "      <td>PASS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25684</th>\n",
       "      <td>19</td>\n",
       "      <td>12740051</td>\n",
       "      <td>12740051</td>\n",
       "      <td>C</td>\n",
       "      <td>-</td>\n",
       "      <td>exonic</td>\n",
       "      <td>ZNF791</td>\n",
       "      <td>.</td>\n",
       "      <td>frameshift deletion</td>\n",
       "      <td>ZNF791:NM_153358:exon4:c.1708delC:p.H570Ifs*2</td>\n",
       "      <td>...</td>\n",
       "      <td>PASS</td>\n",
       "      <td>PASS</td>\n",
       "      <td>PASS</td>\n",
       "      <td>PASS</td>\n",
       "      <td>PASS</td>\n",
       "      <td>PASS</td>\n",
       "      <td>1</td>\n",
       "      <td>PASS</td>\n",
       "      <td>PASS</td>\n",
       "      <td>PASS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25925</th>\n",
       "      <td>19</td>\n",
       "      <td>18895130</td>\n",
       "      <td>18895130</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>exonic</td>\n",
       "      <td>COMP</td>\n",
       "      <td>.</td>\n",
       "      <td>nonsynonymous SNV</td>\n",
       "      <td>COMP:NM_000095:exon17:c.C1958T:p.A653V</td>\n",
       "      <td>...</td>\n",
       "      <td>PASS</td>\n",
       "      <td>PASS</td>\n",
       "      <td>PASS</td>\n",
       "      <td>PASS</td>\n",
       "      <td>PASS</td>\n",
       "      <td>PASS</td>\n",
       "      <td>1</td>\n",
       "      <td>PASS</td>\n",
       "      <td>PASS</td>\n",
       "      <td>PASS</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>10 rows × 224 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      Chr      Start        End    Ref Alt     Func.refGene  \\\n",
       "1571    1  150529731  150529731      A   G  exonic;splicing   \n",
       "1578    1  150940556  150940556      C   A         splicing   \n",
       "6023    3  132036365  132036365      T   C           exonic   \n",
       "7054    4   85678317   85678321  AAGAG   -         splicing   \n",
       "8205    5  110456178  110456178      T   C         splicing   \n",
       "13358   9   21994323   21994323      C   T           exonic   \n",
       "22095  16   30709488   30709488      G   A           exonic   \n",
       "25224  19    3959260    3959260      C   T           exonic   \n",
       "25684  19   12740051   12740051      C   -           exonic   \n",
       "25925  19   18895130   18895130      G   A           exonic   \n",
       "\n",
       "            Gene.refGene                                 GeneDetail.refGene  \\\n",
       "1571   ADAMTSL4;ADAMTSL4                    NM_001288607:exon13:c.1857-7A>G   \n",
       "1578               CERS2  NM_181746:exon4:c.410+3G>T;NM_022075:exon4:c.4...   \n",
       "6023                ACPP                                                  .   \n",
       "7054               WDFY3                             NM_014991:exon33:r.spl   \n",
       "8205               WDR36                      NM_139281:exon18:c.2073-16T>C   \n",
       "13358             CDKN2A                                                  .   \n",
       "22095          LOC730183                                                  .   \n",
       "25224              DAPK3                                                  .   \n",
       "25684             ZNF791                                                  .   \n",
       "25925               COMP                                                  .   \n",
       "\n",
       "        ExonicFunc.refGene                                   AAChange.refGene  \\\n",
       "1571     nonsynonymous SNV  ADAMTSL4:NM_019032:exon12:c.A1967G:p.H656R,ADA...   \n",
       "1578                     .                                                  .   \n",
       "6023     nonsynonymous SNV  ACPP:NM_001099:exon1:c.T65C:p.L22P,ACPP:NM_001...   \n",
       "7054                     .                                                  .   \n",
       "8205                     .                                                  .   \n",
       "13358    nonsynonymous SNV                 CDKN2A:NM_058195:exon1:c.G8A:p.R3H   \n",
       "22095    nonsynonymous SNV        LOC730183:NM_001256932:exon1:c.C142T:p.R48C   \n",
       "25224    nonsynonymous SNV             DAPK3:NM_001348:exon9:c.G1204A:p.A402T   \n",
       "25684  frameshift deletion      ZNF791:NM_153358:exon4:c.1708delC:p.H570Ifs*2   \n",
       "25925    nonsynonymous SNV             COMP:NM_000095:exon17:c.C1958T:p.A653V   \n",
       "\n",
       "       ... MAF_0.1%_FILTER MAF_1%_FILTER FlaggedSNP_FILTER HLAMUC_FILTER  \\\n",
       "1571   ...            PASS          PASS              PASS          PASS   \n",
       "1578   ...            PASS          PASS              PASS          PASS   \n",
       "6023   ...            PASS          PASS              PASS          PASS   \n",
       "7054   ...            PASS          PASS              PASS          PASS   \n",
       "8205   ...            PASS          PASS              PASS          PASS   \n",
       "13358  ...            PASS          PASS              PASS          PASS   \n",
       "22095  ...            PASS          PASS              PASS          PASS   \n",
       "25224  ...            PASS          PASS              PASS          PASS   \n",
       "25684  ...            PASS          PASS              PASS          PASS   \n",
       "25925  ...            PASS          PASS              PASS          PASS   \n",
       "\n",
       "      ExonicSyno_FILTER GQ_FILTER Hetero_count  GT_FILTER  MOI_FILTER  \\\n",
       "1571               PASS      PASS            1       PASS        PASS   \n",
       "1578               PASS      PASS            1       PASS        PASS   \n",
       "6023               PASS      PASS            1       PASS        PASS   \n",
       "7054               PASS      PASS            1       PASS        PASS   \n",
       "8205               PASS      PASS            1       PASS        PASS   \n",
       "13358              PASS      PASS            1       PASS        PASS   \n",
       "22095              PASS      PASS            1       PASS        PASS   \n",
       "25224              PASS      PASS            1       PASS        PASS   \n",
       "25684              PASS      PASS            1       PASS        PASS   \n",
       "25925              PASS      PASS            1       PASS        PASS   \n",
       "\n",
       "       HARD_FILTER  \n",
       "1571          PASS  \n",
       "1578          PASS  \n",
       "6023          PASS  \n",
       "7054          PASS  \n",
       "8205          PASS  \n",
       "13358         PASS  \n",
       "22095         PASS  \n",
       "25224         PASS  \n",
       "25684         PASS  \n",
       "25925         PASS  \n",
       "\n",
       "[10 rows x 224 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dfs.AD[dfs.AD['HARD_FILTER'] == 'PASS']"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Memo, Test & Dev."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "import glob\n",
    "\n",
    "class GeneBasedAnno:\n",
    "    def __init__(self, args: dict):\n",
    "        self._root_path = args['resources']\n",
    "        self.path_to_hgmd = glob.glob(f'{self._root_path}/HGMD*.tsv.gz')\n",
    "        self.path_to_ddg2p = glob.glob(f'{self._root_path}/DDG2P*.gz')\n",
    "        self.path_to_eyeg2p = glob.glob(f'{self._root_path}/EyeG2P*.gz')\n",
    "        self.path_to_sking2p = glob.glob(f'{self._root_path}/SkinG2P*.gz')\n",
    "        self.path_to_cancer2p = glob.glob(f'{self._root_path}/CancerG2P*.gz')\n",
    "        self.path_to_cardiacg2p = glob.glob(f'{self._root_path}/CardiacG2P*.gz')\n",
    "        self.path_to_skeletalg2p = glob.glob(f'{self._root_path}/SkeletalG2P*.gz')\n",
    "\n",
    "    def anno_hgmd(self, df: pd.DataFrame) -> pd.DataFrame:\n",
    "        hgmd = pd.read_table(self.path_to_hgmd[0], header=0, dtype=str)\n",
    "        hgmd = hgmd[\n",
    "            ['gene', 'altsymbol', 'refseq', \n",
    "             'expected_inheritance', 'hgncID', 'omimid', 'DM']\n",
    "             ]\n",
    "        df = pd.merge(\n",
    "            df, hgmd, left_on='Gene.refGene', right_on='gene', how='left'\n",
    "            )\n",
    "        df = df.drop(columns=['gene'])\n",
    "\n",
    "        return df\n",
    "    \n",
    "    def anno_dcpr(self, df: pd.DataFrame) -> pd.DataFrame:\n",
    "        usecols = ['gene symbol', 'disease name', 'allelic requirement',\n",
    "                   'hgnc id']\n",
    "        dd = pd.read_csv(self.path_to_ddg2p[0], header=0, dtype=str)\n",
    "        eye = pd.read_csv(self.path_to_eyeg2p[0], header=0, dtype=str)\n",
    "        ski = pd.read_csv(self.path_to_sking2p[0], header=0, dtype=str)\n",
    "        can = pd.read_csv(self.path_to_cancer2p[0], header=0, dtype=str)\n",
    "        car = pd.read_csv(self.path_to_cardiacg2p[0], header=0, dtype=str)\n",
    "        ske = pd.read_csv(self.path_to_skeletalg2p[0], header=0, dtype=str)\n",
    "\n",
    "        pass\n",
    "\n",
    "\n",
    "    def anno_gnomad(self, da):\n",
    "        pass\n",
    "\n",
    "\n",
    "    def match_g2p_phenotypes(self, df):\n",
    "        pass"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pyhpo"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "gba = GeneBasedAnno(args)\n",
    "df = gba.anno_hgmd(df=df)\n",
    "df = gba.anno_dcpr(df=df)\n",
    "df['MOI'] = df.apply(gba.summarize_moi, axis=1)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "df['AB'] = df['AD'].str.split(',', expand=True)[1].astype(int) / df['DP'].astype(int)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>AD</th>\n",
       "      <th>DP</th>\n",
       "      <th>GT_Pro</th>\n",
       "      <th>AB (Allele Balance)</th>\n",
       "      <th>AB</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>7,9</td>\n",
       "      <td>28</td>\n",
       "      <td>0/1</td>\n",
       "      <td>0.562500</td>\n",
       "      <td>0.321429</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>14,9</td>\n",
       "      <td>23</td>\n",
       "      <td>0/1</td>\n",
       "      <td>0.391304</td>\n",
       "      <td>0.391304</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>9,5</td>\n",
       "      <td>14</td>\n",
       "      <td>0/1</td>\n",
       "      <td>0.357143</td>\n",
       "      <td>0.357143</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>58,26</td>\n",
       "      <td>84</td>\n",
       "      <td>0/1</td>\n",
       "      <td>0.309524</td>\n",
       "      <td>0.309524</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>32,12</td>\n",
       "      <td>44</td>\n",
       "      <td>0/1</td>\n",
       "      <td>0.272727</td>\n",
       "      <td>0.272727</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7856</th>\n",
       "      <td>53,63</td>\n",
       "      <td>116</td>\n",
       "      <td>0/1</td>\n",
       "      <td>0.543103</td>\n",
       "      <td>0.543103</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7860</th>\n",
       "      <td>13,6</td>\n",
       "      <td>23</td>\n",
       "      <td>0/1</td>\n",
       "      <td>0.315789</td>\n",
       "      <td>0.260870</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7861</th>\n",
       "      <td>80,28</td>\n",
       "      <td>108</td>\n",
       "      <td>0/1</td>\n",
       "      <td>0.259259</td>\n",
       "      <td>0.259259</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7862</th>\n",
       "      <td>36,55</td>\n",
       "      <td>91</td>\n",
       "      <td>0/1</td>\n",
       "      <td>0.604396</td>\n",
       "      <td>0.604396</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7863</th>\n",
       "      <td>35,41</td>\n",
       "      <td>76</td>\n",
       "      <td>0/1</td>\n",
       "      <td>0.539474</td>\n",
       "      <td>0.539474</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5404 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         AD   DP GT_Pro  AB (Allele Balance)        AB\n",
       "5       7,9   28    0/1             0.562500  0.321429\n",
       "6      14,9   23    0/1             0.391304  0.391304\n",
       "7       9,5   14    0/1             0.357143  0.357143\n",
       "12    58,26   84    0/1             0.309524  0.309524\n",
       "13    32,12   44    0/1             0.272727  0.272727\n",
       "...     ...  ...    ...                  ...       ...\n",
       "7856  53,63  116    0/1             0.543103  0.543103\n",
       "7860   13,6   23    0/1             0.315789  0.260870\n",
       "7861  80,28  108    0/1             0.259259  0.259259\n",
       "7862  36,55   91    0/1             0.604396  0.604396\n",
       "7863  35,41   76    0/1             0.539474  0.539474\n",
       "\n",
       "[5404 rows x 5 columns]"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.loc[df['GT_Pro'] == '0/1', ['AD', 'DP', 'GT_Pro', 'AB (Allele Balance)', 'AB']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "MOI\n",
       ".     6780\n",
       "AR     608\n",
       "AD     448\n",
       "XL      26\n",
       "YL       1\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df['MOI'].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "metadata": {},
   "outputs": [],
   "source": [
    "def summarize_allelic_requirement(row, not_ad, not_ar, not_xl, not_yl):\n",
    "    inheritance = row['expected_inheritance']\n",
    "    alleleinfo = row['G2P_alleleinfo']\n",
    "    \n",
    "    if inheritance == 'AD':\n",
    "        return 'AD' if not any(word in alleleinfo for word in not_ad) else '.'\n",
    "    elif inheritance == 'AR':\n",
    "        return 'AR' if not any(word in alleleinfo for word in not_ar) else '.'\n",
    "    elif inheritance in ['XLD', 'XLR']:\n",
    "        return 'XL' if not any(word in alleleinfo for word in not_xl) else '.'\n",
    "    elif inheritance == 'YL':\n",
    "        return 'YL' if not any(word in alleleinfo for word in not_yl) else '.'\n",
    "    else:\n",
    "        return '.'\n",
    "\n",
    "df['MOI'] = df.apply(lambda row: summarize_allelic_requirement(row, not_ad, not_ar, not_xl, not_yl), axis=1)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [],
   "source": [
    "df['check'] = df['G2P_alleleinfo'].apply(lambda x: any(word in x for word in not_ad))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#-----   STEP 6. Additional annotations\n",
    "anno = Anno(df=df, args=args)\n",
    "df = anno.anno_scores()\n",
    "\n",
    "#-----   STEP 7. Filtering\n",
    "maffilter = MafFilter(\n",
    "    df=df, mode_samples_info=mode_samples_info, config=configs)\n",
    "df = maffilter.all_filtering()\n",
    "\n",
    "typefilter = TypeFilter(df=df)\n",
    "df = typefilter.exclude_hlamuc_and_exonicsyno()\n",
    "\n",
    "gtfilter = GtFilter(\n",
    "    df=df, mode_samples_info=mode_samples_info)\n",
    "dfs = gtfilter.genotypeing_filter()\n",
    "\n",
    "#-----   STEP 8. Output as an Excel \n",
    "def df_to_excel(dfs: dataclass, output_xlsx) -> None:\n",
    "    sheet_names = ['AD', 'Homo', 'CH', 'XL']\n",
    "    with pd.ExcelWriter(output_xlsx) as writer:\n",
    "        for df, sheet_name in zip([dfs.AD, dfs.Hm, dfs.CH, dfs.XL], sheet_names):\n",
    "            df.to_excel(writer, sheet_name=sheet_name, index=False)\n",
    "\n",
    "df_to_excel(dfs, f\"{output_file_path}.xlsx\")\n",
    "\n",
    "############################################\n",
    "import pickle\n",
    "with open(f\"{output_file_path}.pkl\", mode=\"wb\") as f:\n",
    "    pickle.dump(dfs, f)\n",
    "############################################\n",
    "\n",
    "#-----   STEP (Final Step). Count variants of filtering process\n",
    "countsummery_file = str(Path(output_file_path).parent) + '/CountSummary.xlsx'\n",
    "counter_result = counter(dfs=dfs, output_excel=countsummery_file)\n",
    "print(counter_result)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'libs.filter.gtfilter.ModelDataFrame'>\n"
     ]
    }
   ],
   "source": [
    "import pickle\n",
    "pkl = '/work/Github/TestData/trio/29881/Sample_29881-trio_results/Sample_29881-trio.tsv.pkl'\n",
    "with open(pkl, mode='rb') as f:\n",
    "    dfs = pickle.load(f)\n",
    "\n",
    "print(type(dfs))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "hgmd_pkl = '/work/resources/HGMD/hgmd_info_2023.3'\n",
    "hgmd = pd.read_pickle(hgmd_pkl)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>altsymbol</th>\n",
       "      <th>refseq</th>\n",
       "      <th>expected_inheritance</th>\n",
       "      <th>hgncID</th>\n",
       "      <th>omimid</th>\n",
       "      <th>DFP</th>\n",
       "      <th>DM</th>\n",
       "      <th>DM?</th>\n",
       "      <th>DP</th>\n",
       "      <th>FP</th>\n",
       "      <th>R</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>RBFOX1</td>\n",
       "      <td>2BP1|A2BP1|FOX-1|FOX1|HRNBP1</td>\n",
       "      <td>NM_145891.3</td>\n",
       "      <td>AD</td>\n",
       "      <td>18222</td>\n",
       "      <td>605104</td>\n",
       "      <td>0.0</td>\n",
       "      <td>15.0</td>\n",
       "      <td>47.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ABCA3</td>\n",
       "      <td>ABC-C|ABC3|EST111653|LBM180|SMDP3</td>\n",
       "      <td>NM_001089.3</td>\n",
       "      <td>AR</td>\n",
       "      <td>33</td>\n",
       "      <td>601615</td>\n",
       "      <td>0.0</td>\n",
       "      <td>230.0</td>\n",
       "      <td>152.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AKAP13</td>\n",
       "      <td>AKAP-13|AKAP-Lbc|ARHGEF13|BRX|c-lbc|HA-3|Ht31|...</td>\n",
       "      <td>NM_007200.5</td>\n",
       "      <td>UNK</td>\n",
       "      <td>371</td>\n",
       "      <td>604686</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>11.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>GSS</td>\n",
       "      <td>GSHS|HEL-S-64p|HEL-S-88n</td>\n",
       "      <td>NM_000178.4</td>\n",
       "      <td>AR</td>\n",
       "      <td>4624</td>\n",
       "      <td>601002</td>\n",
       "      <td>0.0</td>\n",
       "      <td>40.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>BRF1</td>\n",
       "      <td>BRF|BRF-1|CFDS|GTF3B|hBRF|HEL-S-76p|TAF3B2|TAF...</td>\n",
       "      <td>NM_001519.4</td>\n",
       "      <td>ADAR</td>\n",
       "      <td>11551</td>\n",
       "      <td>604902</td>\n",
       "      <td>0.0</td>\n",
       "      <td>19.0</td>\n",
       "      <td>15.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17668</th>\n",
       "      <td>TNFRSF21</td>\n",
       "      <td>BM-018|CD358|DR6</td>\n",
       "      <td>NM_014452.5</td>\n",
       "      <td>UNK</td>\n",
       "      <td>13469</td>\n",
       "      <td>605732</td>\n",
       "      <td>0.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>7.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17669</th>\n",
       "      <td>C12orf4</td>\n",
       "      <td>MRT66</td>\n",
       "      <td>NM_020374.4</td>\n",
       "      <td>AR</td>\n",
       "      <td>1184</td>\n",
       "      <td>616082</td>\n",
       "      <td>0.0</td>\n",
       "      <td>7.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17670</th>\n",
       "      <td>DMBX1</td>\n",
       "      <td>Atx|MBX|OTX3|PAXB</td>\n",
       "      <td>NM_147192.3</td>\n",
       "      <td>UNK</td>\n",
       "      <td>19026</td>\n",
       "      <td>607410</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17671</th>\n",
       "      <td>WDR93</td>\n",
       "      <td>C1d-87|CFAP297|FAP297</td>\n",
       "      <td>NM_020212.2</td>\n",
       "      <td>UNK</td>\n",
       "      <td>26924</td>\n",
       "      <td>None</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17672</th>\n",
       "      <td>ETV6</td>\n",
       "      <td>TEL|TEL/ABL|THC5</td>\n",
       "      <td>NM_001987.5</td>\n",
       "      <td>AD</td>\n",
       "      <td>3495</td>\n",
       "      <td>600618</td>\n",
       "      <td>2.0</td>\n",
       "      <td>39.0</td>\n",
       "      <td>38.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>17673 rows × 12 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           gene                                          altsymbol  \\\n",
       "0        RBFOX1                       2BP1|A2BP1|FOX-1|FOX1|HRNBP1   \n",
       "1         ABCA3                  ABC-C|ABC3|EST111653|LBM180|SMDP3   \n",
       "2        AKAP13  AKAP-13|AKAP-Lbc|ARHGEF13|BRX|c-lbc|HA-3|Ht31|...   \n",
       "3           GSS                           GSHS|HEL-S-64p|HEL-S-88n   \n",
       "4          BRF1  BRF|BRF-1|CFDS|GTF3B|hBRF|HEL-S-76p|TAF3B2|TAF...   \n",
       "...         ...                                                ...   \n",
       "17668  TNFRSF21                                   BM-018|CD358|DR6   \n",
       "17669   C12orf4                                              MRT66   \n",
       "17670     DMBX1                                  Atx|MBX|OTX3|PAXB   \n",
       "17671     WDR93                              C1d-87|CFAP297|FAP297   \n",
       "17672      ETV6                                   TEL|TEL/ABL|THC5   \n",
       "\n",
       "            refseq expected_inheritance hgncID  omimid  DFP     DM    DM?  \\\n",
       "0      NM_145891.3                   AD  18222  605104  0.0   15.0   47.0   \n",
       "1      NM_001089.3                   AR     33  601615  0.0  230.0  152.0   \n",
       "2      NM_007200.5                  UNK    371  604686  0.0    0.0   11.0   \n",
       "3      NM_000178.4                   AR   4624  601002  0.0   40.0    5.0   \n",
       "4      NM_001519.4                 ADAR  11551  604902  0.0   19.0   15.0   \n",
       "...            ...                  ...    ...     ...  ...    ...    ...   \n",
       "17668  NM_014452.5                  UNK  13469  605732  0.0    4.0    7.0   \n",
       "17669  NM_020374.4                   AR   1184  616082  0.0    7.0    4.0   \n",
       "17670  NM_147192.3                  UNK  19026  607410  0.0    1.0    5.0   \n",
       "17671  NM_020212.2                  UNK  26924    None  0.0    1.0    1.0   \n",
       "17672  NM_001987.5                   AD   3495  600618  2.0   39.0   38.0   \n",
       "\n",
       "        DP   FP    R  \n",
       "0      4.0  0.0  0.0  \n",
       "1      1.0  2.0  0.0  \n",
       "2      1.0  5.0  0.0  \n",
       "3      0.0  0.0  0.0  \n",
       "4      0.0  0.0  0.0  \n",
       "...    ...  ...  ...  \n",
       "17668  0.0  0.0  0.0  \n",
       "17669  0.0  0.0  0.0  \n",
       "17670  0.0  0.0  0.0  \n",
       "17671  0.0  0.0  0.0  \n",
       "17672  1.0  0.0  0.0  \n",
       "\n",
       "[17673 rows x 12 columns]"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "genocode_file_hg19 = '/resources/GENCODE_Basic_Annotations/gencode.v44lift37.basic.annotation.gtf.gz'\n",
    "genocode_db_hg19 = '/resources/GENCODE_Basic_Annotations/gencode.v44lift37.basic.annotation.db.bak'\n",
    "\n",
    "db = gffutils.FeatureDB(genocode_db_hg19)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import re\n",
    "\n",
    "@dataclass\n",
    "class GencodeInfo:\n",
    "    gene_name: str\n",
    "    hgnc: str\n",
    "    ensg: str\n",
    "    enst: str\n",
    "    ensg_full: str\n",
    "    enst_full: str\n",
    "    strand: str\n",
    "\n",
    "def __anno_gencode_info(row) -> str:\n",
    "    query_region: str = f\"chr{row['CHROM']}:{row['POS']}-{row['POS']}\"\n",
    "    # query_region: str = f\"chr17:1132706-1132706\"\n",
    "    fetched_data = db.region(region=query_region, \n",
    "                            featuretype='gene',\n",
    "                            completely_within=False)\n",
    "    \n",
    "    result = []\n",
    "    while 1:\n",
    "        try:\n",
    "            data = next(fetched_data)\n",
    "        except StopIteration:\n",
    "            break\n",
    "        else:\n",
    "            strand = data.strand\n",
    "            gene_name = data.attributes['gene_name'][0]\n",
    "            \n",
    "            try:\n",
    "                ensg_full = data.attributes['gene_id'][0]\n",
    "            except KeyError:\n",
    "                ensg = '.'\n",
    "            else:\n",
    "                ensg = re.match(r'ENSG\\d+', ensg_full).group()\n",
    "    \n",
    "    # print(result)\n",
    "    return result"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import re\n",
    "\n",
    "@dataclass\n",
    "class GencodeInfo:\n",
    "    gene_name: str\n",
    "    hgnc: str\n",
    "    ensg: str\n",
    "    enst: str\n",
    "    ensg_full: str\n",
    "    enst_full: str\n",
    "    strand: str\n",
    "\n",
    "def __anno_gencode_info(row) -> str:\n",
    "    query_region: str = f\"chr{row['CHROM']}:{row['POS']}-{row['POS']}\"\n",
    "    # query_region: str = f\"chr17:1132706-1132706\"\n",
    "    fetched_data = db.region(region=query_region, \n",
    "                            featuretype='gene',\n",
    "                            completely_within=False)\n",
    "    \n",
    "    result = []\n",
    "    while 1:\n",
    "        try:\n",
    "            data = next(fetched_data)\n",
    "        except StopIteration:\n",
    "            break\n",
    "        else:\n",
    "            strand = data.strand\n",
    "            gene_name = data.attributes['gene_name'][0]\n",
    "\n",
    "            try:\n",
    "                hgnc_info = data.attributes['hgnc_id'][0]\n",
    "            except KeyError:\n",
    "                hgnc = '.'\n",
    "            else:\n",
    "                hgnc = re.search(r'\\d+', hgnc_info).group()\n",
    "            \n",
    "            try:\n",
    "                ensg_full = data.attributes['gene_id'][0]\n",
    "            except KeyError:\n",
    "                ensg = '.'\n",
    "            else:\n",
    "                ensg = re.match(r'ENSG\\d+', ensg_full).group()\n",
    "            \n",
    "            try:\n",
    "                enst_full = data.attributes['transcript_id'][0]\n",
    "            except KeyError:\n",
    "                enst = '.'\n",
    "            else:\n",
    "                enst = re.match(r'ENST\\d+', enst_full).group()\n",
    "\n",
    "            genecode_info = GencodeInfo(\n",
    "                gene_name, hgnc, ensg, enst, ensg_full, enst_full, strand)\n",
    "            result.append(genecode_info)\n",
    "    \n",
    "    # print(result)\n",
    "    return result\n",
    "        \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "query_region: str = f\"chr17:1132706-1132706\"\n",
    "fetched_data = db.region(region=query_region, \n",
    "                            featuretype='transcript',\n",
    "                            completely_within=False)\n",
    "    \n",
    "result = []\n",
    "while 1:\n",
    "    try:\n",
    "        data = next(fetched_data)\n",
    "    except StopIteration:\n",
    "        break\n",
    "    else:\n",
    "        strand = data.strand\n",
    "        gene_name = data.attributes['gene_name'][0]\n",
    "\n",
    "        try:\n",
    "            hgnc_info = data.attributes['hgnc_id'][0]\n",
    "        except KeyError:\n",
    "            hgnc = '.'\n",
    "        else:\n",
    "            hgnc = re.search(r'\\d+', hgnc_info).group()\n",
    "        \n",
    "        try:\n",
    "            ensg_full = data.attributes['gene_id'][0]\n",
    "        except KeyError:\n",
    "            ensg = '.'\n",
    "        else:\n",
    "            ensg = re.match(r'ENSG\\d+', ensg_full).group()\n",
    "        \n",
    "        try:\n",
    "            enst_full = data.attributes['transcript_id'][0]\n",
    "        except KeyError:\n",
    "            enst = '.'\n",
    "        else:\n",
    "            enst = re.match(r'ENST\\d+', enst_full).group()\n",
    "\n",
    "        genecode_info = GencodeInfo(\n",
    "            gene_name, hgnc, ensg, enst, ensg_full, enst_fuœll, strand)\n",
    "        result.append(genecode_info)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "query_region: str = f\"chr17:1132706-1132706\"\n",
    "fetched_gene = db.region(region=query_region, \n",
    "                         featuretype=['gene', 'transcript'],\n",
    "                         completely_within=False)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "gtf = db.children('', featuretype='transcript', order_by='start')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for g in gtf:\n",
    "    print(g)\n",
    "    print(g.attributes['tag'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 6443/6443 [01:25<00:00, 75.63it/s] \n"
     ]
    }
   ],
   "source": [
    "\n",
    "df['gencode'] = df.progress_apply(__anno_gencode_info, axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for row in df.iterrows():\n",
    "    print(row[0], row[1]['Gene.refGene'], row[1]['gencode'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "chr1\tHAVANA\ttranscript\t65419\t71585\t.\t+\t.\tgene_id \"ENSG00000186092.7_9\"; transcript_id \"ENST00000641515.2_5\"; gene_type \"protein_coding\"; gene_name \"OR4F5\"; transcript_type \"protein_coding\"; transcript_name \"OR4F5-201\"; level \"2\"; protein_id \"ENSP00000493376.2\"; hgnc_id \"HGNC:14825\"; tag \"RNA_Seq_supported_partial\"; tag \"basic\"; tag \"Ensembl_canonical\"; tag \"MANE_Select\"; tag \"appris_principal_1\"; tag \"CCDS\"; ccdsid \"CCDS30547.2\"; havana_gene \"OTTHUMG00000001094.4_9\"; havana_transcript \"OTTHUMT00000003223.4_5\"; remap_num_mappings \"1\"; remap_status \"full_contig\"; remap_target_status \"new\";\n",
      "Strand: +\n",
      "ENSG_Full: ENSG00000186092.7_9\n",
      "ENSG     : ENSG00000186092\n",
      "ENST_Full: ENST00000641515.2_5\n",
      "ENST     : ENST00000641515\n",
      "GeneName : OR4F5\n",
      "HGNC_ID  : 14825\n"
     ]
    }
   ],
   "source": [
    "import re\n",
    "\n",
    "query_region: str = f'chr1: 69610-69610'\n",
    "rows = db.region(region=query_region, featuretype='transcript')\n",
    "\n",
    "for row in rows:\n",
    "    print(row)\n",
    "\n",
    "    enst_full = row.attributes['transcript_id'][0]\n",
    "    ensg_full = row.attributes['gene_id'][0]\n",
    "    hgnc_info = row.attributes['hgnc_id'][0]\n",
    "\n",
    "    try:\n",
    "        enst = re.match(r'ENST\\d+', enst_full).group()\n",
    "    except AttributeError:\n",
    "        enst = '.'\n",
    "    try:\n",
    "        ensg = re.match(r'ENSG\\d+', ensg_full).group()\n",
    "    except AttributeError:\n",
    "        ensg = '.'\n",
    "    try:\n",
    "        hgnc = re.search(r'\\d+', hgnc_info).group()\n",
    "    except AttributeError:\n",
    "        hgnc = '.'\n",
    "\n",
    "    print(f\"Strand: {row.strand}\")\n",
    "    print(f\"GeneName : {row.attributes['gene_name'][0]}\")\n",
    "    print(f\"HGNC_ID  : {hgnc}\")\n",
    "    print(f\"ENSG_Full: {ensg_full}\")\n",
    "    print(f\"ENSG     : {ensg}\")\n",
    "    print(f\"ENST_Full: {enst_full}\")\n",
    "    print(f\"ENST     : {enst}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#-----   STEP 5. Pre-processing\n",
    "preprocessing = PreProcessExomeSummary(\n",
    "    df=df, mode_samples_info=mode_samples_info)\n",
    "df = preprocessing.all_pre_processing()\n",
    "\n",
    "#-----   STEP 6. Additional annotations\n",
    "anno = Anno(df=df, args=args)\n",
    "df = anno.anno_scores()\n",
    "\n",
    "#-----   STEP 7. Filtering\n",
    "maffilter = MafFilter(\n",
    "    df=df, mode_samples_info=mode_samples_info, config=config)\n",
    "df = maffilter.all_filtering()\n",
    "\n",
    "typefilter = TypeFilter(df=df)\n",
    "df = typefilter.exclude_hlamuc_and_exonicsyno()\n",
    "\n",
    "gtfilter = GtFilter(\n",
    "    df=df, mode_samples_info=mode_samples_info)\n",
    "dfs = gtfilter.genotypeing_filter()\n",
    "\n",
    "#-----   STEP 9. Output as an Excel \n",
    "def df_to_excel(dfs: dataclass, output_xlsx) -> None:\n",
    "    sheet_names = ['AD', 'Homo', 'CH', 'XL']\n",
    "    with pd.ExcelWriter(output_xlsx) as writer:\n",
    "        for df, sheet_name in zip([dfs.AD, dfs.Hm, dfs.CH, dfs.XL], sheet_names):\n",
    "            df.to_excel(writer, sheet_name=sheet_name, index=False)\n",
    "\n",
    "output_xlsx = f\"{output_file_path}.xlsx\"\n",
    "output_pickle = f\"{output_file_path}.pkl\"\n",
    "df_to_excel(dfs, output_xlsx)\n",
    "dfs.to_pickle(output_pickle)\n",
    "\n",
    "#-----   STEP 8. Count variants of filtering process\n",
    "countsummery_file = str(Path(output_file_path).parent) + '/CountSummary.xlsx'\n",
    "counter_result = counter(dfs=dfs, output_excel=countsummery_file)\n",
    "print(counter_result)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['InHouse_absent_FILTER', 'InHouse_1%_FILTER', 'MAF_0.1%_FILTER',\n",
       "       'MAF_1%_FILTER', 'HLAMUC_FILTER', 'ExonicSyno_FILTER', 'GT_FILTER'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dfs.AD.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/work/Github/TestData/trio/29881/Sample_29881-trio_results/Sample_29881-trio.tsv\n"
     ]
    }
   ],
   "source": [
    "print(output_file_path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#-----   STEP 9. Output as an Excel \n",
    "def df_to_excel(dfs: dataclass, output_xlsx) -> None:\n",
    "    sheet_names = ['AD', 'Homo', 'CH', 'XL']\n",
    "    with pd.ExcelWriter(output_xlsx) as writer:\n",
    "        for df, sheet_name in zip([dfs.AD, dfs.Hm, dfs.CH, dfs.XL], sheet_names):\n",
    "            df.to_excel(writer, sheet_name=sheet_name, index=False)\n",
    "\n",
    "output_xlsx = f\"{output_file_path}.xlsx\"\n",
    "df_to_excel(dfs, output_xlsx)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'Sample_32741-proband_countsummary.txt'"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "hgmd_resource: str = [\n",
    "    str(x) for x in self.resources_dir.glob('HGMD/HGMD_gene_based.tsv.gz')][0]\n",
    "        return pd.read_csv(hgmd_resource, sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "couter_result = counter(dfs=dfs, output_excel=output_file_path)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df = counter(dfs=dfs, output_excel='./test.xlsx')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df.to_pickle('./post7.pkl')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df = pd.read_pickle('./post7.pkl')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df = df.head(100)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#-----   STEP 8. Output as an Excel \n",
    "output_xlsx = './head100.xlsx'\n",
    "def df_to_excel(dfs: dataclass, output_xlsx) -> None:\n",
    "    sheet_names = ['AD', 'Homo', 'CH', 'XL']\n",
    "    with pd.ExcelWriter(output_xlsx) as writer:\n",
    "        for df, sheet_name in zip([dfs.AD, dfs.Hm, dfs.CH, dfs.XL], sheet_names):\n",
    "            df.to_excel(writer, sheet_name=sheet_name, index=False)\n",
    "\n",
    "df_to_excel(dfs, 'output_xlsx')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 102,
   "metadata": {},
   "outputs": [],
   "source": [
    "from cyvcf2 import VCF\n",
    "from pysam import VariantFile\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 103,
   "metadata": {},
   "outputs": [],
   "source": [
    "tommo_bulk_file = '/Users/utsu/Downloads/tommo-54kjpn-20230626r3-GRCh37_lifted_from_GRCh38-af-chrX_PAR2.vcf.gz'\n",
    "tbx = pysam.TabixFile(tommo_bulk_file)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 126,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1.9999999494757503e-05\n"
     ]
    }
   ],
   "source": [
    "query_start, query_end = 60145, 60146\n",
    "query_chrom: str = 'X'\n",
    "query_ref, query_alt = 'G', 'A'\n",
    "query_variant = f\"{query_chrom}:{query_end}-{query_ref}-{query_alt}\"\n",
    "\n",
    "# for r in tbx.fetch(query_chrom, query_start, query_end, parser=pysam.asVCF()):\n",
    "#     print(r.info)\n",
    "\n",
    "vcfin = VariantFile(tommo_bulk_file)\n",
    "for rec in vcfin.fetch(query_chrom, query_start, query_end):\n",
    "    if len(rec.alts) > 1:\n",
    "        continue\n",
    "    if rec.filter.keys()[0] != 'PASS':\n",
    "        continue\n",
    "    \n",
    "    tommo_id = f\"{rec.contig}:{rec.pos}-{rec.ref}-{rec.alts[0]}\"\n",
    "    if tommo_id == query_variant:\n",
    "        print(rec.info['AF'][0])\n",
    "        break\n",
    "    else:\n",
    "        pass\n",
    "        \n",
    "\n",
    "    \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [],
   "source": [
    "df = pd.read_table('/Users/utsu/work/Github/TestData/proband/32487_IRUD_TTR/annovar/exome_summary.20230413_004616.txt', header=0, dtype=str)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [],
   "source": [
    "from pysam import VariantFile\n",
    "tommo_bulk_file = '/Users/utsu/Downloads/tommo-54kjpn-20230626r3-GRCh37_lifted_from_GRCh38-af-chrX_PAR2.vcf.gz'\n",
    "tommovcf = VariantFile(tommo_bulk_file)\n",
    "\n",
    "def anno_tommo(row, tommovcf):\n",
    "    # Set queries\n",
    "    query_chrom: str = row['CHROM']\n",
    "    query_start: int = int(row['POS']) - 1\n",
    "    query_end: int = int(row['POS'])\n",
    "    query_variant_id = f\"{query_chrom}:{query_end}-{row['REF']}-{row['ALT']}\"\n",
    "\n",
    "    for rec in tommovcf.fetch(query_chrom, query_start, query_end):\n",
    "        # Filter out non-PASS variants\n",
    "        if rec.filter.keys()[0] != 'PASS':\n",
    "            return '.'\n",
    "        \n",
    "        for i, alt in enumerate(rec.alts):\n",
    "            tommo_variant_id = f\"{rec.contig}:{rec.pos}-{rec.ref}-{alt}\"\n",
    "            if tommo_variant_id == query_variant_id:\n",
    "                return rec.info['AF'][i]\n",
    "            else:\n",
    "                pass\n",
    "\n",
    "        return '.'       "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "  0%|          | 0/7940 [00:00<?, ?it/s]"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 7940/7940 [00:01<00:00, 6623.29it/s] \n"
     ]
    }
   ],
   "source": [
    "\n",
    "df['ToMMo_AF'] = df.progress_apply(anno_tommo, tommovcf=tommovcf, axis=1)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Replace '.' to np.nan in 'ToMMo_AF' column\n",
    "df['ToMMo_AF'] = df['ToMMo_AF'].replace('.', np.nan)\n",
    "df = df.astype({'ToMMo_AF': float})\n",
    "df.loc[((df['CHROM'] == 'X') & (df['ToMMo_AF'] < 0.01)), 'ToMMo_AF']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "CHROM:  X\n",
      "POS:    60170\n",
      "ID:     rs2068139808\n",
      "REF:    C\n",
      "ALT:    ['G']\n",
      "QUAL:   47.290000915527344\n",
      "INFO:   <cyvcf2.cyvcf2.INFO object at 0x108081c60>\n",
      "starat: 60169\n",
      "end:    60170\n",
      "          \n",
      "('AC', 2)\n",
      "('AF', 1.9999999494757503e-05)\n",
      "('AN', 102418)\n",
      "('DP', 807194)\n",
      "('ExcessHet', 0.008299999870359898)\n",
      "('FS', 0.0)\n",
      "('InbreedingCoeff', 0.0013000000035390258)\n",
      "('MLEAC', 1)\n",
      "('MLEAF', 7.679999725951348e-06)\n",
      "('QD', 15.760000228881836)\n",
      "('SOR', 2.8329999446868896)\n",
      "('NEGATIVE_TRAIN_SITE', True)\n",
      "('VQSLOD', -3.8850998878479004)\n",
      "('culprit', 'DP')\n",
      "('AC_XX', 2)\n",
      "('AN_XX', 67988)\n",
      "('AF_XX', 2.9000000722589903e-05)\n",
      "('AC_XY', 0)\n",
      "('AN_XY', 34430)\n",
      "('AF_XY', 0.0)\n",
      "('AC_Illumina_HiSeq_2500_162PE', 0)\n",
      "('AN_Illumina_HiSeq_2500_162PE', 5876)\n",
      "('AF_Illumina_HiSeq_2500_162PE', 0.0)\n",
      "('AC_Illumina_NovaSeq_6000_150PE', 0)\n",
      "('AN_Illumina_NovaSeq_6000_150PE', 45688)\n",
      "('AF_Illumina_NovaSeq_6000_150PE', 0.0)\n",
      "('AC_Illumina_NovaSeq_6000_161PE', 2)\n",
      "('AN_Illumina_NovaSeq_6000_161PE', 33674)\n",
      "('AF_Illumina_NovaSeq_6000_161PE', 5.900000178371556e-05)\n",
      "('AC_MGI_DNBSeq_T7_150PE', 0)\n",
      "('AN_MGI_DNBSeq_T7_150PE', 13994)\n",
      "('AF_MGI_DNBSeq_T7_150PE', 0.0)\n",
      "('AC_Illumina_HiSeq_2500_162PE_XX', 0)\n",
      "('AN_Illumina_HiSeq_2500_162PE_XX', 3146)\n",
      "('AF_Illumina_HiSeq_2500_162PE_XX', 0.0)\n",
      "('AC_Illumina_HiSeq_2500_162PE_XY', 0)\n",
      "('AN_Illumina_HiSeq_2500_162PE_XY', 2730)\n",
      "('AF_Illumina_HiSeq_2500_162PE_XY', 0.0)\n",
      "('AC_Illumina_NovaSeq_6000_150PE_XX', 0)\n",
      "('AN_Illumina_NovaSeq_6000_150PE_XX', 33442)\n",
      "('AF_Illumina_NovaSeq_6000_150PE_XX', 0.0)\n",
      "('AC_Illumina_NovaSeq_6000_150PE_XY', 0)\n",
      "('AN_Illumina_NovaSeq_6000_150PE_XY', 12246)\n",
      "('AF_Illumina_NovaSeq_6000_150PE_XY', 0.0)\n",
      "('AC_Illumina_NovaSeq_6000_161PE_XX', 2)\n",
      "('AN_Illumina_NovaSeq_6000_161PE_XX', 21844)\n",
      "('AF_Illumina_NovaSeq_6000_161PE_XX', 9.200000204145908e-05)\n",
      "('AC_Illumina_NovaSeq_6000_161PE_XY', 0)\n",
      "('AN_Illumina_NovaSeq_6000_161PE_XY', 11830)\n",
      "('AF_Illumina_NovaSeq_6000_161PE_XY', 0.0)\n",
      "('AC_MGI_DNBSeq_T7_150PE_XX', 0)\n",
      "('AN_MGI_DNBSeq_T7_150PE_XX', 7882)\n",
      "('AF_MGI_DNBSeq_T7_150PE_XX', 0.0)\n",
      "('AC_MGI_DNBSeq_T7_150PE_XY', 0)\n",
      "('AN_MGI_DNBSeq_T7_150PE_XY', 6112)\n",
      "('AF_MGI_DNBSeq_T7_150PE_XY', 0.0)\n",
      "('TOMMO_PLATFORM_BIAS_TEST_ALLELE', 'G')\n",
      "('TOMMO_PLATFORM_BIAS_TEST_PVALUE', 0.502249002456665)\n",
      "('HWE', 9.764000424183905e-06)\n",
      "('ANN', 'G|intergenic_region|MODIFIER|CHR_START-PLCXD1|CHR_START-PLCXD1|intergenic_region|CHR_START-PLCXD1|||n.10170C>G||||||')\n",
      "('ORIGINAL_CHROM', 'chrX')\n",
      "('ORIGINAL_POS', 10170)\n",
      "('ORIGINAL_STRAND', '+')\n",
      "\n",
      "CHROM:  X\n",
      "POS:    60181\n",
      "ID:     rs189980076\n",
      "REF:    G\n",
      "ALT:    ['C']\n",
      "QUAL:   108188.0\n",
      "INFO:   <cyvcf2.cyvcf2.INFO object at 0x137f99480>\n",
      "starat: 60180\n",
      "end:    60181\n",
      "          \n",
      "('AC', 540)\n",
      "('AF', 0.0052709998562932014)\n",
      "('AN', 102438)\n",
      "('BaseQRankSum', 0.0)\n",
      "('DP', 854128)\n",
      "('ExcessHet', -0.0)\n",
      "('FS', 1.3450000286102295)\n",
      "('InbreedingCoeff', 0.20280000567436218)\n",
      "('MLEAC', 660)\n",
      "('MLEAF', 0.00506999995559454)\n",
      "('MQRankSum', 0.0)\n",
      "('QD', 13.760000228881836)\n",
      "('ReadPosRankSum', 0.12800000607967377)\n",
      "('SOR', 0.8180000185966492)\n",
      "('NEGATIVE_TRAIN_SITE', True)\n",
      "('VQSLOD', -1.7177000045776367)\n",
      "('culprit', 'MQRankSum')\n",
      "('AC_XX', 360)\n",
      "('AN_XX', 67994)\n",
      "('AF_XX', 0.005295000039041042)\n",
      "('AC_XY', 180)\n",
      "('AN_XY', 34444)\n",
      "('AF_XY', 0.005226000212132931)\n",
      "('AC_Illumina_HiSeq_2500_162PE', 28)\n",
      "('AN_Illumina_HiSeq_2500_162PE', 5876)\n",
      "('AF_Illumina_HiSeq_2500_162PE', 0.004765000194311142)\n",
      "('AC_Illumina_NovaSeq_6000_150PE', 231)\n",
      "('AN_Illumina_NovaSeq_6000_150PE', 45696)\n",
      "('AF_Illumina_NovaSeq_6000_150PE', 0.005055000074207783)\n",
      "('AC_Illumina_NovaSeq_6000_161PE', 194)\n",
      "('AN_Illumina_NovaSeq_6000_161PE', 33682)\n",
      "('AF_Illumina_NovaSeq_6000_161PE', 0.005760000087320805)\n",
      "('AC_MGI_DNBSeq_T7_150PE', 73)\n",
      "('AN_MGI_DNBSeq_T7_150PE', 14000)\n",
      "('AF_MGI_DNBSeq_T7_150PE', 0.00521400012075901)\n",
      "('AC_Illumina_HiSeq_2500_162PE_XX', 17)\n",
      "('AN_Illumina_HiSeq_2500_162PE_XX', 3148)\n",
      "('AF_Illumina_HiSeq_2500_162PE_XX', 0.005400000140070915)\n",
      "('AC_Illumina_HiSeq_2500_162PE_XY', 11)\n",
      "('AN_Illumina_HiSeq_2500_162PE_XY', 2728)\n",
      "('AF_Illumina_HiSeq_2500_162PE_XY', 0.004031999967992306)\n",
      "('AC_Illumina_NovaSeq_6000_150PE_XX', 170)\n",
      "('AN_Illumina_NovaSeq_6000_150PE_XX', 33444)\n",
      "('AF_Illumina_NovaSeq_6000_150PE_XX', 0.00508299982175231)\n",
      "('AC_Illumina_NovaSeq_6000_150PE_XY', 61)\n",
      "('AN_Illumina_NovaSeq_6000_150PE_XY', 12252)\n",
      "('AF_Illumina_NovaSeq_6000_150PE_XY', 0.004978999961167574)\n",
      "('AC_Illumina_NovaSeq_6000_161PE_XX', 125)\n",
      "('AN_Illumina_NovaSeq_6000_161PE_XX', 21846)\n",
      "('AF_Illumina_NovaSeq_6000_161PE_XX', 0.0057219997979700565)\n",
      "('AC_Illumina_NovaSeq_6000_161PE_XY', 69)\n",
      "('AN_Illumina_NovaSeq_6000_161PE_XY', 11836)\n",
      "('AF_Illumina_NovaSeq_6000_161PE_XY', 0.005830000154674053)\n",
      "('AC_MGI_DNBSeq_T7_150PE_XX', 45)\n",
      "('AN_MGI_DNBSeq_T7_150PE_XX', 7884)\n",
      "('AF_MGI_DNBSeq_T7_150PE_XX', 0.005708000157028437)\n",
      "('AC_MGI_DNBSeq_T7_150PE_XY', 28)\n",
      "('AN_MGI_DNBSeq_T7_150PE_XY', 6116)\n",
      "('AF_MGI_DNBSeq_T7_150PE_XY', 0.0045779999345541)\n",
      "('TOMMO_PLATFORM_BIAS_TEST_ALLELE', 'C')\n",
      "('TOMMO_PLATFORM_BIAS_TEST_PVALUE', 0.6921539902687073)\n",
      "('HWE', 0.0)\n",
      "('ANN', 'C|intergenic_region|MODIFIER|CHR_START-PLCXD1|CHR_START-PLCXD1|intergenic_region|CHR_START-PLCXD1|||n.10181G>C||||||')\n",
      "('ORIGINAL_CHROM', 'chrX')\n",
      "('ORIGINAL_POS', 10181)\n",
      "('ORIGINAL_STRAND', '+')\n"
     ]
    }
   ],
   "source": [
    "vcf = VCF(tommo_bulk_file)\n",
    "i = 0\n",
    "while 1:\n",
    "    i += 1\n",
    "    if i > 2:\n",
    "        break\n",
    "\n",
    "    v = next(vcf)\n",
    "    \n",
    "    print(f\"\"\"\n",
    "CHROM:  {v.CHROM}\n",
    "POS:    {v.POS}\n",
    "ID:     {v.ID}\n",
    "REF:    {v.REF}\n",
    "ALT:    {v.ALT}\n",
    "QUAL:   {v.QUAL}\n",
    "INFO:   {v.INFO}\n",
    "starat: {v.start}\n",
    "end:    {v.end}\n",
    "          \"\"\")\n",
    "    \n",
    "    for inf in v.INFO:\n",
    "        print(inf)\n",
    "    \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pysam"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#----   STEP 9. Insert hyperlinks\n",
    "# 0. 最初にエクセルを読み込む\n",
    "# 1. メモ用の列を作る (1-2列目)\n",
    "# 2. リンク挿入用の列を作る (3-7列目)\n",
    "# 3. リンクの挿入"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Formatng ......\n"
     ]
    }
   ],
   "source": [
    "from libs.excelibs.excel_format import ExcelFormat\n",
    "from libs.excelibs.hyperlinks import HyperLinks\n",
    "\n",
    "excelformat = ExcelFormat('./test1.xlsx')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "excelformat.insert_comment_cols()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "excelformat.insert_hyperlink_cols()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "excelformat.workbook.save('./test2.xlsx')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### CREATE dbs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Create db for gffutils\n",
    "import gffutils\n",
    "import gffutils.pybedtools_integration\n",
    "\n",
    "\n",
    "genocode_file_hg19 = '/resources/GENCODE_Basic_Annotations/gencode.v44lift37.basic.annotation.gtf.gz'\n",
    "genocode_db_hg19 = '/resources/GENCODE_Basic_Annotations/gencode.v44lift37.basic.annotation.db'\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "db = gffutils.create_db(data=genocode_file_hg19, dbfn=genocode_db_hg19, \n",
    "                        disable_infer_genes=True,\n",
    "                        disable_infer_transcripts=True,\n",
    "                        keep_order=True, \n",
    "                        force=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<BedTool(/resources/GENCODE_Basic_Annotations/gencode.v44lift37.basic.annotation.intron.db)>"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genocode_db_intron_hg19 = '/resources/GENCODE_Basic_Annotations/gencode.v44lift37.basic.annotation.intron.db'\n",
    "\n",
    "db = gffutils.FeatureDB(genocode_db_hg19)\n",
    "introns = db.create_introns(exon_featuretype='exon', \n",
    "                            new_featuretype='intron', \n",
    "                            merge_attributes=True, \n",
    "                            numeric_sort=True)\n",
    "pybed = gffutils.pybedtools_integration.to_bedtool(introns)\n",
    "pybed.saveas(genocode_db_intron_hg19)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def generate_intoron_gtf(db: gffutils.FeatureDB, output: str) -> None:\n",
    "    introns = db.create_introns(exon_featuretype='exon', \n",
    "                                new_featuretype='intron', \n",
    "                                merge_attributes=True, \n",
    "                                numeric_sort=True)\n",
    "    pybed = gffutils.pybedtools_integration.to_bedtool(introns)\n",
    "    pybed.saveas(output)\n",
    "    \n",
    "    return"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "wesanno",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.18"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
