{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "INFO: Pandarallel will run on 4 workers.\n",
      "INFO: Pandarallel will use Memory file system to transfer data between the main process and workers.\n"
     ]
    }
   ],
   "source": [
    "# Standard modules\n",
    "from dataclasses import dataclass\n",
    "from logging import getLogger, StreamHandler, DEBUG, Formatter\n",
    "\n",
    "# Third party modules\n",
    "import gffutils\n",
    "import pybedtools\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "from pathlib2 import Path\n",
    "from tqdm import tqdm\n",
    "\n",
    "# Local modules\n",
    "from libs.args import parser_setting\n",
    "from libs.utils import load_config, OutputSettings\n",
    "from libs.preprocess import PreProcessExomeSummary\n",
    "from libs.modesamples import ModeSamples\n",
    "from libs.annolibs.anno import Anno\n",
    "from libs.filter.maffilter import MafFilter\n",
    "from libs.filter.typefilter import TypeFilter\n",
    "from libs.filter.gtfilter import GtFilter\n",
    "from libs.filter.counter import counter\n",
    "\n",
    "# Settings\n",
    "tqdm.pandas()\n",
    "logger = getLogger(__name__)\n",
    "handler = StreamHandler()\n",
    "handler.setLevel(DEBUG)\n",
    "logger.setLevel(DEBUG)\n",
    "logger.addHandler(handler)\n",
    "\n",
    "# http://www.openbioinformatics.org/annovar/download/0wgxR2rIVP/annovar.latest.tar.gz"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pysam"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "     input: /work/Github/TestData/trio/29881/exome_summary.20230420_014141.txt\n",
      "     output: None\n",
      "     xhmm: /work/Github/TestData/proband/xhmm/data.segdup.strvar.haplo.deciph.omim.xcnv.gene.uniq\n",
      "     vcf: None\n",
      "     phenotype: None\n",
      "     mode: auto\n",
      "     samples: auto\n",
      "     config: /work/Github/playground/wesanno/config/config.toml\n",
      "     resources: /resources\n",
      "     no_gnomad: False\n",
      "     no_hgmd: False\n",
      "     no_decipher: False\n",
      "     no_ddg2p: False\n",
      "     no_jarvis: False\n",
      "     no_spliceai: False\n",
      "     no_syno: None\n",
      "     no_alphamissense: False\n",
      "     no_revel: False\n",
      "     no_trap: False\n",
      "     excel_formating: True\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "#-----   STEP 1. Argument settings\n",
    "args = parser_setting()\n",
    "\n",
    "#-----   STEP 2. Input file and Output settings\n",
    "df = pd.read_table(args['input'], header=0, dtype=str)\n",
    "config: dict = load_config(args['config'])\n",
    "\n",
    "#-----   STEP 3. Get Mode and Samples information\n",
    "modesamples = ModeSamples(df=df, args=args)\n",
    "mode_samples_info = modesamples.get_mode_samples_info()\n",
    "\n",
    "#-----   STEP 4. Output settings\n",
    "output_settings = OutputSettings(\n",
    "    args=args, mode_samples_info=mode_samples_info)\n",
    "output_file_path = output_settings.get_saving_file_path()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "metadata": {},
   "outputs": [],
   "source": [
    "def process_dataframe_optimized(df):\n",
    "    # 処理結果を格納するためのリストを初期化\n",
    "    rows_list = []\n",
    "\n",
    "    # データフレームの各行に対して処理を行う\n",
    "    for row in df.itertuples(index=False):\n",
    "        # REF列の値をカンマで分割\n",
    "        # refs = row['ALT'].split(',')\n",
    "        alts = row.ALT.split(',')\n",
    "        # 分割した値ごとに新しい行を作成\n",
    "        for alt in alts:\n",
    "            new_row = row._asdict()\n",
    "            new_row['SplitALT'] = alt\n",
    "            rows_list.append(new_row)\n",
    "\n",
    "    # 新しいデータフレームを作成\n",
    "    result_df = pd.DataFrame(rows_list)\n",
    "\n",
    "    return result_df\n",
    "\n",
    "# 関数を実行し、結果を表示\n",
    "# processed_df = process_dataframe_optimized(df)\n",
    "# print(processed_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "metadata": {},
   "outputs": [],
   "source": [
    "processed_df = process_dataframe_optimized(df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "metadata": {},
   "outputs": [],
   "source": [
    "def split_qc_col(df: pd.DataFrame) -> pd.DataFrame:\n",
    "    for i, new_col in enumerate(['GT', 'AD', 'DP', 'GQ', 'PL']):\n",
    "        df[new_col] = df[mode_samples_info.proband_id].str.split(':').str[i]\n",
    "    df.fillna({'GQ': 0}, inplace=True)\n",
    "    df = df.astype({'GQ': 'int32'})\n",
    "   \n",
    "    return df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "metadata": {},
   "outputs": [],
   "source": [
    "df = split_qc_col(df)\n",
    "processed_df = split_qc_col(processed_df)\n",
    "df2 = processed_df.drop_duplicates(subset=['CHROM', 'POS', 'ALT', 'SplitALT'])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 82,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['0/1', '0/0', '1/1', './.', '0/2', '1/2', '1/3', '2/2', '2/3',\n",
       "       '0/3', '3/3'], dtype=object)"
      ]
     },
     "execution_count": 82,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df['GT'].unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 88,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>GT</th>\n",
       "      <th>PL</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2827</th>\n",
       "      <td>ATGTG</td>\n",
       "      <td>A,ATG,ATGTGTGTGTG</td>\n",
       "      <td>2/3</td>\n",
       "      <td>978,1028,1563,771,1053,990,270,291,0,625</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2828</th>\n",
       "      <td>ATGTG</td>\n",
       "      <td>A,ATG,ATGTGTGTGTG</td>\n",
       "      <td>2/3</td>\n",
       "      <td>978,1028,1563,771,1053,990,270,291,0,625</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2829</th>\n",
       "      <td>ATGTG</td>\n",
       "      <td>A,ATG,ATGTGTGTGTG</td>\n",
       "      <td>2/3</td>\n",
       "      <td>978,1028,1563,771,1053,990,270,291,0,625</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3564</th>\n",
       "      <td>CTCTCTGTGTGTGTGTGTGTGTG</td>\n",
       "      <td>CTGTGTGTGTGTG,CTG,C</td>\n",
       "      <td>2/3</td>\n",
       "      <td>2783,1303,1255,243,214,195,221,167,0,173</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3565</th>\n",
       "      <td>CTCTCTGTGTGTGTGTGTGTGTG</td>\n",
       "      <td>CTGTGTGTGTGTG,CTG,C</td>\n",
       "      <td>2/3</td>\n",
       "      <td>2783,1303,1255,243,214,195,221,167,0,173</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3566</th>\n",
       "      <td>CTCTCTGTGTGTGTGTGTGTGTG</td>\n",
       "      <td>CTGTGTGTGTGTG,CTG,C</td>\n",
       "      <td>2/3</td>\n",
       "      <td>2783,1303,1255,243,214,195,221,167,0,173</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4296</th>\n",
       "      <td>TAGCAGCAGC</td>\n",
       "      <td>TAGC,T,TAGCAGC</td>\n",
       "      <td>2/3</td>\n",
       "      <td>4058,2218,2300,1039,1097,1811,2190,1230,0,2077</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4297</th>\n",
       "      <td>TAGCAGCAGC</td>\n",
       "      <td>TAGC,T,TAGCAGC</td>\n",
       "      <td>2/3</td>\n",
       "      <td>4058,2218,2300,1039,1097,1811,2190,1230,0,2077</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9610</th>\n",
       "      <td>G</td>\n",
       "      <td>GGGCTCCGTAGGTTCCATGGGCTCCGTAGGTTCCATGGGCTCCGTA...</td>\n",
       "      <td>2/3</td>\n",
       "      <td>32680,4972,4719,6973,3392,6720,306,138,0,53</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9611</th>\n",
       "      <td>G</td>\n",
       "      <td>GGGCTCCGTAGGTTCCATGGGCTCCGTAGGTTCCATGGGCTCCGTA...</td>\n",
       "      <td>2/3</td>\n",
       "      <td>32680,4972,4719,6973,3392,6720,306,138,0,53</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9612</th>\n",
       "      <td>G</td>\n",
       "      <td>GGGCTCCGTAGGTTCCATGGGCTCCGTAGGTTCCATGGGCTCCGTA...</td>\n",
       "      <td>2/3</td>\n",
       "      <td>32680,4972,4719,6973,3392,6720,306,138,0,53</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                          REF  \\\n",
       "2827                    ATGTG   \n",
       "2828                    ATGTG   \n",
       "2829                    ATGTG   \n",
       "3564  CTCTCTGTGTGTGTGTGTGTGTG   \n",
       "3565  CTCTCTGTGTGTGTGTGTGTGTG   \n",
       "3566  CTCTCTGTGTGTGTGTGTGTGTG   \n",
       "4296               TAGCAGCAGC   \n",
       "4297               TAGCAGCAGC   \n",
       "9610                        G   \n",
       "9611                        G   \n",
       "9612                        G   \n",
       "\n",
       "                                                    ALT   GT  \\\n",
       "2827                                  A,ATG,ATGTGTGTGTG  2/3   \n",
       "2828                                  A,ATG,ATGTGTGTGTG  2/3   \n",
       "2829                                  A,ATG,ATGTGTGTGTG  2/3   \n",
       "3564                                CTGTGTGTGTGTG,CTG,C  2/3   \n",
       "3565                                CTGTGTGTGTGTG,CTG,C  2/3   \n",
       "3566                                CTGTGTGTGTGTG,CTG,C  2/3   \n",
       "4296                                     TAGC,T,TAGCAGC  2/3   \n",
       "4297                                     TAGC,T,TAGCAGC  2/3   \n",
       "9610  GGGCTCCGTAGGTTCCATGGGCTCCGTAGGTTCCATGGGCTCCGTA...  2/3   \n",
       "9611  GGGCTCCGTAGGTTCCATGGGCTCCGTAGGTTCCATGGGCTCCGTA...  2/3   \n",
       "9612  GGGCTCCGTAGGTTCCATGGGCTCCGTAGGTTCCATGGGCTCCGTA...  2/3   \n",
       "\n",
       "                                                  PL  \n",
       "2827        978,1028,1563,771,1053,990,270,291,0,625  \n",
       "2828        978,1028,1563,771,1053,990,270,291,0,625  \n",
       "2829        978,1028,1563,771,1053,990,270,291,0,625  \n",
       "3564        2783,1303,1255,243,214,195,221,167,0,173  \n",
       "3565        2783,1303,1255,243,214,195,221,167,0,173  \n",
       "3566        2783,1303,1255,243,214,195,221,167,0,173  \n",
       "4296  4058,2218,2300,1039,1097,1811,2190,1230,0,2077  \n",
       "4297  4058,2218,2300,1039,1097,1811,2190,1230,0,2077  \n",
       "9610     32680,4972,4719,6973,3392,6720,306,138,0,53  \n",
       "9611     32680,4972,4719,6973,3392,6720,306,138,0,53  \n",
       "9612     32680,4972,4719,6973,3392,6720,306,138,0,53  "
      ]
     },
     "execution_count": 88,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.loc[((df['GT'] == '2/3')), ['REF', 'ALT', 'GT', 'PL']].head(50)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 89,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>SplitALT</th>\n",
       "      <th>GT</th>\n",
       "      <th>PL</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>3044</th>\n",
       "      <td>ATGTG</td>\n",
       "      <td>A,ATG,ATGTGTGTGTG</td>\n",
       "      <td>A</td>\n",
       "      <td>2/3</td>\n",
       "      <td>978,1028,1563,771,1053,990,270,291,0,625</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3045</th>\n",
       "      <td>ATGTG</td>\n",
       "      <td>A,ATG,ATGTGTGTGTG</td>\n",
       "      <td>ATG</td>\n",
       "      <td>2/3</td>\n",
       "      <td>978,1028,1563,771,1053,990,270,291,0,625</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3046</th>\n",
       "      <td>ATGTG</td>\n",
       "      <td>A,ATG,ATGTGTGTGTG</td>\n",
       "      <td>ATGTGTGTGTG</td>\n",
       "      <td>2/3</td>\n",
       "      <td>978,1028,1563,771,1053,990,270,291,0,625</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3047</th>\n",
       "      <td>ATGTG</td>\n",
       "      <td>A,ATG,ATGTGTGTGTG</td>\n",
       "      <td>A</td>\n",
       "      <td>2/3</td>\n",
       "      <td>978,1028,1563,771,1053,990,270,291,0,625</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3048</th>\n",
       "      <td>ATGTG</td>\n",
       "      <td>A,ATG,ATGTGTGTGTG</td>\n",
       "      <td>ATG</td>\n",
       "      <td>2/3</td>\n",
       "      <td>978,1028,1563,771,1053,990,270,291,0,625</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3049</th>\n",
       "      <td>ATGTG</td>\n",
       "      <td>A,ATG,ATGTGTGTGTG</td>\n",
       "      <td>ATGTGTGTGTG</td>\n",
       "      <td>2/3</td>\n",
       "      <td>978,1028,1563,771,1053,990,270,291,0,625</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3050</th>\n",
       "      <td>ATGTG</td>\n",
       "      <td>A,ATG,ATGTGTGTGTG</td>\n",
       "      <td>A</td>\n",
       "      <td>2/3</td>\n",
       "      <td>978,1028,1563,771,1053,990,270,291,0,625</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3051</th>\n",
       "      <td>ATGTG</td>\n",
       "      <td>A,ATG,ATGTGTGTGTG</td>\n",
       "      <td>ATG</td>\n",
       "      <td>2/3</td>\n",
       "      <td>978,1028,1563,771,1053,990,270,291,0,625</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3052</th>\n",
       "      <td>ATGTG</td>\n",
       "      <td>A,ATG,ATGTGTGTGTG</td>\n",
       "      <td>ATGTGTGTGTG</td>\n",
       "      <td>2/3</td>\n",
       "      <td>978,1028,1563,771,1053,990,270,291,0,625</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3845</th>\n",
       "      <td>CTCTCTGTGTGTGTGTGTGTGTG</td>\n",
       "      <td>CTGTGTGTGTGTG,CTG,C</td>\n",
       "      <td>CTGTGTGTGTGTG</td>\n",
       "      <td>2/3</td>\n",
       "      <td>2783,1303,1255,243,214,195,221,167,0,173</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3846</th>\n",
       "      <td>CTCTCTGTGTGTGTGTGTGTGTG</td>\n",
       "      <td>CTGTGTGTGTGTG,CTG,C</td>\n",
       "      <td>CTG</td>\n",
       "      <td>2/3</td>\n",
       "      <td>2783,1303,1255,243,214,195,221,167,0,173</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3847</th>\n",
       "      <td>CTCTCTGTGTGTGTGTGTGTGTG</td>\n",
       "      <td>CTGTGTGTGTGTG,CTG,C</td>\n",
       "      <td>C</td>\n",
       "      <td>2/3</td>\n",
       "      <td>2783,1303,1255,243,214,195,221,167,0,173</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3848</th>\n",
       "      <td>CTCTCTGTGTGTGTGTGTGTGTG</td>\n",
       "      <td>CTGTGTGTGTGTG,CTG,C</td>\n",
       "      <td>CTGTGTGTGTGTG</td>\n",
       "      <td>2/3</td>\n",
       "      <td>2783,1303,1255,243,214,195,221,167,0,173</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3849</th>\n",
       "      <td>CTCTCTGTGTGTGTGTGTGTGTG</td>\n",
       "      <td>CTGTGTGTGTGTG,CTG,C</td>\n",
       "      <td>CTG</td>\n",
       "      <td>2/3</td>\n",
       "      <td>2783,1303,1255,243,214,195,221,167,0,173</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3850</th>\n",
       "      <td>CTCTCTGTGTGTGTGTGTGTGTG</td>\n",
       "      <td>CTGTGTGTGTGTG,CTG,C</td>\n",
       "      <td>C</td>\n",
       "      <td>2/3</td>\n",
       "      <td>2783,1303,1255,243,214,195,221,167,0,173</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3851</th>\n",
       "      <td>CTCTCTGTGTGTGTGTGTGTGTG</td>\n",
       "      <td>CTGTGTGTGTGTG,CTG,C</td>\n",
       "      <td>CTGTGTGTGTGTG</td>\n",
       "      <td>2/3</td>\n",
       "      <td>2783,1303,1255,243,214,195,221,167,0,173</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3852</th>\n",
       "      <td>CTCTCTGTGTGTGTGTGTGTGTG</td>\n",
       "      <td>CTGTGTGTGTGTG,CTG,C</td>\n",
       "      <td>CTG</td>\n",
       "      <td>2/3</td>\n",
       "      <td>2783,1303,1255,243,214,195,221,167,0,173</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3853</th>\n",
       "      <td>CTCTCTGTGTGTGTGTGTGTGTG</td>\n",
       "      <td>CTGTGTGTGTGTG,CTG,C</td>\n",
       "      <td>C</td>\n",
       "      <td>2/3</td>\n",
       "      <td>2783,1303,1255,243,214,195,221,167,0,173</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4628</th>\n",
       "      <td>TAGCAGCAGC</td>\n",
       "      <td>TAGC,T,TAGCAGC</td>\n",
       "      <td>TAGC</td>\n",
       "      <td>2/3</td>\n",
       "      <td>4058,2218,2300,1039,1097,1811,2190,1230,0,2077</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4629</th>\n",
       "      <td>TAGCAGCAGC</td>\n",
       "      <td>TAGC,T,TAGCAGC</td>\n",
       "      <td>T</td>\n",
       "      <td>2/3</td>\n",
       "      <td>4058,2218,2300,1039,1097,1811,2190,1230,0,2077</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4630</th>\n",
       "      <td>TAGCAGCAGC</td>\n",
       "      <td>TAGC,T,TAGCAGC</td>\n",
       "      <td>TAGCAGC</td>\n",
       "      <td>2/3</td>\n",
       "      <td>4058,2218,2300,1039,1097,1811,2190,1230,0,2077</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4631</th>\n",
       "      <td>TAGCAGCAGC</td>\n",
       "      <td>TAGC,T,TAGCAGC</td>\n",
       "      <td>TAGC</td>\n",
       "      <td>2/3</td>\n",
       "      <td>4058,2218,2300,1039,1097,1811,2190,1230,0,2077</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4632</th>\n",
       "      <td>TAGCAGCAGC</td>\n",
       "      <td>TAGC,T,TAGCAGC</td>\n",
       "      <td>T</td>\n",
       "      <td>2/3</td>\n",
       "      <td>4058,2218,2300,1039,1097,1811,2190,1230,0,2077</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4633</th>\n",
       "      <td>TAGCAGCAGC</td>\n",
       "      <td>TAGC,T,TAGCAGC</td>\n",
       "      <td>TAGCAGC</td>\n",
       "      <td>2/3</td>\n",
       "      <td>4058,2218,2300,1039,1097,1811,2190,1230,0,2077</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10234</th>\n",
       "      <td>G</td>\n",
       "      <td>GGGCTCCGTAGGTTCCATGGGCTCCGTAGGTTCCATGGGCTCCGTA...</td>\n",
       "      <td>GGGCTCCGTAGGTTCCATGGGCTCCGTAGGTTCCATGGGCTCCGTA...</td>\n",
       "      <td>2/3</td>\n",
       "      <td>32680,4972,4719,6973,3392,6720,306,138,0,53</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10235</th>\n",
       "      <td>G</td>\n",
       "      <td>GGGCTCCGTAGGTTCCATGGGCTCCGTAGGTTCCATGGGCTCCGTA...</td>\n",
       "      <td>GGGCTCCGTAGGTTCCATGGGCTCCGTAGGTTCCATGGGCTCCGTA...</td>\n",
       "      <td>2/3</td>\n",
       "      <td>32680,4972,4719,6973,3392,6720,306,138,0,53</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10236</th>\n",
       "      <td>G</td>\n",
       "      <td>GGGCTCCGTAGGTTCCATGGGCTCCGTAGGTTCCATGGGCTCCGTA...</td>\n",
       "      <td>GGGCTCCGTAGGTTCCATGGGCTCCGTAGGTTCCATGGGCTCCGTA...</td>\n",
       "      <td>2/3</td>\n",
       "      <td>32680,4972,4719,6973,3392,6720,306,138,0,53</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10237</th>\n",
       "      <td>G</td>\n",
       "      <td>GGGCTCCGTAGGTTCCATGGGCTCCGTAGGTTCCATGGGCTCCGTA...</td>\n",
       "      <td>GGGCTCCGTAGGTTCCATGGGCTCCGTAGGTTCCATGGGCTCCGTA...</td>\n",
       "      <td>2/3</td>\n",
       "      <td>32680,4972,4719,6973,3392,6720,306,138,0,53</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10238</th>\n",
       "      <td>G</td>\n",
       "      <td>GGGCTCCGTAGGTTCCATGGGCTCCGTAGGTTCCATGGGCTCCGTA...</td>\n",
       "      <td>GGGCTCCGTAGGTTCCATGGGCTCCGTAGGTTCCATGGGCTCCGTA...</td>\n",
       "      <td>2/3</td>\n",
       "      <td>32680,4972,4719,6973,3392,6720,306,138,0,53</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10239</th>\n",
       "      <td>G</td>\n",
       "      <td>GGGCTCCGTAGGTTCCATGGGCTCCGTAGGTTCCATGGGCTCCGTA...</td>\n",
       "      <td>GGGCTCCGTAGGTTCCATGGGCTCCGTAGGTTCCATGGGCTCCGTA...</td>\n",
       "      <td>2/3</td>\n",
       "      <td>32680,4972,4719,6973,3392,6720,306,138,0,53</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10240</th>\n",
       "      <td>G</td>\n",
       "      <td>GGGCTCCGTAGGTTCCATGGGCTCCGTAGGTTCCATGGGCTCCGTA...</td>\n",
       "      <td>GGGCTCCGTAGGTTCCATGGGCTCCGTAGGTTCCATGGGCTCCGTA...</td>\n",
       "      <td>2/3</td>\n",
       "      <td>32680,4972,4719,6973,3392,6720,306,138,0,53</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10241</th>\n",
       "      <td>G</td>\n",
       "      <td>GGGCTCCGTAGGTTCCATGGGCTCCGTAGGTTCCATGGGCTCCGTA...</td>\n",
       "      <td>GGGCTCCGTAGGTTCCATGGGCTCCGTAGGTTCCATGGGCTCCGTA...</td>\n",
       "      <td>2/3</td>\n",
       "      <td>32680,4972,4719,6973,3392,6720,306,138,0,53</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10242</th>\n",
       "      <td>G</td>\n",
       "      <td>GGGCTCCGTAGGTTCCATGGGCTCCGTAGGTTCCATGGGCTCCGTA...</td>\n",
       "      <td>GGGCTCCGTAGGTTCCATGGGCTCCGTAGGTTCCATGGGCTCCGTA...</td>\n",
       "      <td>2/3</td>\n",
       "      <td>32680,4972,4719,6973,3392,6720,306,138,0,53</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                           REF  \\\n",
       "3044                     ATGTG   \n",
       "3045                     ATGTG   \n",
       "3046                     ATGTG   \n",
       "3047                     ATGTG   \n",
       "3048                     ATGTG   \n",
       "3049                     ATGTG   \n",
       "3050                     ATGTG   \n",
       "3051                     ATGTG   \n",
       "3052                     ATGTG   \n",
       "3845   CTCTCTGTGTGTGTGTGTGTGTG   \n",
       "3846   CTCTCTGTGTGTGTGTGTGTGTG   \n",
       "3847   CTCTCTGTGTGTGTGTGTGTGTG   \n",
       "3848   CTCTCTGTGTGTGTGTGTGTGTG   \n",
       "3849   CTCTCTGTGTGTGTGTGTGTGTG   \n",
       "3850   CTCTCTGTGTGTGTGTGTGTGTG   \n",
       "3851   CTCTCTGTGTGTGTGTGTGTGTG   \n",
       "3852   CTCTCTGTGTGTGTGTGTGTGTG   \n",
       "3853   CTCTCTGTGTGTGTGTGTGTGTG   \n",
       "4628                TAGCAGCAGC   \n",
       "4629                TAGCAGCAGC   \n",
       "4630                TAGCAGCAGC   \n",
       "4631                TAGCAGCAGC   \n",
       "4632                TAGCAGCAGC   \n",
       "4633                TAGCAGCAGC   \n",
       "10234                        G   \n",
       "10235                        G   \n",
       "10236                        G   \n",
       "10237                        G   \n",
       "10238                        G   \n",
       "10239                        G   \n",
       "10240                        G   \n",
       "10241                        G   \n",
       "10242                        G   \n",
       "\n",
       "                                                     ALT  \\\n",
       "3044                                   A,ATG,ATGTGTGTGTG   \n",
       "3045                                   A,ATG,ATGTGTGTGTG   \n",
       "3046                                   A,ATG,ATGTGTGTGTG   \n",
       "3047                                   A,ATG,ATGTGTGTGTG   \n",
       "3048                                   A,ATG,ATGTGTGTGTG   \n",
       "3049                                   A,ATG,ATGTGTGTGTG   \n",
       "3050                                   A,ATG,ATGTGTGTGTG   \n",
       "3051                                   A,ATG,ATGTGTGTGTG   \n",
       "3052                                   A,ATG,ATGTGTGTGTG   \n",
       "3845                                 CTGTGTGTGTGTG,CTG,C   \n",
       "3846                                 CTGTGTGTGTGTG,CTG,C   \n",
       "3847                                 CTGTGTGTGTGTG,CTG,C   \n",
       "3848                                 CTGTGTGTGTGTG,CTG,C   \n",
       "3849                                 CTGTGTGTGTGTG,CTG,C   \n",
       "3850                                 CTGTGTGTGTGTG,CTG,C   \n",
       "3851                                 CTGTGTGTGTGTG,CTG,C   \n",
       "3852                                 CTGTGTGTGTGTG,CTG,C   \n",
       "3853                                 CTGTGTGTGTGTG,CTG,C   \n",
       "4628                                      TAGC,T,TAGCAGC   \n",
       "4629                                      TAGC,T,TAGCAGC   \n",
       "4630                                      TAGC,T,TAGCAGC   \n",
       "4631                                      TAGC,T,TAGCAGC   \n",
       "4632                                      TAGC,T,TAGCAGC   \n",
       "4633                                      TAGC,T,TAGCAGC   \n",
       "10234  GGGCTCCGTAGGTTCCATGGGCTCCGTAGGTTCCATGGGCTCCGTA...   \n",
       "10235  GGGCTCCGTAGGTTCCATGGGCTCCGTAGGTTCCATGGGCTCCGTA...   \n",
       "10236  GGGCTCCGTAGGTTCCATGGGCTCCGTAGGTTCCATGGGCTCCGTA...   \n",
       "10237  GGGCTCCGTAGGTTCCATGGGCTCCGTAGGTTCCATGGGCTCCGTA...   \n",
       "10238  GGGCTCCGTAGGTTCCATGGGCTCCGTAGGTTCCATGGGCTCCGTA...   \n",
       "10239  GGGCTCCGTAGGTTCCATGGGCTCCGTAGGTTCCATGGGCTCCGTA...   \n",
       "10240  GGGCTCCGTAGGTTCCATGGGCTCCGTAGGTTCCATGGGCTCCGTA...   \n",
       "10241  GGGCTCCGTAGGTTCCATGGGCTCCGTAGGTTCCATGGGCTCCGTA...   \n",
       "10242  GGGCTCCGTAGGTTCCATGGGCTCCGTAGGTTCCATGGGCTCCGTA...   \n",
       "\n",
       "                                                SplitALT   GT  \\\n",
       "3044                                                   A  2/3   \n",
       "3045                                                 ATG  2/3   \n",
       "3046                                         ATGTGTGTGTG  2/3   \n",
       "3047                                                   A  2/3   \n",
       "3048                                                 ATG  2/3   \n",
       "3049                                         ATGTGTGTGTG  2/3   \n",
       "3050                                                   A  2/3   \n",
       "3051                                                 ATG  2/3   \n",
       "3052                                         ATGTGTGTGTG  2/3   \n",
       "3845                                       CTGTGTGTGTGTG  2/3   \n",
       "3846                                                 CTG  2/3   \n",
       "3847                                                   C  2/3   \n",
       "3848                                       CTGTGTGTGTGTG  2/3   \n",
       "3849                                                 CTG  2/3   \n",
       "3850                                                   C  2/3   \n",
       "3851                                       CTGTGTGTGTGTG  2/3   \n",
       "3852                                                 CTG  2/3   \n",
       "3853                                                   C  2/3   \n",
       "4628                                                TAGC  2/3   \n",
       "4629                                                   T  2/3   \n",
       "4630                                             TAGCAGC  2/3   \n",
       "4631                                                TAGC  2/3   \n",
       "4632                                                   T  2/3   \n",
       "4633                                             TAGCAGC  2/3   \n",
       "10234  GGGCTCCGTAGGTTCCATGGGCTCCGTAGGTTCCATGGGCTCCGTA...  2/3   \n",
       "10235  GGGCTCCGTAGGTTCCATGGGCTCCGTAGGTTCCATGGGCTCCGTA...  2/3   \n",
       "10236  GGGCTCCGTAGGTTCCATGGGCTCCGTAGGTTCCATGGGCTCCGTA...  2/3   \n",
       "10237  GGGCTCCGTAGGTTCCATGGGCTCCGTAGGTTCCATGGGCTCCGTA...  2/3   \n",
       "10238  GGGCTCCGTAGGTTCCATGGGCTCCGTAGGTTCCATGGGCTCCGTA...  2/3   \n",
       "10239  GGGCTCCGTAGGTTCCATGGGCTCCGTAGGTTCCATGGGCTCCGTA...  2/3   \n",
       "10240  GGGCTCCGTAGGTTCCATGGGCTCCGTAGGTTCCATGGGCTCCGTA...  2/3   \n",
       "10241  GGGCTCCGTAGGTTCCATGGGCTCCGTAGGTTCCATGGGCTCCGTA...  2/3   \n",
       "10242  GGGCTCCGTAGGTTCCATGGGCTCCGTAGGTTCCATGGGCTCCGTA...  2/3   \n",
       "\n",
       "                                                   PL  \n",
       "3044         978,1028,1563,771,1053,990,270,291,0,625  \n",
       "3045         978,1028,1563,771,1053,990,270,291,0,625  \n",
       "3046         978,1028,1563,771,1053,990,270,291,0,625  \n",
       "3047         978,1028,1563,771,1053,990,270,291,0,625  \n",
       "3048         978,1028,1563,771,1053,990,270,291,0,625  \n",
       "3049         978,1028,1563,771,1053,990,270,291,0,625  \n",
       "3050         978,1028,1563,771,1053,990,270,291,0,625  \n",
       "3051         978,1028,1563,771,1053,990,270,291,0,625  \n",
       "3052         978,1028,1563,771,1053,990,270,291,0,625  \n",
       "3845         2783,1303,1255,243,214,195,221,167,0,173  \n",
       "3846         2783,1303,1255,243,214,195,221,167,0,173  \n",
       "3847         2783,1303,1255,243,214,195,221,167,0,173  \n",
       "3848         2783,1303,1255,243,214,195,221,167,0,173  \n",
       "3849         2783,1303,1255,243,214,195,221,167,0,173  \n",
       "3850         2783,1303,1255,243,214,195,221,167,0,173  \n",
       "3851         2783,1303,1255,243,214,195,221,167,0,173  \n",
       "3852         2783,1303,1255,243,214,195,221,167,0,173  \n",
       "3853         2783,1303,1255,243,214,195,221,167,0,173  \n",
       "4628   4058,2218,2300,1039,1097,1811,2190,1230,0,2077  \n",
       "4629   4058,2218,2300,1039,1097,1811,2190,1230,0,2077  \n",
       "4630   4058,2218,2300,1039,1097,1811,2190,1230,0,2077  \n",
       "4631   4058,2218,2300,1039,1097,1811,2190,1230,0,2077  \n",
       "4632   4058,2218,2300,1039,1097,1811,2190,1230,0,2077  \n",
       "4633   4058,2218,2300,1039,1097,1811,2190,1230,0,2077  \n",
       "10234     32680,4972,4719,6973,3392,6720,306,138,0,53  \n",
       "10235     32680,4972,4719,6973,3392,6720,306,138,0,53  \n",
       "10236     32680,4972,4719,6973,3392,6720,306,138,0,53  \n",
       "10237     32680,4972,4719,6973,3392,6720,306,138,0,53  \n",
       "10238     32680,4972,4719,6973,3392,6720,306,138,0,53  \n",
       "10239     32680,4972,4719,6973,3392,6720,306,138,0,53  \n",
       "10240     32680,4972,4719,6973,3392,6720,306,138,0,53  \n",
       "10241     32680,4972,4719,6973,3392,6720,306,138,0,53  \n",
       "10242     32680,4972,4719,6973,3392,6720,306,138,0,53  "
      ]
     },
     "execution_count": 89,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "processed_df.loc[((processed_df['GT'] == '2/3')), ['REF', 'ALT', 'SplitALT', 'GT', 'PL']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 90,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>SplitALT</th>\n",
       "      <th>GT</th>\n",
       "      <th>PL</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>3044</th>\n",
       "      <td>ATGTG</td>\n",
       "      <td>A,ATG,ATGTGTGTGTG</td>\n",
       "      <td>A</td>\n",
       "      <td>2/3</td>\n",
       "      <td>978,1028,1563,771,1053,990,270,291,0,625</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3045</th>\n",
       "      <td>ATGTG</td>\n",
       "      <td>A,ATG,ATGTGTGTGTG</td>\n",
       "      <td>ATG</td>\n",
       "      <td>2/3</td>\n",
       "      <td>978,1028,1563,771,1053,990,270,291,0,625</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3046</th>\n",
       "      <td>ATGTG</td>\n",
       "      <td>A,ATG,ATGTGTGTGTG</td>\n",
       "      <td>ATGTGTGTGTG</td>\n",
       "      <td>2/3</td>\n",
       "      <td>978,1028,1563,771,1053,990,270,291,0,625</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3845</th>\n",
       "      <td>CTCTCTGTGTGTGTGTGTGTGTG</td>\n",
       "      <td>CTGTGTGTGTGTG,CTG,C</td>\n",
       "      <td>CTGTGTGTGTGTG</td>\n",
       "      <td>2/3</td>\n",
       "      <td>2783,1303,1255,243,214,195,221,167,0,173</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3846</th>\n",
       "      <td>CTCTCTGTGTGTGTGTGTGTGTG</td>\n",
       "      <td>CTGTGTGTGTGTG,CTG,C</td>\n",
       "      <td>CTG</td>\n",
       "      <td>2/3</td>\n",
       "      <td>2783,1303,1255,243,214,195,221,167,0,173</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3847</th>\n",
       "      <td>CTCTCTGTGTGTGTGTGTGTGTG</td>\n",
       "      <td>CTGTGTGTGTGTG,CTG,C</td>\n",
       "      <td>C</td>\n",
       "      <td>2/3</td>\n",
       "      <td>2783,1303,1255,243,214,195,221,167,0,173</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4628</th>\n",
       "      <td>TAGCAGCAGC</td>\n",
       "      <td>TAGC,T,TAGCAGC</td>\n",
       "      <td>TAGC</td>\n",
       "      <td>2/3</td>\n",
       "      <td>4058,2218,2300,1039,1097,1811,2190,1230,0,2077</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4629</th>\n",
       "      <td>TAGCAGCAGC</td>\n",
       "      <td>TAGC,T,TAGCAGC</td>\n",
       "      <td>T</td>\n",
       "      <td>2/3</td>\n",
       "      <td>4058,2218,2300,1039,1097,1811,2190,1230,0,2077</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4630</th>\n",
       "      <td>TAGCAGCAGC</td>\n",
       "      <td>TAGC,T,TAGCAGC</td>\n",
       "      <td>TAGCAGC</td>\n",
       "      <td>2/3</td>\n",
       "      <td>4058,2218,2300,1039,1097,1811,2190,1230,0,2077</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10234</th>\n",
       "      <td>G</td>\n",
       "      <td>GGGCTCCGTAGGTTCCATGGGCTCCGTAGGTTCCATGGGCTCCGTA...</td>\n",
       "      <td>GGGCTCCGTAGGTTCCATGGGCTCCGTAGGTTCCATGGGCTCCGTA...</td>\n",
       "      <td>2/3</td>\n",
       "      <td>32680,4972,4719,6973,3392,6720,306,138,0,53</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10235</th>\n",
       "      <td>G</td>\n",
       "      <td>GGGCTCCGTAGGTTCCATGGGCTCCGTAGGTTCCATGGGCTCCGTA...</td>\n",
       "      <td>GGGCTCCGTAGGTTCCATGGGCTCCGTAGGTTCCATGGGCTCCGTA...</td>\n",
       "      <td>2/3</td>\n",
       "      <td>32680,4972,4719,6973,3392,6720,306,138,0,53</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10236</th>\n",
       "      <td>G</td>\n",
       "      <td>GGGCTCCGTAGGTTCCATGGGCTCCGTAGGTTCCATGGGCTCCGTA...</td>\n",
       "      <td>GGGCTCCGTAGGTTCCATGGGCTCCGTAGGTTCCATGGGCTCCGTA...</td>\n",
       "      <td>2/3</td>\n",
       "      <td>32680,4972,4719,6973,3392,6720,306,138,0,53</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                           REF  \\\n",
       "3044                     ATGTG   \n",
       "3045                     ATGTG   \n",
       "3046                     ATGTG   \n",
       "3845   CTCTCTGTGTGTGTGTGTGTGTG   \n",
       "3846   CTCTCTGTGTGTGTGTGTGTGTG   \n",
       "3847   CTCTCTGTGTGTGTGTGTGTGTG   \n",
       "4628                TAGCAGCAGC   \n",
       "4629                TAGCAGCAGC   \n",
       "4630                TAGCAGCAGC   \n",
       "10234                        G   \n",
       "10235                        G   \n",
       "10236                        G   \n",
       "\n",
       "                                                     ALT  \\\n",
       "3044                                   A,ATG,ATGTGTGTGTG   \n",
       "3045                                   A,ATG,ATGTGTGTGTG   \n",
       "3046                                   A,ATG,ATGTGTGTGTG   \n",
       "3845                                 CTGTGTGTGTGTG,CTG,C   \n",
       "3846                                 CTGTGTGTGTGTG,CTG,C   \n",
       "3847                                 CTGTGTGTGTGTG,CTG,C   \n",
       "4628                                      TAGC,T,TAGCAGC   \n",
       "4629                                      TAGC,T,TAGCAGC   \n",
       "4630                                      TAGC,T,TAGCAGC   \n",
       "10234  GGGCTCCGTAGGTTCCATGGGCTCCGTAGGTTCCATGGGCTCCGTA...   \n",
       "10235  GGGCTCCGTAGGTTCCATGGGCTCCGTAGGTTCCATGGGCTCCGTA...   \n",
       "10236  GGGCTCCGTAGGTTCCATGGGCTCCGTAGGTTCCATGGGCTCCGTA...   \n",
       "\n",
       "                                                SplitALT   GT  \\\n",
       "3044                                                   A  2/3   \n",
       "3045                                                 ATG  2/3   \n",
       "3046                                         ATGTGTGTGTG  2/3   \n",
       "3845                                       CTGTGTGTGTGTG  2/3   \n",
       "3846                                                 CTG  2/3   \n",
       "3847                                                   C  2/3   \n",
       "4628                                                TAGC  2/3   \n",
       "4629                                                   T  2/3   \n",
       "4630                                             TAGCAGC  2/3   \n",
       "10234  GGGCTCCGTAGGTTCCATGGGCTCCGTAGGTTCCATGGGCTCCGTA...  2/3   \n",
       "10235  GGGCTCCGTAGGTTCCATGGGCTCCGTAGGTTCCATGGGCTCCGTA...  2/3   \n",
       "10236  GGGCTCCGTAGGTTCCATGGGCTCCGTAGGTTCCATGGGCTCCGTA...  2/3   \n",
       "\n",
       "                                                   PL  \n",
       "3044         978,1028,1563,771,1053,990,270,291,0,625  \n",
       "3045         978,1028,1563,771,1053,990,270,291,0,625  \n",
       "3046         978,1028,1563,771,1053,990,270,291,0,625  \n",
       "3845         2783,1303,1255,243,214,195,221,167,0,173  \n",
       "3846         2783,1303,1255,243,214,195,221,167,0,173  \n",
       "3847         2783,1303,1255,243,214,195,221,167,0,173  \n",
       "4628   4058,2218,2300,1039,1097,1811,2190,1230,0,2077  \n",
       "4629   4058,2218,2300,1039,1097,1811,2190,1230,0,2077  \n",
       "4630   4058,2218,2300,1039,1097,1811,2190,1230,0,2077  \n",
       "10234     32680,4972,4719,6973,3392,6720,306,138,0,53  \n",
       "10235     32680,4972,4719,6973,3392,6720,306,138,0,53  \n",
       "10236     32680,4972,4719,6973,3392,6720,306,138,0,53  "
      ]
     },
     "execution_count": 90,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df2.loc[((df2['GT'] == '2/3')), ['REF', 'ALT', 'SplitALT', 'GT', 'PL']].head(50)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "metadata": {},
   "outputs": [],
   "source": [
    "df2 = processed_df.drop_duplicates(subset=['CHROM', 'POS', 'ALT', 'SplitALT', 'PL'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "11471\n",
      "12181\n",
      "11481\n"
     ]
    }
   ],
   "source": [
    "print(len(df))\n",
    "print(len(processed_df))\n",
    "print(len(df2))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chr</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Ref</th>\n",
       "      <th>Alt</th>\n",
       "      <th>Func.refGene</th>\n",
       "      <th>Gene.refGene</th>\n",
       "      <th>GeneDetail.refGene</th>\n",
       "      <th>ExonicFunc.refGene</th>\n",
       "      <th>AAChange.refGene</th>\n",
       "      <th>...</th>\n",
       "      <th>_32</th>\n",
       "      <th>_39</th>\n",
       "      <th>_57</th>\n",
       "      <th>_58</th>\n",
       "      <th>_65</th>\n",
       "      <th>_66</th>\n",
       "      <th>_67</th>\n",
       "      <th>_68</th>\n",
       "      <th>SplitALT</th>\n",
       "      <th>_merge</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>1</td>\n",
       "      <td>1247579</td>\n",
       "      <td>1247579</td>\n",
       "      <td>G</td>\n",
       "      <td>-</td>\n",
       "      <td>splicing</td>\n",
       "      <td>INTS11</td>\n",
       "      <td>NM_001256463:exon13:c.1304+27C&gt;-;NM_001256462:...</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>left_only</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>381</th>\n",
       "      <td>1</td>\n",
       "      <td>47610436</td>\n",
       "      <td>47610436</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>splicing</td>\n",
       "      <td>CYP4A22</td>\n",
       "      <td>NM_001308102:exon7:c.743-21G&gt;C;NM_001010969:ex...</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>left_only</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>401</th>\n",
       "      <td>1</td>\n",
       "      <td>54605318</td>\n",
       "      <td>54605318</td>\n",
       "      <td>-</td>\n",
       "      <td>G</td>\n",
       "      <td>exonic</td>\n",
       "      <td>CDCP2</td>\n",
       "      <td>.</td>\n",
       "      <td>frameshift insertion</td>\n",
       "      <td>CDCP2:NM_201546:exon5:c.1224dupC:p.L409fs</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>left_only</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>410</th>\n",
       "      <td>1</td>\n",
       "      <td>55505558</td>\n",
       "      <td>55505558</td>\n",
       "      <td>-</td>\n",
       "      <td>CTG</td>\n",
       "      <td>exonic</td>\n",
       "      <td>PCSK9</td>\n",
       "      <td>.</td>\n",
       "      <td>nonframeshift insertion</td>\n",
       "      <td>PCSK9:NM_174936:exon1:c.48_49insCTG:p.L23_G24insL</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>left_only</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>413</th>\n",
       "      <td>1</td>\n",
       "      <td>55524340</td>\n",
       "      <td>55524355</td>\n",
       "      <td>GTGTGTGTGTGTGTGT</td>\n",
       "      <td>-</td>\n",
       "      <td>splicing</td>\n",
       "      <td>PCSK9</td>\n",
       "      <td>NM_174936:exon9:r.spl</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>left_only</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11583</th>\n",
       "      <td>X</td>\n",
       "      <td>27765403</td>\n",
       "      <td>27765405</td>\n",
       "      <td>GAG</td>\n",
       "      <td>-</td>\n",
       "      <td>exonic</td>\n",
       "      <td>DCAF8L2</td>\n",
       "      <td>.</td>\n",
       "      <td>unknown</td>\n",
       "      <td>UNKNOWN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>left_only</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11595</th>\n",
       "      <td>X</td>\n",
       "      <td>44935925</td>\n",
       "      <td>44935926</td>\n",
       "      <td>TT</td>\n",
       "      <td>-</td>\n",
       "      <td>splicing</td>\n",
       "      <td>KDM6A</td>\n",
       "      <td>NM_001291418:exon16:r.spl;NM_001291421:exon16:...</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>left_only</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11635</th>\n",
       "      <td>X</td>\n",
       "      <td>56296780</td>\n",
       "      <td>56296780</td>\n",
       "      <td>-</td>\n",
       "      <td>ACACAC</td>\n",
       "      <td>splicing</td>\n",
       "      <td>KLF8</td>\n",
       "      <td>NM_001324105:exon5:c.889+26-&gt;ACACAC;NM_0013241...</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>left_only</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11645</th>\n",
       "      <td>X</td>\n",
       "      <td>66765159</td>\n",
       "      <td>66765161</td>\n",
       "      <td>GCA</td>\n",
       "      <td>-</td>\n",
       "      <td>exonic</td>\n",
       "      <td>AR</td>\n",
       "      <td>.</td>\n",
       "      <td>nonframeshift deletion</td>\n",
       "      <td>AR:NM_000044:exon1:c.171_173del:p.Q80del,AR:NM...</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>left_only</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11733</th>\n",
       "      <td>X</td>\n",
       "      <td>118985711</td>\n",
       "      <td>118985711</td>\n",
       "      <td>-</td>\n",
       "      <td>TTTTTTTTTTTT</td>\n",
       "      <td>splicing</td>\n",
       "      <td>UPF3B</td>\n",
       "      <td>NM_023010:exon2:c.263+19-&gt;AAAAAAAAAAAA;NM_0806...</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>left_only</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>329 rows × 109 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      Chr      Start        End               Ref           Alt Func.refGene  \\\n",
       "25      1    1247579    1247579                 G             -     splicing   \n",
       "381     1   47610436   47610436                 G             C     splicing   \n",
       "401     1   54605318   54605318                 -             G       exonic   \n",
       "410     1   55505558   55505558                 -           CTG       exonic   \n",
       "413     1   55524340   55524355  GTGTGTGTGTGTGTGT             -     splicing   \n",
       "...    ..        ...        ...               ...           ...          ...   \n",
       "11583   X   27765403   27765405               GAG             -       exonic   \n",
       "11595   X   44935925   44935926                TT             -     splicing   \n",
       "11635   X   56296780   56296780                 -        ACACAC     splicing   \n",
       "11645   X   66765159   66765161               GCA             -       exonic   \n",
       "11733   X  118985711  118985711                 -  TTTTTTTTTTTT     splicing   \n",
       "\n",
       "      Gene.refGene                                 GeneDetail.refGene  \\\n",
       "25          INTS11  NM_001256463:exon13:c.1304+27C>-;NM_001256462:...   \n",
       "381        CYP4A22  NM_001308102:exon7:c.743-21G>C;NM_001010969:ex...   \n",
       "401          CDCP2                                                  .   \n",
       "410          PCSK9                                                  .   \n",
       "413          PCSK9                              NM_174936:exon9:r.spl   \n",
       "...            ...                                                ...   \n",
       "11583      DCAF8L2                                                  .   \n",
       "11595        KDM6A  NM_001291418:exon16:r.spl;NM_001291421:exon16:...   \n",
       "11635         KLF8  NM_001324105:exon5:c.889+26->ACACAC;NM_0013241...   \n",
       "11645           AR                                                  .   \n",
       "11733        UPF3B  NM_023010:exon2:c.263+19->AAAAAAAAAAAA;NM_0806...   \n",
       "\n",
       "            ExonicFunc.refGene  \\\n",
       "25                           .   \n",
       "381                          .   \n",
       "401       frameshift insertion   \n",
       "410    nonframeshift insertion   \n",
       "413                          .   \n",
       "...                        ...   \n",
       "11583                  unknown   \n",
       "11595                        .   \n",
       "11635                        .   \n",
       "11645   nonframeshift deletion   \n",
       "11733                        .   \n",
       "\n",
       "                                        AAChange.refGene  ...  _32  _39  _57  \\\n",
       "25                                                     .  ...  NaN  NaN  NaN   \n",
       "381                                                    .  ...  NaN  NaN  NaN   \n",
       "401            CDCP2:NM_201546:exon5:c.1224dupC:p.L409fs  ...  NaN  NaN  NaN   \n",
       "410    PCSK9:NM_174936:exon1:c.48_49insCTG:p.L23_G24insL  ...  NaN  NaN  NaN   \n",
       "413                                                    .  ...  NaN  NaN  NaN   \n",
       "...                                                  ...  ...  ...  ...  ...   \n",
       "11583                                            UNKNOWN  ...  NaN  NaN  NaN   \n",
       "11595                                                  .  ...  NaN  NaN  NaN   \n",
       "11635                                                  .  ...  NaN  NaN  NaN   \n",
       "11645  AR:NM_000044:exon1:c.171_173del:p.Q80del,AR:NM...  ...  NaN  NaN  NaN   \n",
       "11733                                                  .  ...  NaN  NaN  NaN   \n",
       "\n",
       "       _58  _65  _66  _67  _68 SplitALT     _merge  \n",
       "25     NaN  NaN  NaN  NaN  NaN      NaN  left_only  \n",
       "381    NaN  NaN  NaN  NaN  NaN      NaN  left_only  \n",
       "401    NaN  NaN  NaN  NaN  NaN      NaN  left_only  \n",
       "410    NaN  NaN  NaN  NaN  NaN      NaN  left_only  \n",
       "413    NaN  NaN  NaN  NaN  NaN      NaN  left_only  \n",
       "...    ...  ...  ...  ...  ...      ...        ...  \n",
       "11583  NaN  NaN  NaN  NaN  NaN      NaN  left_only  \n",
       "11595  NaN  NaN  NaN  NaN  NaN      NaN  left_only  \n",
       "11635  NaN  NaN  NaN  NaN  NaN      NaN  left_only  \n",
       "11645  NaN  NaN  NaN  NaN  NaN      NaN  left_only  \n",
       "11733  NaN  NaN  NaN  NaN  NaN      NaN  left_only  \n",
       "\n",
       "[329 rows x 109 columns]"
      ]
     },
     "execution_count": 62,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pd.merge(df,df2, how =\"outer\", indicator=True).query(f'_merge != \"both\"')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def separate_alt(info):\n",
    "    info = info.split(':')\n",
    "    separated_alt = []\n",
    "    return separated_alt\n",
    "\n",
    "def generate_tmp_vcf(df, threads) -> pysam.TabixFile:\n",
    "    #1. Sort by CHROM and POS\n",
    "    df = df.astype({'CHROM': str, 'POS': int})\n",
    "    df = df.sort_values(by=['CHROM', 'POS'])\n",
    "\n",
    "    #2. Generate VCF file\n",
    "    df_vcf = df[['CHROM', 'POS', 'ID', 'REF', 'ALT', 'FILTER', 'QUAL', mode_samples_info.proband_id, 'FORMAT']]\n",
    "    df_vcf.columns = ['#CHROM', 'POS', 'ID', 'REF', 'ALT', 'FILTER', 'QUAL', 'INFO', 'FORMAT']\n",
    "    df_vcf = df_vcf.drop_duplicates(subset=['#CHROM', 'POS', 'REF', 'ALT'])\n",
    "    df_vcf.to_csv('./temp.vcf', sep='\\t', index=False, header=True)\n",
    "    \n",
    "    #3. Compress and Index\n",
    "    pysam.tabix_compress('./temp.vcf', './temp.vcf.gz', force=True)\n",
    "    pysam.tabix_index('./temp.vcf.gz', preset='vcf', force=True)\n",
    "    \n",
    "    #4. Load and Return as TabixFile\n",
    "    tbx_vcf = pysam.TabixFile('./temp.vcf.gz', threads=threads)\n",
    "\n",
    "    return tbx_vcf\n",
    "\n",
    "{\n",
    "            'gnomad40_genome_AF': '.', 'gnomad40_genome_AF_afr': '.',\n",
    "            'gnomad40_genome_AF_ami': '.', 'gnomad40_genome_AF_amr': '.',\n",
    "            'gnomad40_genome_AF_asj': '.', 'gnomad40_genome_AF_eas': '.',\n",
    "            'gnomad40_genome_AF_fin': '.', 'gnomad40_genome_AF_mid': '.',\n",
    "            'gnomad40_genome_AF_nfe': '.', 'gnomad40_genome_AF_remaining': '.',\n",
    "            'gnomad40_genome_AF_sas': '.', 'ALL.sites.2015_08': '.',\n",
    "            'HRC_AF': '.', 'HRC_non1000G_AF': '.', 'Kaviar_AF': '.', \n",
    "            'GME_AF': '.', 'GME_NWA': '.', 'GME_NEA': '.', 'GME_AP': '.', \n",
    "            'GME_Israel': '.', 'GME_SD': '.', 'GME_TP': '.', 'GME_CA': '.'\n",
    "            }\n",
    "\n",
    "df = df.astype({'CHROM': str, 'POS': int})\n",
    "df = df.sort_values(by=['CHROM', 'POS'])\n",
    "df_vcf = df[['CHROM', 'POS', 'ID', 'REF', 'ALT', 'FILTER', 'QUAL', mode_samples_info.proband_id, 'FORMAT']]\n",
    "df_vcf.columns = ['#CHROM', 'POS', 'ID', 'REF', 'ALT', 'FILTER', 'QUAL', 'INFO', 'FORMAT']\n",
    "df_vcf = df_vcf.drop_duplicates(subset=['#CHROM', 'POS', 'REF', 'ALT'])\n",
    "df_vcf.to_csv('./temp.vcf', sep='\\t', index=False, header=True)\n",
    "\n",
    "pysam.tabix_compress('./temp.vcf', './temp.vcf.gz', force=True)\n",
    "pysam.tabix_index('./temp.vcf.gz', preset='vcf', force=True)\n",
    "tbx_vcf = pysam.TabixFile('./temp.vcf.gz', threads=4)\n",
    "\n",
    "\n",
    "for row in tbx_vcf.fetch(parser=pysam.asVCF()):\n",
    "    gt = row.info.split(':')[0]\n",
    "    for alt in row.alt.split(','):\n",
    "        print(alt, gt, sep='\\t')\n",
    "        break\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "o = './temp_out.vcf'\n",
    "with open(o, 'w') as outfh:\n",
    "    vcf = tbx_vcf.fetch(parser=pysam.asVCF())\n",
    "    for record in vcf:\n",
    "        for sample in snp_record.samples:\n",
    "            gt_list.append(sample['GT'])\n",
    "        \n",
    "        new_record = []\n",
    "        \n",
    "        new_record.extend(gt_list)\n",
    "        \n",
    "        outfh.write('\\t'.join(new_record) + '\\n')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for row in tbx_vcf.fetch(parser=pysam.asVCF()):\n",
    "    # 各行（row）に対する処理をここに記述\n",
    "    # 例えば、行の内容を印刷する\n",
    "    print(row)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'./temp.tsv.gz'"
      ]
     },
     "execution_count": 61,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pysam.tabix_index('./temp.tsv.gz', preset='vcf', force=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "\n",
    "# VCFデータのDataFrameを作成\n",
    "df_vcf = df[['CHROM', 'POS', 'REF', 'ALT', 'FILTER', 'QUAL', 'FORMAT', mode_samples_info.proband_id]]\n",
    "df_vcf.columns = ['#CHROM', 'POS', 'REF', 'ALT', 'FILTER', 'QUAL', 'FORMAT', 'INFO']\n",
    "\n",
    "# VCFのヘッダーを用意\n",
    "vcf_header = \"\"\"##fileformat=VCFv4.2\n",
    "\n",
    "\n",
    "#CHROM\\tPOS\\tID\\tREF\\tALT\\tQUAL\\tFILTER\\tINFO\\tFORMAT\\tSAMPLE\n",
    "\"\"\"\n",
    "\n",
    "# ヘッダーとデータを組み合わせてファイルに書き出す\n",
    "with open('./temp.vcf', 'w') as outfile:\n",
    "    outfile.write(vcf_header)\n",
    "    df_vcf.to_csv(outfile, sep='\\t', index=False, header=False)\n",
    "\n",
    "# pysamでVCFファイルを読み込む\n",
    "vcf = pysam.VariantFile('./temp.vcf')\n",
    "# filename = './temp.vcf.idx'\n",
    "# vcf.index(filename=filename)\n",
    "# # ファイルを閉じる\n",
    "# vcf.close()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "[W::vcf_parse] Contig '1' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse_filter] FILTER 'GT:AD:DP:GQ:PL:JpnAD2' is not defined in the header\n",
      "[W::vcf_parse_info] INFO '0/1:18,16:34:99:324,0,405:0.471' is not defined in the header, assuming Type=String\n",
      "[E::bcf_hdr_parse_line] Could not parse the header line: \"##INFO=<ID=0/1:18,16:34:99:324,0,405:0.471,Number=1,Type=String,Description=\\\"Dummy\\\">\"\n",
      "[E::vcf_parse_info] Could not add dummy header for INFO '0/1:18,16:34:99:324,0,405:0.471' at 1:871193\n"
     ]
    },
    {
     "ename": "OSError",
     "evalue": "unable to parse next record",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mOSError\u001b[0m                                   Traceback (most recent call last)",
      "\u001b[1;32m/work/Github/wesanno/wesanno/test.ipynb Cell 7\u001b[0m line \u001b[0;36m1\n\u001b[0;32m----> <a href='vscode-notebook-cell://attached-container%2B7b22636f6e7461696e65724e616d65223a222f696e73706972696e675f6c616c616e6465222c2273657474696e6773223a7b22636f6e74657874223a226465736b746f702d6c696e7578227d7d/work/Github/wesanno/wesanno/test.ipynb#X65sdnNjb2RlLXJlbW90ZQ%3D%3D?line=0'>1</a>\u001b[0m \u001b[39mfor\u001b[39;00m rec \u001b[39min\u001b[39;00m vcf:\n\u001b[1;32m      <a href='vscode-notebook-cell://attached-container%2B7b22636f6e7461696e65724e616d65223a222f696e73706972696e675f6c616c616e6465222c2273657474696e6773223a7b22636f6e74657874223a226465736b746f702d6c696e7578227d7d/work/Github/wesanno/wesanno/test.ipynb#X65sdnNjb2RlLXJlbW90ZQ%3D%3D?line=1'>2</a>\u001b[0m     \u001b[39mprint\u001b[39m(rec)\n\u001b[1;32m      <a href='vscode-notebook-cell://attached-container%2B7b22636f6e7461696e65724e616d65223a222f696e73706972696e675f6c616c616e6465222c2273657474696e6773223a7b22636f6e74657874223a226465736b746f702d6c696e7578227d7d/work/Github/wesanno/wesanno/test.ipynb#X65sdnNjb2RlLXJlbW90ZQ%3D%3D?line=2'>3</a>\u001b[0m     \u001b[39mbreak\u001b[39;00m\n",
      "File \u001b[0;32m/opt/conda/envs/wesanno/lib/python3.10/site-packages/pysam/libcbcf.pyx:4189\u001b[0m, in \u001b[0;36mpysam.libcbcf.VariantFile.__next__\u001b[0;34m()\u001b[0m\n",
      "\u001b[0;31mOSError\u001b[0m: unable to parse next record"
     ]
    }
   ],
   "source": [
    "for rec in vcf:\n",
    "    print(rec)\n",
    "    break"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"GT:AD:DP:GQ:PL:JpnAD2\"\"\"\n",
    "# df['AD'] = df[mode_samples_info.proband_id].str.split(':')[1]\n",
    "# df['DP'] = df[mode_samples_info.proband_id].str.split(':')[2]\n",
    "# df['GQ'] = df[mode_samples_info.proband_id].str.split(':')[3]\n",
    "# df['PL'] = df[mode_samples_info.proband_id].str.split(':')[4]\n",
    "\n",
    "for i, new_col in enumerate(['GT', 'AD', 'DP', 'GQ', 'PL']):\n",
    "    df[new_col] = df[mode_samples_info.proband_id].str.split(':').str[i]\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>GT</th>\n",
       "      <th>GQ</th>\n",
       "      <th>PL</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1433</th>\n",
       "      <td>TAA</td>\n",
       "      <td>TA,T</td>\n",
       "      <td>1/2</td>\n",
       "      <td>17</td>\n",
       "      <td>373,20,17,301,0,389</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1434</th>\n",
       "      <td>TAA</td>\n",
       "      <td>TA,T</td>\n",
       "      <td>1/2</td>\n",
       "      <td>17</td>\n",
       "      <td>373,20,17,301,0,389</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2789</th>\n",
       "      <td>G</td>\n",
       "      <td>GGTGT,GGT</td>\n",
       "      <td>1/2</td>\n",
       "      <td>6</td>\n",
       "      <td>148,32,151,6,0,61</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2790</th>\n",
       "      <td>G</td>\n",
       "      <td>GGTGT,GGT</td>\n",
       "      <td>1/2</td>\n",
       "      <td>6</td>\n",
       "      <td>148,32,151,6,0,61</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3148</th>\n",
       "      <td>GAA</td>\n",
       "      <td>GA,G</td>\n",
       "      <td>1/2</td>\n",
       "      <td>14</td>\n",
       "      <td>363,14,37,252,0,410</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3149</th>\n",
       "      <td>GAA</td>\n",
       "      <td>GA,G</td>\n",
       "      <td>1/2</td>\n",
       "      <td>14</td>\n",
       "      <td>363,14,37,252,0,410</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4053</th>\n",
       "      <td>T</td>\n",
       "      <td>TAA,TA</td>\n",
       "      <td>1/2</td>\n",
       "      <td>16</td>\n",
       "      <td>242,16,153,49,0,128</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4054</th>\n",
       "      <td>T</td>\n",
       "      <td>TAA,TA</td>\n",
       "      <td>1/2</td>\n",
       "      <td>16</td>\n",
       "      <td>242,16,153,49,0,128</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4164</th>\n",
       "      <td>G</td>\n",
       "      <td>GAAAAAAA,GAAAAAA</td>\n",
       "      <td>1/2</td>\n",
       "      <td>15</td>\n",
       "      <td>522,72,96,15,0,19</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4165</th>\n",
       "      <td>G</td>\n",
       "      <td>GAAAAAAA,GAAAAAA</td>\n",
       "      <td>1/2</td>\n",
       "      <td>15</td>\n",
       "      <td>522,72,96,15,0,19</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5449</th>\n",
       "      <td>GAAAA</td>\n",
       "      <td>G,GA</td>\n",
       "      <td>1/2</td>\n",
       "      <td>19</td>\n",
       "      <td>431,164,180,57,0,19</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5450</th>\n",
       "      <td>GAAAA</td>\n",
       "      <td>G,GA</td>\n",
       "      <td>1/2</td>\n",
       "      <td>19</td>\n",
       "      <td>431,164,180,57,0,19</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6557</th>\n",
       "      <td>AGTGT</td>\n",
       "      <td>A,AGT</td>\n",
       "      <td>1/2</td>\n",
       "      <td>16</td>\n",
       "      <td>188,140,211,31,0,16</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6558</th>\n",
       "      <td>AGTGT</td>\n",
       "      <td>A,AGT</td>\n",
       "      <td>1/2</td>\n",
       "      <td>16</td>\n",
       "      <td>188,140,211,31,0,16</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7743</th>\n",
       "      <td>TAA</td>\n",
       "      <td>TA,T</td>\n",
       "      <td>1/2</td>\n",
       "      <td>16</td>\n",
       "      <td>179,19,16,131,0,192</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7744</th>\n",
       "      <td>TAA</td>\n",
       "      <td>TA,T</td>\n",
       "      <td>1/2</td>\n",
       "      <td>16</td>\n",
       "      <td>179,19,16,131,0,192</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10678</th>\n",
       "      <td>ATGTGTGTGTG</td>\n",
       "      <td>ATGTG,A</td>\n",
       "      <td>1/2</td>\n",
       "      <td>16</td>\n",
       "      <td>600,43,16,334,0,331</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10687</th>\n",
       "      <td>TA</td>\n",
       "      <td>TAA,T</td>\n",
       "      <td>1/2</td>\n",
       "      <td>17</td>\n",
       "      <td>151,17,215,50,0,244</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10688</th>\n",
       "      <td>TA</td>\n",
       "      <td>TAA,T</td>\n",
       "      <td>1/2</td>\n",
       "      <td>17</td>\n",
       "      <td>151,17,215,50,0,244</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11393</th>\n",
       "      <td>C</td>\n",
       "      <td>CTTTTTTTTTTTTT,CTTTTTTTTTTTT</td>\n",
       "      <td>1/2</td>\n",
       "      <td>16</td>\n",
       "      <td>566,34,16,40,0,22</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11394</th>\n",
       "      <td>C</td>\n",
       "      <td>CTTTTTTTTTTTTT,CTTTTTTTTTTTT</td>\n",
       "      <td>1/2</td>\n",
       "      <td>16</td>\n",
       "      <td>566,34,16,40,0,22</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "               REF                           ALT   GT  GQ                   PL\n",
       "1433           TAA                          TA,T  1/2  17  373,20,17,301,0,389\n",
       "1434           TAA                          TA,T  1/2  17  373,20,17,301,0,389\n",
       "2789             G                     GGTGT,GGT  1/2   6    148,32,151,6,0,61\n",
       "2790             G                     GGTGT,GGT  1/2   6    148,32,151,6,0,61\n",
       "3148           GAA                          GA,G  1/2  14  363,14,37,252,0,410\n",
       "3149           GAA                          GA,G  1/2  14  363,14,37,252,0,410\n",
       "4053             T                        TAA,TA  1/2  16  242,16,153,49,0,128\n",
       "4054             T                        TAA,TA  1/2  16  242,16,153,49,0,128\n",
       "4164             G              GAAAAAAA,GAAAAAA  1/2  15    522,72,96,15,0,19\n",
       "4165             G              GAAAAAAA,GAAAAAA  1/2  15    522,72,96,15,0,19\n",
       "5449         GAAAA                          G,GA  1/2  19  431,164,180,57,0,19\n",
       "5450         GAAAA                          G,GA  1/2  19  431,164,180,57,0,19\n",
       "6557         AGTGT                         A,AGT  1/2  16  188,140,211,31,0,16\n",
       "6558         AGTGT                         A,AGT  1/2  16  188,140,211,31,0,16\n",
       "7743           TAA                          TA,T  1/2  16  179,19,16,131,0,192\n",
       "7744           TAA                          TA,T  1/2  16  179,19,16,131,0,192\n",
       "10678  ATGTGTGTGTG                       ATGTG,A  1/2  16  600,43,16,334,0,331\n",
       "10687           TA                         TAA,T  1/2  17  151,17,215,50,0,244\n",
       "10688           TA                         TAA,T  1/2  17  151,17,215,50,0,244\n",
       "11393            C  CTTTTTTTTTTTTT,CTTTTTTTTTTTT  1/2  16    566,34,16,40,0,22\n",
       "11394            C  CTTTTTTTTTTTTT,CTTTTTTTTTTTT  1/2  16    566,34,16,40,0,22"
      ]
     },
     "execution_count": 54,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import numpy as np\n",
    "df.fillna({'GQ': 0}, inplace=True)\n",
    "df = df.astype({'GQ': 'int32'})\n",
    "df.loc[((df['GT'] == '1/2') & (df['GQ'] < 20)), ['REF', 'ALT', 'GT', 'GQ', 'PL']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "#-----   STEP 5. Pre-processing\n",
    "preprocessing = PreProcessExomeSummary(\n",
    "    df=df, mode_samples_info=mode_samples_info)\n",
    "df = preprocessing.all_pre_processing()\n",
    "\n",
    "#-----   STEP 6. Additional annotations\n",
    "anno = Anno(df=df, args=args)\n",
    "df = anno.anno_scores()\n",
    "\n",
    "#-----   STEP 7. Filtering\n",
    "maffilter = MafFilter(\n",
    "    df=df, mode_samples_info=mode_samples_info, config=config)\n",
    "df = maffilter.all_filtering()\n",
    "\n",
    "typefilter = TypeFilter(df=df)\n",
    "df = typefilter.exclude_hlamuc_and_exonicsyno()\n",
    "\n",
    "gtfilter = GtFilter(\n",
    "    df=df, mode_samples_info=mode_samples_info)\n",
    "dfs = gtfilter.genotypeing_filter()\n",
    "\n",
    "#-----   STEP 8. Output as an Excel \n",
    "def df_to_excel(dfs: dataclass, output_xlsx) -> None:\n",
    "    sheet_names = ['AD', 'Homo', 'CH', 'XL']\n",
    "    with pd.ExcelWriter(output_xlsx) as writer:\n",
    "        for df, sheet_name in zip([dfs.AD, dfs.Hm, dfs.CH, dfs.XL], sheet_names):\n",
    "            df.to_excel(writer, sheet_name=sheet_name, index=False)\n",
    "\n",
    "df_to_excel(dfs, f\"{output_file_path}.xlsx\")\n",
    "\n",
    "############################################\n",
    "import pickle\n",
    "with open(f\"{output_file_path}.pkl\", mode=\"wb\") as f:\n",
    "    pickle.dump(dfs, f)\n",
    "############################################\n",
    "\n",
    "#-----   STEP (Final Step). Count variants of filtering process\n",
    "countsummery_file = str(Path(output_file_path).parent) + '/CountSummary.xlsx'\n",
    "counter_result = counter(dfs=dfs, output_excel=countsummery_file)\n",
    "print(counter_result)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'libs.filter.gtfilter.ModelDataFrame'>\n"
     ]
    }
   ],
   "source": [
    "import pickle\n",
    "pkl = '/work/Github/TestData/trio/29881/Sample_29881-trio_results/Sample_29881-trio.tsv.pkl'\n",
    "with open(pkl, mode='rb') as f:\n",
    "    dfs = pickle.load(f)\n",
    "\n",
    "print(type(dfs))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "hgmd_pkl = '/work/resources/HGMD/hgmd_info_2023.3'\n",
    "hgmd = pd.read_pickle(hgmd_pkl)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>altsymbol</th>\n",
       "      <th>refseq</th>\n",
       "      <th>expected_inheritance</th>\n",
       "      <th>hgncID</th>\n",
       "      <th>omimid</th>\n",
       "      <th>DFP</th>\n",
       "      <th>DM</th>\n",
       "      <th>DM?</th>\n",
       "      <th>DP</th>\n",
       "      <th>FP</th>\n",
       "      <th>R</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>RBFOX1</td>\n",
       "      <td>2BP1|A2BP1|FOX-1|FOX1|HRNBP1</td>\n",
       "      <td>NM_145891.3</td>\n",
       "      <td>AD</td>\n",
       "      <td>18222</td>\n",
       "      <td>605104</td>\n",
       "      <td>0.0</td>\n",
       "      <td>15.0</td>\n",
       "      <td>47.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ABCA3</td>\n",
       "      <td>ABC-C|ABC3|EST111653|LBM180|SMDP3</td>\n",
       "      <td>NM_001089.3</td>\n",
       "      <td>AR</td>\n",
       "      <td>33</td>\n",
       "      <td>601615</td>\n",
       "      <td>0.0</td>\n",
       "      <td>230.0</td>\n",
       "      <td>152.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AKAP13</td>\n",
       "      <td>AKAP-13|AKAP-Lbc|ARHGEF13|BRX|c-lbc|HA-3|Ht31|...</td>\n",
       "      <td>NM_007200.5</td>\n",
       "      <td>UNK</td>\n",
       "      <td>371</td>\n",
       "      <td>604686</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>11.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>GSS</td>\n",
       "      <td>GSHS|HEL-S-64p|HEL-S-88n</td>\n",
       "      <td>NM_000178.4</td>\n",
       "      <td>AR</td>\n",
       "      <td>4624</td>\n",
       "      <td>601002</td>\n",
       "      <td>0.0</td>\n",
       "      <td>40.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>BRF1</td>\n",
       "      <td>BRF|BRF-1|CFDS|GTF3B|hBRF|HEL-S-76p|TAF3B2|TAF...</td>\n",
       "      <td>NM_001519.4</td>\n",
       "      <td>ADAR</td>\n",
       "      <td>11551</td>\n",
       "      <td>604902</td>\n",
       "      <td>0.0</td>\n",
       "      <td>19.0</td>\n",
       "      <td>15.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17668</th>\n",
       "      <td>TNFRSF21</td>\n",
       "      <td>BM-018|CD358|DR6</td>\n",
       "      <td>NM_014452.5</td>\n",
       "      <td>UNK</td>\n",
       "      <td>13469</td>\n",
       "      <td>605732</td>\n",
       "      <td>0.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>7.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17669</th>\n",
       "      <td>C12orf4</td>\n",
       "      <td>MRT66</td>\n",
       "      <td>NM_020374.4</td>\n",
       "      <td>AR</td>\n",
       "      <td>1184</td>\n",
       "      <td>616082</td>\n",
       "      <td>0.0</td>\n",
       "      <td>7.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17670</th>\n",
       "      <td>DMBX1</td>\n",
       "      <td>Atx|MBX|OTX3|PAXB</td>\n",
       "      <td>NM_147192.3</td>\n",
       "      <td>UNK</td>\n",
       "      <td>19026</td>\n",
       "      <td>607410</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17671</th>\n",
       "      <td>WDR93</td>\n",
       "      <td>C1d-87|CFAP297|FAP297</td>\n",
       "      <td>NM_020212.2</td>\n",
       "      <td>UNK</td>\n",
       "      <td>26924</td>\n",
       "      <td>None</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17672</th>\n",
       "      <td>ETV6</td>\n",
       "      <td>TEL|TEL/ABL|THC5</td>\n",
       "      <td>NM_001987.5</td>\n",
       "      <td>AD</td>\n",
       "      <td>3495</td>\n",
       "      <td>600618</td>\n",
       "      <td>2.0</td>\n",
       "      <td>39.0</td>\n",
       "      <td>38.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>17673 rows × 12 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           gene                                          altsymbol  \\\n",
       "0        RBFOX1                       2BP1|A2BP1|FOX-1|FOX1|HRNBP1   \n",
       "1         ABCA3                  ABC-C|ABC3|EST111653|LBM180|SMDP3   \n",
       "2        AKAP13  AKAP-13|AKAP-Lbc|ARHGEF13|BRX|c-lbc|HA-3|Ht31|...   \n",
       "3           GSS                           GSHS|HEL-S-64p|HEL-S-88n   \n",
       "4          BRF1  BRF|BRF-1|CFDS|GTF3B|hBRF|HEL-S-76p|TAF3B2|TAF...   \n",
       "...         ...                                                ...   \n",
       "17668  TNFRSF21                                   BM-018|CD358|DR6   \n",
       "17669   C12orf4                                              MRT66   \n",
       "17670     DMBX1                                  Atx|MBX|OTX3|PAXB   \n",
       "17671     WDR93                              C1d-87|CFAP297|FAP297   \n",
       "17672      ETV6                                   TEL|TEL/ABL|THC5   \n",
       "\n",
       "            refseq expected_inheritance hgncID  omimid  DFP     DM    DM?  \\\n",
       "0      NM_145891.3                   AD  18222  605104  0.0   15.0   47.0   \n",
       "1      NM_001089.3                   AR     33  601615  0.0  230.0  152.0   \n",
       "2      NM_007200.5                  UNK    371  604686  0.0    0.0   11.0   \n",
       "3      NM_000178.4                   AR   4624  601002  0.0   40.0    5.0   \n",
       "4      NM_001519.4                 ADAR  11551  604902  0.0   19.0   15.0   \n",
       "...            ...                  ...    ...     ...  ...    ...    ...   \n",
       "17668  NM_014452.5                  UNK  13469  605732  0.0    4.0    7.0   \n",
       "17669  NM_020374.4                   AR   1184  616082  0.0    7.0    4.0   \n",
       "17670  NM_147192.3                  UNK  19026  607410  0.0    1.0    5.0   \n",
       "17671  NM_020212.2                  UNK  26924    None  0.0    1.0    1.0   \n",
       "17672  NM_001987.5                   AD   3495  600618  2.0   39.0   38.0   \n",
       "\n",
       "        DP   FP    R  \n",
       "0      4.0  0.0  0.0  \n",
       "1      1.0  2.0  0.0  \n",
       "2      1.0  5.0  0.0  \n",
       "3      0.0  0.0  0.0  \n",
       "4      0.0  0.0  0.0  \n",
       "...    ...  ...  ...  \n",
       "17668  0.0  0.0  0.0  \n",
       "17669  0.0  0.0  0.0  \n",
       "17670  0.0  0.0  0.0  \n",
       "17671  0.0  0.0  0.0  \n",
       "17672  1.0  0.0  0.0  \n",
       "\n",
       "[17673 rows x 12 columns]"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "genocode_file_hg19 = '/resources/GENCODE_Basic_Annotations/gencode.v44lift37.basic.annotation.gtf.gz'\n",
    "genocode_db_hg19 = '/resources/GENCODE_Basic_Annotations/gencode.v44lift37.basic.annotation.db.bak'\n",
    "\n",
    "db = gffutils.FeatureDB(genocode_db_hg19)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import re\n",
    "\n",
    "@dataclass\n",
    "class GencodeInfo:\n",
    "    gene_name: str\n",
    "    hgnc: str\n",
    "    ensg: str\n",
    "    enst: str\n",
    "    ensg_full: str\n",
    "    enst_full: str\n",
    "    strand: str\n",
    "\n",
    "def __anno_gencode_info(row) -> str:\n",
    "    query_region: str = f\"chr{row['CHROM']}:{row['POS']}-{row['POS']}\"\n",
    "    # query_region: str = f\"chr17:1132706-1132706\"\n",
    "    fetched_data = db.region(region=query_region, \n",
    "                            featuretype='gene',\n",
    "                            completely_within=False)\n",
    "    \n",
    "    result = []\n",
    "    while 1:\n",
    "        try:\n",
    "            data = next(fetched_data)\n",
    "        except StopIteration:\n",
    "            break\n",
    "        else:\n",
    "            strand = data.strand\n",
    "            gene_name = data.attributes['gene_name'][0]\n",
    "            \n",
    "            try:\n",
    "                ensg_full = data.attributes['gene_id'][0]\n",
    "            except KeyError:\n",
    "                ensg = '.'\n",
    "            else:\n",
    "                ensg = re.match(r'ENSG\\d+', ensg_full).group()\n",
    "    \n",
    "    # print(result)\n",
    "    return result"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import re\n",
    "\n",
    "@dataclass\n",
    "class GencodeInfo:\n",
    "    gene_name: str\n",
    "    hgnc: str\n",
    "    ensg: str\n",
    "    enst: str\n",
    "    ensg_full: str\n",
    "    enst_full: str\n",
    "    strand: str\n",
    "\n",
    "def __anno_gencode_info(row) -> str:\n",
    "    query_region: str = f\"chr{row['CHROM']}:{row['POS']}-{row['POS']}\"\n",
    "    # query_region: str = f\"chr17:1132706-1132706\"\n",
    "    fetched_data = db.region(region=query_region, \n",
    "                            featuretype='gene',\n",
    "                            completely_within=False)\n",
    "    \n",
    "    result = []\n",
    "    while 1:\n",
    "        try:\n",
    "            data = next(fetched_data)\n",
    "        except StopIteration:\n",
    "            break\n",
    "        else:\n",
    "            strand = data.strand\n",
    "            gene_name = data.attributes['gene_name'][0]\n",
    "\n",
    "            try:\n",
    "                hgnc_info = data.attributes['hgnc_id'][0]\n",
    "            except KeyError:\n",
    "                hgnc = '.'\n",
    "            else:\n",
    "                hgnc = re.search(r'\\d+', hgnc_info).group()\n",
    "            \n",
    "            try:\n",
    "                ensg_full = data.attributes['gene_id'][0]\n",
    "            except KeyError:\n",
    "                ensg = '.'\n",
    "            else:\n",
    "                ensg = re.match(r'ENSG\\d+', ensg_full).group()\n",
    "            \n",
    "            try:\n",
    "                enst_full = data.attributes['transcript_id'][0]\n",
    "            except KeyError:\n",
    "                enst = '.'\n",
    "            else:\n",
    "                enst = re.match(r'ENST\\d+', enst_full).group()\n",
    "\n",
    "            genecode_info = GencodeInfo(\n",
    "                gene_name, hgnc, ensg, enst, ensg_full, enst_full, strand)\n",
    "            result.append(genecode_info)\n",
    "    \n",
    "    # print(result)\n",
    "    return result\n",
    "        \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "query_region: str = f\"chr17:1132706-1132706\"\n",
    "fetched_data = db.region(region=query_region, \n",
    "                            featuretype='transcript',\n",
    "                            completely_within=False)\n",
    "    \n",
    "result = []\n",
    "while 1:\n",
    "    try:\n",
    "        data = next(fetched_data)\n",
    "    except StopIteration:\n",
    "        break\n",
    "    else:\n",
    "        strand = data.strand\n",
    "        gene_name = data.attributes['gene_name'][0]\n",
    "\n",
    "        try:\n",
    "            hgnc_info = data.attributes['hgnc_id'][0]\n",
    "        except KeyError:\n",
    "            hgnc = '.'\n",
    "        else:\n",
    "            hgnc = re.search(r'\\d+', hgnc_info).group()\n",
    "        \n",
    "        try:\n",
    "            ensg_full = data.attributes['gene_id'][0]\n",
    "        except KeyError:\n",
    "            ensg = '.'\n",
    "        else:\n",
    "            ensg = re.match(r'ENSG\\d+', ensg_full).group()\n",
    "        \n",
    "        try:\n",
    "            enst_full = data.attributes['transcript_id'][0]\n",
    "        except KeyError:\n",
    "            enst = '.'\n",
    "        else:\n",
    "            enst = re.match(r'ENST\\d+', enst_full).group()\n",
    "\n",
    "        genecode_info = GencodeInfo(\n",
    "            gene_name, hgnc, ensg, enst, ensg_full, enst_fuœll, strand)\n",
    "        result.append(genecode_info)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "query_region: str = f\"chr17:1132706-1132706\"\n",
    "fetched_gene = db.region(region=query_region, \n",
    "                         featuretype=['gene', 'transcript'],\n",
    "                         completely_within=False)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "chr17\tHAVANA\tgene\t906759\t1133032\t.\t-\t.\tgene_id \"ENSG00000159842.16_14\"; gene_type \"protein_coding\"; gene_name \"ABR\"; level \"1\"; hgnc_id \"HGNC:81\"; tag \"ncRNA_host\"; havana_gene \"OTTHUMG00000090313.18_14\"; remap_status \"full_contig\"; remap_num_mappings \"1\"; remap_target_status \"overlap\";\n"
     ]
    }
   ],
   "source": [
    "for data in fetched_gene:\n",
    "    print(data)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "gtf = db.children('', featuretype='transcript', order_by='start')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for g in gtf:\n",
    "    print(g)\n",
    "    print(g.attributes['tag'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 6443/6443 [01:25<00:00, 75.63it/s] \n"
     ]
    }
   ],
   "source": [
    "\n",
    "df['gencode'] = df.progress_apply(__anno_gencode_info, axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for row in df.iterrows():\n",
    "    print(row[0], row[1]['Gene.refGene'], row[1]['gencode'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "chr1\tHAVANA\ttranscript\t65419\t71585\t.\t+\t.\tgene_id \"ENSG00000186092.7_9\"; transcript_id \"ENST00000641515.2_5\"; gene_type \"protein_coding\"; gene_name \"OR4F5\"; transcript_type \"protein_coding\"; transcript_name \"OR4F5-201\"; level \"2\"; protein_id \"ENSP00000493376.2\"; hgnc_id \"HGNC:14825\"; tag \"RNA_Seq_supported_partial\"; tag \"basic\"; tag \"Ensembl_canonical\"; tag \"MANE_Select\"; tag \"appris_principal_1\"; tag \"CCDS\"; ccdsid \"CCDS30547.2\"; havana_gene \"OTTHUMG00000001094.4_9\"; havana_transcript \"OTTHUMT00000003223.4_5\"; remap_num_mappings \"1\"; remap_status \"full_contig\"; remap_target_status \"new\";\n",
      "Strand: +\n",
      "ENSG_Full: ENSG00000186092.7_9\n",
      "ENSG     : ENSG00000186092\n",
      "ENST_Full: ENST00000641515.2_5\n",
      "ENST     : ENST00000641515\n",
      "GeneName : OR4F5\n",
      "HGNC_ID  : 14825\n"
     ]
    }
   ],
   "source": [
    "import re\n",
    "\n",
    "query_region: str = f'chr1: 69610-69610'\n",
    "rows = db.region(region=query_region, featuretype='transcript')\n",
    "\n",
    "for row in rows:\n",
    "    print(row)\n",
    "\n",
    "    enst_full = row.attributes['transcript_id'][0]\n",
    "    ensg_full = row.attributes['gene_id'][0]\n",
    "    hgnc_info = row.attributes['hgnc_id'][0]\n",
    "\n",
    "    try:\n",
    "        enst = re.match(r'ENST\\d+', enst_full).group()\n",
    "    except AttributeError:\n",
    "        enst = '.'\n",
    "    try:\n",
    "        ensg = re.match(r'ENSG\\d+', ensg_full).group()\n",
    "    except AttributeError:\n",
    "        ensg = '.'\n",
    "    try:\n",
    "        hgnc = re.search(r'\\d+', hgnc_info).group()\n",
    "    except AttributeError:\n",
    "        hgnc = '.'\n",
    "\n",
    "    print(f\"Strand: {row.strand}\")\n",
    "    print(f\"GeneName : {row.attributes['gene_name'][0]}\")\n",
    "    print(f\"HGNC_ID  : {hgnc}\")\n",
    "    print(f\"ENSG_Full: {ensg_full}\")\n",
    "    print(f\"ENSG     : {ensg}\")\n",
    "    print(f\"ENST_Full: {enst_full}\")\n",
    "    print(f\"ENST     : {enst}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#-----   STEP 5. Pre-processing\n",
    "preprocessing = PreProcessExomeSummary(\n",
    "    df=df, mode_samples_info=mode_samples_info)\n",
    "df = preprocessing.all_pre_processing()\n",
    "\n",
    "#-----   STEP 6. Additional annotations\n",
    "anno = Anno(df=df, args=args)\n",
    "df = anno.anno_scores()\n",
    "\n",
    "#-----   STEP 7. Filtering\n",
    "maffilter = MafFilter(\n",
    "    df=df, mode_samples_info=mode_samples_info, config=config)\n",
    "df = maffilter.all_filtering()\n",
    "\n",
    "typefilter = TypeFilter(df=df)\n",
    "df = typefilter.exclude_hlamuc_and_exonicsyno()\n",
    "\n",
    "gtfilter = GtFilter(\n",
    "    df=df, mode_samples_info=mode_samples_info)\n",
    "dfs = gtfilter.genotypeing_filter()\n",
    "\n",
    "#-----   STEP 9. Output as an Excel \n",
    "def df_to_excel(dfs: dataclass, output_xlsx) -> None:\n",
    "    sheet_names = ['AD', 'Homo', 'CH', 'XL']\n",
    "    with pd.ExcelWriter(output_xlsx) as writer:\n",
    "        for df, sheet_name in zip([dfs.AD, dfs.Hm, dfs.CH, dfs.XL], sheet_names):\n",
    "            df.to_excel(writer, sheet_name=sheet_name, index=False)\n",
    "\n",
    "output_xlsx = f\"{output_file_path}.xlsx\"\n",
    "output_pickle = f\"{output_file_path}.pkl\"\n",
    "df_to_excel(dfs, output_xlsx)\n",
    "dfs.to_pickle(output_pickle)\n",
    "\n",
    "#-----   STEP 8. Count variants of filtering process\n",
    "countsummery_file = str(Path(output_file_path).parent) + '/CountSummary.xlsx'\n",
    "counter_result = counter(dfs=dfs, output_excel=countsummery_file)\n",
    "print(counter_result)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['InHouse_absent_FILTER', 'InHouse_1%_FILTER', 'MAF_0.1%_FILTER',\n",
       "       'MAF_1%_FILTER', 'HLAMUC_FILTER', 'ExonicSyno_FILTER', 'GT_FILTER'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dfs.AD.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/work/Github/TestData/trio/29881/Sample_29881-trio_results/Sample_29881-trio.tsv\n"
     ]
    }
   ],
   "source": [
    "print(output_file_path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#-----   STEP 9. Output as an Excel \n",
    "def df_to_excel(dfs: dataclass, output_xlsx) -> None:\n",
    "    sheet_names = ['AD', 'Homo', 'CH', 'XL']\n",
    "    with pd.ExcelWriter(output_xlsx) as writer:\n",
    "        for df, sheet_name in zip([dfs.AD, dfs.Hm, dfs.CH, dfs.XL], sheet_names):\n",
    "            df.to_excel(writer, sheet_name=sheet_name, index=False)\n",
    "\n",
    "output_xlsx = f\"{output_file_path}.xlsx\"\n",
    "df_to_excel(dfs, output_xlsx)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'Sample_32741-proband_countsummary.txt'"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "hgmd_resource: str = [\n",
    "    str(x) for x in self.resources_dir.glob('HGMD/HGMD_gene_based.tsv.gz')][0]\n",
    "        return pd.read_csv(hgmd_resource, sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "couter_result = counter(dfs=dfs, output_excel=output_file_path)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df = counter(dfs=dfs, output_excel='./test.xlsx')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df.to_pickle('./post7.pkl')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df = pd.read_pickle('./post7.pkl')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df = df.head(100)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#-----   STEP 8. Output as an Excel \n",
    "output_xlsx = './head100.xlsx'\n",
    "def df_to_excel(dfs: dataclass, output_xlsx) -> None:\n",
    "    sheet_names = ['AD', 'Homo', 'CH', 'XL']\n",
    "    with pd.ExcelWriter(output_xlsx) as writer:\n",
    "        for df, sheet_name in zip([dfs.AD, dfs.Hm, dfs.CH, dfs.XL], sheet_names):\n",
    "            df.to_excel(writer, sheet_name=sheet_name, index=False)\n",
    "\n",
    "df_to_excel(dfs, 'output_xlsx')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#----   STEP 9. Insert hyperlinks\n",
    "# 0. 最初にエクセルを読み込む\n",
    "# 1. メモ用の列を作る (1-2列目)\n",
    "# 2. リンク挿入用の列を作る (3-7列目)\n",
    "# 3. リンクの挿入"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Formatng ......\n"
     ]
    }
   ],
   "source": [
    "from libs.excelibs.excel_format import ExcelFormat\n",
    "from libs.excelibs.hyperlinks import HyperLinks\n",
    "\n",
    "excelformat = ExcelFormat('./test1.xlsx')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "excelformat.insert_comment_cols()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "excelformat.insert_hyperlink_cols()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "excelformat.workbook.save('./test2.xlsx')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### CREATE dbs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Create db for gffutils\n",
    "import gffutils\n",
    "import gffutils.pybedtools_integration\n",
    "\n",
    "\n",
    "genocode_file_hg19 = '/resources/GENCODE_Basic_Annotations/gencode.v44lift37.basic.annotation.gtf.gz'\n",
    "genocode_db_hg19 = '/resources/GENCODE_Basic_Annotations/gencode.v44lift37.basic.annotation.db'\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "db = gffutils.create_db(data=genocode_file_hg19, dbfn=genocode_db_hg19, \n",
    "                        disable_infer_genes=True,\n",
    "                        disable_infer_transcripts=True,\n",
    "                        keep_order=True, \n",
    "                        force=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<BedTool(/resources/GENCODE_Basic_Annotations/gencode.v44lift37.basic.annotation.intron.db)>"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genocode_db_intron_hg19 = '/resources/GENCODE_Basic_Annotations/gencode.v44lift37.basic.annotation.intron.db'\n",
    "\n",
    "db = gffutils.FeatureDB(genocode_db_hg19)\n",
    "introns = db.create_introns(exon_featuretype='exon', \n",
    "                            new_featuretype='intron', \n",
    "                            merge_attributes=True, \n",
    "                            numeric_sort=True)\n",
    "pybed = gffutils.pybedtools_integration.to_bedtool(introns)\n",
    "pybed.saveas(genocode_db_intron_hg19)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def generate_intoron_gtf(db: gffutils.FeatureDB, output: str) -> None:\n",
    "    introns = db.create_introns(exon_featuretype='exon', \n",
    "                                new_featuretype='intron', \n",
    "                                merge_attributes=True, \n",
    "                                numeric_sort=True)\n",
    "    pybed = gffutils.pybedtools_integration.to_bedtool(introns)\n",
    "    pybed.saveas(output)\n",
    "    \n",
    "    return"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "wesanno",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.18"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
